// @osdk/client/build/browser/observable/internal/ObservableClientImpl.js
/*
 * Copyright 2025 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Subscription } from "rxjs";
import { UnsubscribableWrapper } from "./UnsubscribableWrapper.js";

/**
 * Implementation of the public ObservableClient interface.
 * - Delegates all operations to the Store for consistency
 * - Serves as the entry point for reactive data management
 * - Ensures proper method binding and API exposure
 *
 * @internal
 */
export class ObservableClientImpl {
  constructor(store) {
    this.__experimentalStore = store;
    this.applyAction = store.applyAction.bind(store);
    this.validateAction = store.validateAction.bind(store);
  }
  observeObject = (apiName, pk, options, subFn) => {
    return this.__experimentalStore.objects.observe({
      ...options,
      apiName,
      pk
    },
    // cast to cross typed to untyped barrier
    subFn);
  };
  observeList = (options, subFn) => {
    return this.__experimentalStore.lists.observe(options,
    // cast to cross typed to untyped barrier
    subFn);
  };
  observeAggregation = (options, subFn) => {
    return this.__experimentalStore.aggregations.observe(options, subFn);
  };
  observeFunction = (queryDef, params, options, subFn) => {
    const dependsOn = options.dependsOn?.map(dep => typeof dep === "string" ? dep : dep.apiName);
    const dependsOnObjects = options.dependsOnObjects?.map(obj => ({
      $apiName: obj.$objectType ?? obj.$apiName,
      $primaryKey: obj.$primaryKey
    }));
    return this.__experimentalStore.functions.observe({
      ...options,
      queryDef,
      params,
      dependsOn,
      dependsOnObjects
    }, subFn);
  };
  observeLinks = (objects, linkName, options, subFn) => {
    const objectsArray = Array.isArray(objects) ? objects : [objects];
    const observer = subFn;
    return objectsArray.length <= 1 ? observeSingleLink(this.__experimentalStore, objectsArray, linkName, options, observer) : observeMultiLinks(this.__experimentalStore, objectsArray, linkName, options, observer);
  };
  observeObjectSet(baseObjectSet, options, subFn) {
    return this.__experimentalStore.objectSets.observe({
      baseObjectSet,
      ...options
    },
    // cast to cross typed to untyped barrier
    subFn);
  }
  invalidateAll() {
    return this.__experimentalStore.invalidateAll();
  }
  invalidateObjects(objects) {
    return this.__experimentalStore.invalidateObjects(objects);
  }
  invalidateObjectType(type) {
    return this.__experimentalStore.invalidateObjectType(type, undefined);
  }
  invalidateFunction(apiName, params) {
    return this.__experimentalStore.invalidateFunction(apiName, params);
  }
  invalidateFunctionsByObject(apiName, primaryKey) {
    return this.__experimentalStore.invalidateFunctionsByObject(apiName, primaryKey);
  }
  canonicalizeWhereClause(where) {
    return this.__experimentalStore.whereCanonicalizer.canonicalize(where);
  }
}
function observeSingleLink(store, objectsArray, linkName, options, observer) {
  if (objectsArray.length === 0) {
    observer.next({
      resolvedList: [],
      isOptimistic: false,
      lastUpdated: 0,
      fetchMore: () => Promise.resolve(),
      hasMore: false,
      status: "loaded",
      totalCount: "0"
    });
    return new UnsubscribableWrapper(new Subscription());
  }
  const parentSub = new Subscription();
  for (const obj of objectsArray) {
    parentSub.add(store.links.observe({
      ...options,
      srcType: {
        type: "object",
        apiName: obj.$objectType ?? obj.$apiName
      },
      linkName,
      pk: obj.$primaryKey
    }, observer));
  }
  return new UnsubscribableWrapper(parentSub);
}
function observeMultiLinks(store, objectsArray, linkName, options, observer) {
  const parentSub = new Subscription();
  const totalExpected = objectsArray.length;
  const perObjectResults = new Map();
  let errored = false;
  function mergeAndEmit() {
    if (errored) {
      return;
    }
    const seen = new Map();
    const fetchMores = [];
    let latestUpdated = 0;
    let hasMore = false;
    let isOptimistic = false;
    for (const payload of perObjectResults.values()) {
      for (const obj of payload.resolvedList) {
        seen.set(`${obj.$objectType}:${obj.$primaryKey}`, obj);
      }
      if (payload.lastUpdated > latestUpdated) {
        latestUpdated = payload.lastUpdated;
      }
      if (payload.isOptimistic) {
        isOptimistic = true;
      }
      if (payload.hasMore) {
        hasMore = true;
        fetchMores.push(payload.fetchMore);
      }
    }
    const payloads = [...perObjectResults.values()];
    const loading = perObjectResults.size < totalExpected || payloads.some(p => p.status === "init" || p.status === "loading");
    observer.next({
      resolvedList: Array.from(seen.values()),
      isOptimistic,
      lastUpdated: latestUpdated,
      fetchMore: hasMore ? () => Promise.all(fetchMores.map(fn => fn())).then(() => {}) : () => Promise.resolve(),
      hasMore,
      status: loading ? "loading" : payloads.some(p => p.status === "error") ? "error" : "loaded",
      ...(!hasMore ? {
        totalCount: String(seen.size)
      } : {})
    });
  }
  for (const obj of objectsArray) {
    const objKey = `${obj.$objectType ?? obj.$apiName}:${obj.$primaryKey}`;
    parentSub.add(store.links.observe({
      ...options,
      srcType: {
        type: "object",
        apiName: obj.$objectType ?? obj.$apiName
      },
      linkName,
      pk: obj.$primaryKey
    }, {
      next: payload => {
        if (errored) {
          return;
        }
        perObjectResults.set(objKey, payload);
        mergeAndEmit();
      },
      error: err => {
        if (errored) {
          return;
        }
        errored = true;
        parentSub.unsubscribe();
        observer.error(err);
      },
      // store link queries are long-lived and do not complete
      complete: () => {}
    }));
  }
  return new UnsubscribableWrapper(parentSub);
}
