// @osdk/client/build/esm/observable/internal/aggregation/AggregationQuery.js
/*
 * Copyright 2025 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { BehaviorSubject, connectable, map } from "rxjs";
import { additionalContext } from "../../../Client.js";
import { Query } from "../Query.js";
import { AGGREGATE_IDX, API_NAME_IDX, RDP_IDX, WHERE_IDX } from "./AggregationCacheKey.js";
export class AggregationQuery extends Query {
  constructor(store, subject, cacheKey, opts) {
    super(store, subject, opts, cacheKey, process.env.NODE_ENV !== "production" ? store.client[additionalContext].logger?.child({}, {
      msgPrefix: `AggregationQuery<${cacheKey.otherKeys.map(x => JSON.stringify(x)).join(", ")}>`
    }) : undefined);
    this.apiName = cacheKey.otherKeys[API_NAME_IDX];
    this.canonicalWhere = cacheKey.otherKeys[WHERE_IDX];
    this.rdpConfig = cacheKey.otherKeys[RDP_IDX];
    this.canonicalAggregate = cacheKey.otherKeys[AGGREGATE_IDX];
  }
  _createConnectable(subject) {
    return connectable(subject.pipe(map(x => {
      return {
        status: x.status,
        result: x.value,
        lastUpdated: x.lastUpdated,
        error: x.status === "error" ? new Error("Aggregation failed") : undefined
      };
    })), {
      connector: () => new BehaviorSubject({
        status: "init",
        result: undefined,
        lastUpdated: 0
      })
    });
  }
  async _fetchAndStore() {
    if (process.env.NODE_ENV !== "production") {
      this.logger?.child({
        methodName: "_fetchAndStore"
      }).debug("calling _fetchAndStore");
    }
    try {
      const result = await this._fetchAggregation();
      this.store.batch({}, batch => {
        this.writeToStore(result, "loaded", batch);
      });
    } catch (err) {
      this.store.batch({}, batch => {
        this.writeToStore(undefined, "error", batch);
      });
    }
  }
  writeToStore(data, status, batch) {
    batch.write(this.cacheKey, data, status);
    batch.changes.modified.add(this.cacheKey);
    return batch.read(this.cacheKey);
  }
  invalidateObjectType = (objectType, changes) => {
    if (this.apiName === objectType) {
      changes?.modified.add(this.cacheKey);
      return this.revalidate(true);
    }
    return Promise.resolve();
  };
}
