// @osdk/foundry.mediasets/build/esm/public/MediaSet.js
/*
 * Copyright 2024 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { foundryPlatformFetch as $foundryPlatformFetch } from "@osdk/shared.net.platformapi";
//
const _abort = [1, "/v2/mediasets/{0}/transactions/{1}/abort", 2];
/**
 * Aborts an open transaction. Items uploaded to the media set during this transaction will be deleted.
 *
 * @beta
 *
 * Required Scopes: [api:mediasets-write]
 * URL: /v2/mediasets/{mediaSetRid}/transactions/{transactionId}/abort
 */
export function abort($ctx, ...args) {
  return $foundryPlatformFetch($ctx, _abort, ...args);
}
const _calculate = [0, "/v2/mediasets/{0}/items/{1}/transform/imagery/thumbnail/calculate", 6];
/**
 * Starts calculation of a thumbnail for a given image.
 *
 * @alpha
 *
 * Required Scopes: [api:mediasets-read]
 * URL: /v2/mediasets/{mediaSetRid}/items/{mediaItemRid}/transform/imagery/thumbnail/calculate
 */
export function calculate($ctx, ...args) {
  return $foundryPlatformFetch($ctx, _calculate, ...args);
}
const _commit = [1, "/v2/mediasets/{0}/transactions/{1}/commit", 2];
/**
 * Commits an open transaction. On success, items uploaded to the media set during this transaction will become available.
 *
 * @beta
 *
 * Required Scopes: [api:mediasets-write]
 * URL: /v2/mediasets/{mediaSetRid}/transactions/{transactionId}/commit
 */
export function commit($ctx, ...args) {
  return $foundryPlatformFetch($ctx, _commit, ...args);
}
const _create = [1, "/v2/mediasets/{0}/transactions", 2];
/**
 * Creates a new transaction. Items uploaded to the media set while this transaction is open will not be reflected until the transaction is committed.
 *
 * @beta
 *
 * Required Scopes: [api:mediasets-write]
 * URL: /v2/mediasets/{mediaSetRid}/transactions
 */
export function create($ctx, ...args) {
  return $foundryPlatformFetch($ctx, _create, ...args);
}
const _info = [0, "/v2/mediasets/{0}/items/{1}", 6];
/**
 * Gets information about the media item.
 *
 * @beta
 *
 * Required Scopes: [api:mediasets-read]
 * URL: /v2/mediasets/{mediaSetRid}/items/{mediaItemRid}
 */
export function info($ctx, ...args) {
  return $foundryPlatformFetch($ctx, _info, ...args);
}
const _reference = [0, "/v2/mediasets/{0}/items/{1}/reference", 6];
/**
 * Gets the [media reference](https://www.palantir.com/docs/foundry/data-integration/media-sets/#media-references) for this media item.
 *
 * @beta
 *
 * Required Scopes: [api:mediasets-read]
 * URL: /v2/mediasets/{mediaSetRid}/items/{mediaItemRid}/reference
 */
export function reference($ctx, ...args) {
  return $foundryPlatformFetch($ctx, _reference, ...args);
}
const _getRidByPath = [0, "/v2/mediasets/{0}/items/getRidByPath", 2];
/**
 * Returns the media item RID for the media item with the specified path.
 *
 * @beta
 *
 * Required Scopes: [api:mediasets-read]
 * URL: /v2/mediasets/{mediaSetRid}/items/getRidByPath
 */
export function getRidByPath($ctx, ...args) {
  return $foundryPlatformFetch($ctx, _getRidByPath, ...args);
}
const _upload = [1, "/v2/mediasets/{0}/items", 3, "*/*"];
/**
 * Uploads a media item to an existing media set.
 * The body of the request must contain the binary content of the file and the `Content-Type` header must be `application/octet-stream`.
 * A branch name, or branch rid, or view rid may optionally be specified.  If none is specified, the item will be uploaded to the default branch. If more than one is specified, an error is thrown.
 *
 * @beta
 *
 * Required Scopes: [api:mediasets-write]
 * URL: /v2/mediasets/{mediaSetRid}/items
 */
export function upload($ctx, ...args) {
  return $foundryPlatformFetch($ctx, _upload, ...args);
}
const _read = [0, "/v2/mediasets/{0}/items/{1}/content", 6,, "*/*"];
/**
 * Gets the content of a media item.
 *
 * @beta
 *
 * Required Scopes: [api:mediasets-read]
 * URL: /v2/mediasets/{mediaSetRid}/items/{mediaItemRid}/content
 */
export function read($ctx, ...args) {
  return $foundryPlatformFetch($ctx, _read, ...args);
}
const _readOriginal = [0, "/v2/mediasets/{0}/items/{1}/original", 6,, "*/*"];
/**
 * Gets the content of an original file uploaded to the media item, even if it was transformed on upload due to being an additional input format.
 *
 * @beta
 *
 * Required Scopes: [api:mediasets-read]
 * URL: /v2/mediasets/{mediaSetRid}/items/{mediaItemRid}/original
 */
export function readOriginal($ctx, ...args) {
  return $foundryPlatformFetch($ctx, _readOriginal, ...args);
}
const _retrieve = [0, "/v2/mediasets/{0}/items/{1}/transform/imagery/thumbnail/retrieve", 6,, "*/*"];
/**
 * Retrieves a successfully calculated thumbnail for a given image.
 *
 * Thumbnails are 200px wide in the format of `image/webp`
 *
 * @alpha
 *
 * Required Scopes: [api:mediasets-read]
 * URL: /v2/mediasets/{mediaSetRid}/items/{mediaItemRid}/transform/imagery/thumbnail/retrieve
 */
export function retrieve($ctx, ...args) {
  return $foundryPlatformFetch($ctx, _retrieve, ...args);
}
const _uploadMedia = [2, "/v2/mediasets/media/upload", 7, "*/*"];
/**
 * Uploads a temporary media item. If the media item isn't persisted within 1 hour, the item will be deleted.
 *
 * If multiple resources are attributed to, usage will be attributed to the first one in the list.
 *
 * The body of the request must contain the binary content of the file and the `Content-Type` header must be `application/octet-stream`.
 * Third-party applications using this endpoint via OAuth2 must request the following operation scopes: `api:ontologies-read api:ontologies-write`.
 *
 * @beta
 *
 * Required Scopes: [api:ontologies-read, api:ontologies-write]
 * URL: /v2/mediasets/media/upload
 */
export function uploadMedia($ctx, ...args) {
  return $foundryPlatformFetch($ctx, _uploadMedia, ...args);
}
