// @osdk/generator-converters/build/cjs/index.cjs
'use strict';

// src/valueFormatting/convertBooleanValueFormattingRule.ts
function convertBooleanFormattingRule(wireFormatting) {
  return {
    type: "boolean",
    valueIfTrue: wireFormatting.valueIfTrue,
    valueIfFalse: wireFormatting.valueIfFalse
  };
}

// src/valueFormatting/commonValueFormattingUtils.ts
function convertPropertyTypeReferenceOrStringConstant(value) {
  switch (value.type) {
    case "constant":
      return {
        type: "constant",
        value: value.value
      };
    case "propertyType":
      return {
        type: "propertyType",
        propertyApiName: value.propertyApiName
      };
    default:
      throw new Error(`Unknown property reference type: ${value.type}`);
  }
}

// src/valueFormatting/convertDateAndTimestampValueFormattingRule.ts
function convertDateFormattingRule(wireFormatting) {
  return {
    type: "date",
    format: convertDatetimeFormat(wireFormatting.format)
  };
}
function convertTimestampFormattingRule(wireFormatting) {
  return {
    type: "timestamp",
    format: convertDatetimeFormat(wireFormatting.format),
    displayTimezone: convertDatetimeTimezone(wireFormatting.displayTimezone)
  };
}
function convertDatetimeFormat(wireFormat) {
  switch (wireFormat.type) {
    case "stringFormat":
      return {
        type: "stringFormat",
        pattern: wireFormat.pattern
      };
    case "localizedFormat":
      return {
        type: "localizedFormat",
        format: wireFormat.format
      };
    default:
      throw new Error(`Unknown datetime format type: ${wireFormat.type}`);
  }
}
function convertDatetimeTimezone(wireTimezone) {
  switch (wireTimezone.type) {
    case "static":
      return {
        type: "static",
        zoneId: convertPropertyTypeReferenceOrStringConstant(wireTimezone.zoneId)
      };
    case "user":
      return {
        type: "user"
      };
    default:
      throw new Error(`Unknown timezone type: ${wireTimezone.type}`);
  }
}

// src/valueFormatting/convertKnownTypeValueFormattingRule.ts
function convertKnownTypeFormattingRule(wireFormatting) {
  switch (wireFormatting.knownType) {
    case "ARTIFACT_GID":
      return {
        type: "knownType",
        knownType: "ARTIFACT_GID"
      };
    case "RESOURCE_RID":
      return {
        type: "knownType",
        knownType: "RESOURCE_RID"
      };
    case "USER_OR_GROUP_ID":
      return {
        type: "knownType",
        knownType: "USER_OR_GROUP_ID"
      };
    default:
      wireFormatting.knownType;
      return void 0;
  }
}

// src/valueFormatting/convertNumberValueFormattingRule.ts
function convertNumberFormattingRule(wireFormatting) {
  const numberType = convertNumberFormattingType(wireFormatting.numberType);
  if (!numberType) {
    return void 0;
  }
  return {
    type: "number",
    numberType
  };
}
function convertNumberFormatOptions(wireOptions) {
  return {
    useGrouping: wireOptions.useGrouping,
    convertNegativeToParenthesis: wireOptions.convertNegativeToParenthesis,
    minimumIntegerDigits: wireOptions.minimumIntegerDigits,
    minimumFractionDigits: wireOptions.minimumFractionDigits,
    maximumFractionDigits: wireOptions.maximumFractionDigits,
    minimumSignificantDigits: wireOptions.minimumSignificantDigits,
    maximumSignificantDigits: wireOptions.maximumSignificantDigits,
    notation: wireOptions.notation,
    roundingMode: wireOptions.roundingMode
  };
}
function convertDurationFormatStyle(formatStyle) {
  switch (formatStyle.type) {
    case "humanReadable":
      return {
        type: "humanReadable",
        showFullUnits: formatStyle.showFullUnits
      };
    case "timecode":
      return {
        type: "timecode"
      };
    default:
      throw new Error(`Unknown duration format style: ${formatStyle.type}`);
  }
}
function convertNumberFormattingType(wireNumberType) {
  switch (wireNumberType.type) {
    case "standard":
      return {
        type: "standard",
        baseFormatOptions: convertNumberFormatOptions(wireNumberType.baseFormatOptions)
      };
    case "fixedValues":
      return {
        type: "fixedValues",
        values: wireNumberType.values
      };
    case "currency": {
      return {
        type: "currency",
        baseFormatOptions: convertNumberFormatOptions(wireNumberType.baseFormatOptions),
        style: wireNumberType.style,
        currencyCode: convertPropertyTypeReferenceOrStringConstant(wireNumberType.currencyCode)
      };
    }
    case "standardUnit": {
      return {
        type: "standardUnit",
        baseFormatOptions: convertNumberFormatOptions(wireNumberType.baseFormatOptions),
        unit: convertPropertyTypeReferenceOrStringConstant(wireNumberType.unit)
      };
    }
    case "customUnit": {
      return {
        type: "customUnit",
        baseFormatOptions: convertNumberFormatOptions(wireNumberType.baseFormatOptions),
        unit: convertPropertyTypeReferenceOrStringConstant(wireNumberType.unit)
      };
    }
    case "affix": {
      return {
        type: "affix",
        baseFormatOptions: convertNumberFormatOptions(wireNumberType.baseFormatOptions),
        affix: {
          prefix: wireNumberType.affix?.prefix != null ? convertPropertyTypeReferenceOrStringConstant(wireNumberType.affix.prefix) : void 0,
          postfix: wireNumberType.affix?.postfix != null ? convertPropertyTypeReferenceOrStringConstant(wireNumberType.affix.postfix) : void 0
        }
      };
    }
    case "duration": {
      return {
        type: "duration",
        formatStyle: convertDurationFormatStyle(wireNumberType.formatStyle),
        precision: wireNumberType.precision,
        baseValue: wireNumberType.baseValue
      };
    }
    case "scale": {
      return {
        type: "scale",
        scaleType: wireNumberType.scaleType,
        baseFormatOptions: convertNumberFormatOptions(wireNumberType.baseFormatOptions)
      };
    }
    case "ratio": {
      return {
        type: "ratio",
        ratioType: wireNumberType.ratioType,
        baseFormatOptions: convertNumberFormatOptions(wireNumberType.baseFormatOptions)
      };
    }
    default:
      throw new Error(`Unknown number format type: ${wireNumberType.type}`);
  }
}

// src/wirePropertyFormattingToSdkFormatting.ts
function wirePropertyFormattingToSdkFormatting(wireFormatting, log) {
  try {
    return convertPropertyFormattingInternal(wireFormatting);
  } catch (error) {
    log?.info(`Failed to convert property formatting rule: ${error}`);
    return void 0;
  }
}
function convertPropertyFormattingInternal(wireFormatting) {
  switch (wireFormatting.type) {
    case "date":
      return convertDateFormattingRule(wireFormatting);
    case "timestamp":
      return convertTimestampFormattingRule(wireFormatting);
    case "boolean":
      return convertBooleanFormattingRule(wireFormatting);
    case "knownType":
      return convertKnownTypeFormattingRule(wireFormatting);
    case "number":
      return convertNumberFormattingRule(wireFormatting);
    default:
      throw new Error(`Unknown formatting type: ${wireFormatting.type}`);
  }
}

// src/wirePropertyV2ToSdkPropertyDefinition.ts
function wirePropertyV2ToSdkPropertyDefinition(input, isNullable = true, log) {
  const sdkPropDefinition = objectPropertyTypeToSdkPropertyDefinition(input.dataType, log);
  if (sdkPropDefinition == null) {
    return void 0;
  }
  switch (input.dataType.type) {
    case "integer":
    case "string":
    case "byte":
    case "decimal":
    case "double":
    case "float":
    case "long":
    case "short":
    case "boolean":
    case "date":
    case "attachment":
    case "mediaReference":
    case "geopoint":
    case "geoshape":
    case "timestamp":
    case "timeseries":
    case "marking":
    case "geotimeSeriesReference":
    case "struct":
    case "vector":
      return {
        displayName: input.displayName,
        multiplicity: false,
        description: input.description,
        type: sdkPropDefinition,
        nullable: input.nullable == null ? isNullable : input.nullable,
        valueTypeApiName: input.valueTypeApiName,
        valueFormatting: input.valueFormatting != null ? wirePropertyFormattingToSdkFormatting(input.valueFormatting, log) : void 0
      };
    case "array": {
      return {
        displayName: input.displayName,
        multiplicity: true,
        description: input.description,
        type: sdkPropDefinition,
        nullable: true,
        valueTypeApiName: input.valueTypeApiName,
        valueFormatting: input.valueFormatting != null ? wirePropertyFormattingToSdkFormatting(input.valueFormatting, log) : void 0
      };
    }
    case "cipherText": {
      log?.info(`${JSON.stringify(input.dataType.type)} is not a supported dataType`);
      return void 0;
    }
    default:
      input.dataType;
      log?.info(`${JSON.stringify(input.dataType)} is not a supported dataType`);
      return void 0;
  }
}
function objectPropertyTypeToSdkPropertyDefinition(propertyType, log) {
  switch (propertyType.type) {
    case "integer":
    case "string":
    case "byte":
    case "decimal":
    case "double":
    case "float":
    case "long":
    case "short":
    case "boolean":
    case "attachment":
    case "geopoint":
    case "geoshape":
    case "timestamp":
    case "marking":
    case "geotimeSeriesReference":
    case "mediaReference":
    case "vector":
      return propertyType.type;
    case "date":
      return "datetime";
    case "array":
      return objectPropertyTypeToSdkPropertyDefinition(propertyType.subType);
    case "timeseries":
      if (propertyType.itemType?.type === "string") {
        return "stringTimeseries";
      } else if (propertyType.itemType?.type === "double") {
        return "numericTimeseries";
      } else return "sensorTimeseries";
    case "struct": {
      return propertyType.structFieldTypes.reduce((structMap, structField) => {
        structMap[structField.apiName] = objectPropertyTypeToSdkPropertyDefinition(structField.dataType);
        return structMap;
      }, {});
    }
    case "cipherText": {
      log?.info(`${JSON.stringify(propertyType.type)} is not a supported propertyType`);
      return void 0;
    }
    default: {
      log?.info(`${JSON.stringify(propertyType)} is not a supported propertyType`);
      return void 0;
    }
  }
}

// src/__UNSTABLE_wireInterfaceTypeV2ToSdkObjectDefinition.ts
function __UNSTABLE_wireInterfaceTypeV2ToSdkObjectDefinition(interfaceType, v2, log) {
  return {
    type: "interface",
    rid: interfaceType.rid,
    apiName: interfaceType.apiName,
    displayName: interfaceType.displayName,
    description: interfaceType.description,
    implements: interfaceType.allExtendsInterfaces ? [...interfaceType.allExtendsInterfaces].sort((a, b) => a.localeCompare(b)) : interfaceType.extendsInterfaces ? [...interfaceType.extendsInterfaces].sort((a, b) => a.localeCompare(b)) : void 0,
    properties: Object.fromEntries(Object.entries(
      // prefer V2 if available and non-empty, otherwise fall back to V1
      interfaceType.allPropertiesV2 && Object.keys(interfaceType.allPropertiesV2).length > 0 ? interfaceType.allPropertiesV2 : interfaceType.allProperties ?? interfaceType.properties
    ).map(([key, value]) => {
      return [key, wirePropertyV2ToSdkPropertyDefinition(value, true, log)];
    }).filter(([_, value]) => value != null)),
    links: Object.fromEntries(Object.entries(interfaceType.allLinks ?? interfaceType.links ?? {}).map(([linkApiName, linkType]) => [linkApiName, {
      multiplicity: linkType.cardinality === "MANY",
      targetTypeApiName: linkType.linkedEntityApiName.apiName,
      targetType: linkType.linkedEntityApiName.type === "objectTypeApiName" ? "object" : "interface"
    }])),
    implementedBy: interfaceType.implementedByObjectTypes ? [...interfaceType.implementedByObjectTypes].sort((a, b) => a.localeCompare(b)) : interfaceType.implementedByObjectTypes
  };
}

// src/getEditedEntities.ts
function getModifiedEntityTypes(action) {
  const addedObjects = /* @__PURE__ */ new Set();
  const modifiedObjects = /* @__PURE__ */ new Set();
  for (const operation of action.operations) {
    switch (operation.type) {
      case "createObject":
        addedObjects.add(operation.objectTypeApiName);
        break;
      case "modifyObject":
        modifiedObjects.add(operation.objectTypeApiName);
        break;
    }
  }
  return {
    addedObjects,
    modifiedObjects
  };
}

// src/wirePropertyV2ToSdkPrimaryKeyTypeDefinition.ts
function wirePropertyV2ToSdkPrimaryKeyTypeDefinition(input) {
  switch (input.dataType.type) {
    case "integer":
    case "double":
    case "string":
    case "byte":
    case "long":
    case "short":
    case "timestamp": {
      return input.dataType.type;
    }
    case "date": {
      return "datetime";
    }
    case "boolean":
    case "geopoint":
    case "geoshape":
    case "decimal":
    case "attachment":
    case "timeseries":
    case "array":
    case "marking":
    case "float":
    case "geotimeSeriesReference":
    case "mediaReference":
    case "struct":
    case "cipherText":
    case "vector":
      throw new Error(`Primary key of type ${input.dataType.type} is not supported`);
    default:
      input.dataType;
      throw new Error(`Unknown type encountered for primaryKey: ${input.dataType}`);
  }
}

// src/wireObjectTypeFullMetadataToSdkObjectMetadata.ts
function wireObjectTypeFullMetadataToSdkObjectMetadata(objectTypeWithLink, v2, log) {
  if (objectTypeWithLink.objectType.properties[objectTypeWithLink.objectType.primaryKey] === void 0) {
    throw new Error(`Primary key ${objectTypeWithLink.objectType.primaryKey} not found in ${objectTypeWithLink.objectType.apiName}`);
  }
  if (objectTypeWithLink.implementsInterfaces2 == null && objectTypeWithLink.implementsInterfaces != null) {
    throw new Error("Your ontology.json file is missing the implementsInterfaces2 field. Please regenerate it.");
  }
  const interfaceMap = objectTypeWithLink.implementsInterfaces2 ? Object.fromEntries(Object.entries(objectTypeWithLink.implementsInterfaces2).map(([interfaceApiName, impl]) => {
    if (impl.propertiesV2 && Object.keys(impl.propertiesV2).length > 0) {
      const propMap = {};
      for (const [iptApiName, implementation] of Object.entries(impl.propertiesV2)) {
        if (implementation.type === "localPropertyImplementation") {
          propMap[iptApiName] = implementation.propertyApiName;
        }
      }
      return [interfaceApiName, propMap];
    }
    return [interfaceApiName, impl.properties];
  })) : {};
  return {
    type: "object",
    apiName: objectTypeWithLink.objectType.apiName,
    description: objectTypeWithLink.objectType.description,
    primaryKeyApiName: objectTypeWithLink.objectType.primaryKey,
    primaryKeyType: wirePropertyV2ToSdkPrimaryKeyTypeDefinition(objectTypeWithLink.objectType.properties[objectTypeWithLink.objectType.primaryKey]),
    links: Object.fromEntries([...objectTypeWithLink.linkTypes].sort((a, b) => a.apiName.localeCompare(b.apiName)).map((linkType) => {
      return [linkType.apiName, {
        multiplicity: linkType.cardinality === "MANY",
        targetType: linkType.objectTypeApiName
      }];
    })),
    properties: Object.fromEntries(Object.entries(objectTypeWithLink.objectType.properties).map(([key, value]) => [key, wirePropertyV2ToSdkPropertyDefinition(value, !(v2 && objectTypeWithLink.objectType.primaryKey === key), log)]).filter(([_, value]) => value != null)),
    implements: objectTypeWithLink.implementsInterfaces ? [...objectTypeWithLink.implementsInterfaces].sort((a, b) => a.localeCompare(b)) : objectTypeWithLink.implementsInterfaces,
    interfaceMap,
    inverseInterfaceMap: Object.fromEntries(Object.entries(interfaceMap).map(([interfaceApiName, props]) => [interfaceApiName, invertProps(props)])),
    icon: supportedIconTypes.includes(objectTypeWithLink.objectType.icon.type) ? objectTypeWithLink.objectType.icon : void 0,
    titleProperty: objectTypeWithLink.objectType.titleProperty,
    displayName: objectTypeWithLink.objectType.displayName,
    pluralDisplayName: objectTypeWithLink.objectType.pluralDisplayName,
    status: ensureStringEnumSupportedOrUndefined(objectTypeWithLink.objectType.status, supportedReleaseStatus),
    rid: objectTypeWithLink.objectType.rid,
    visibility: ensureStringEnumSupportedOrUndefined(objectTypeWithLink.objectType.visibility, supportedObjectTypeVisibility)
  };
}
function invertProps(a) {
  return a ? Object.fromEntries(Object.entries(a).map(([k, v]) => [v, k])) : void 0;
}
var supportedIconTypes = ["blueprint"];
var supportedReleaseStatus = ["ACTIVE", "EXPERIMENTAL", "DEPRECATED", "ENDORSED"];
var supportedObjectTypeVisibility = ["NORMAL", "PROMINENT", "HIDDEN"];
function ensureStringEnumSupportedOrUndefined(value, supportedValues) {
  return value && supportedValues.includes(value) ? value : void 0;
}

// src/wireActionTypeV2ToSdkActionMetadata.ts
function wireActionTypeV2ToSdkActionMetadata(input) {
  const modifiedEntityTypes = getModifiedEntityTypes(input);
  return {
    type: "action",
    apiName: input.apiName,
    parameters: Object.fromEntries(Object.entries(input.parameters).map(([key, value]) => [key, wireActionParameterV2ToSdkParameterDefinition(value)])),
    displayName: input.displayName,
    description: input.description,
    modifiedEntities: createModifiedEntities(modifiedEntityTypes.addedObjects, modifiedEntityTypes.modifiedObjects),
    rid: input.rid,
    status: ensureStringEnumSupportedOrUndefined(input.status, supportedReleaseStatus)
  };
}
function wireActionParameterV2ToSdkParameterDefinition(value) {
  return {
    multiplicity: value.dataType.type === "array",
    type: actionPropertyToSdkPropertyDefinition(value.dataType.type === "array" ? value.dataType.subType : value.dataType),
    nullable: !value.required,
    description: value.description
  };
}
function actionPropertyToSdkPropertyDefinition(parameterType) {
  switch (parameterType.type) {
    case "string":
    case "boolean":
    case "attachment":
    case "double":
    case "integer":
    case "long":
    case "timestamp":
    case "mediaReference":
    case "marking":
    case "objectType":
    case "geohash":
    case "geoshape":
      return parameterType.type;
    case "date":
      return "datetime";
    case "objectSet":
      return {
        type: "objectSet",
        objectSet: parameterType.objectTypeApiName
      };
    case "object":
      return {
        type: "object",
        object: parameterType.objectTypeApiName
      };
    case "array":
      return actionPropertyToSdkPropertyDefinition(parameterType.subType);
    case "interfaceObject":
      return {
        type: "interface",
        interface: parameterType.interfaceTypeApiName
      };
    case "struct":
      return {
        type: "struct",
        struct: parameterType.fields.reduce((structMap, structField) => {
          structMap[structField.name] = actionPropertyToSdkPropertyDefinition(structField.fieldType);
          return structMap;
        }, {})
      };
    default:
      throw new Error(`Unsupported action parameter type: ${JSON.stringify(parameterType)}`);
  }
}
function createModifiedEntities(addedObjects, modifiedObjects) {
  const entities = {};
  for (const key of addedObjects) {
    entities[key] = {
      created: true,
      modified: false
    };
  }
  for (const key of modifiedObjects) {
    if (entities[key]) {
      entities[key].modified = true;
    } else {
      entities[key] = {
        created: false,
        modified: true
      };
    }
  }
  return entities;
}

// src/isNullableQueryDataType.ts
function isNullableQueryDataType(input) {
  if (input.type === "null") {
    return true;
  }
  if (input.type === "union") {
    return input.unionTypes.some((t) => isNullableQueryDataType(t));
  }
  return false;
}

// src/wireQueryDataTypeToQueryDataTypeDefinition.ts
function wireQueryDataTypeToQueryDataTypeDefinition(input) {
  switch (input.type) {
    case "double":
    case "float":
    case "integer":
    case "long":
    case "attachment":
    case "boolean":
    case "date":
    case "string":
    case "timestamp":
      return {
        type: input.type,
        nullable: false
      };
    case "object":
      return {
        type: "object",
        object: input.objectTypeApiName,
        nullable: false
      };
    case "objectSet":
      return {
        type: "objectSet",
        objectSet: input.objectTypeApiName,
        nullable: false
      };
    case "interfaceObject":
      return {
        type: "interface",
        interface: input.interfaceTypeApiName,
        nullable: false
      };
    case "interfaceObjectSet":
      return {
        type: "interfaceObjectSet",
        objectSet: input.interfaceTypeApiName,
        nullable: false
      };
    case "array":
      return {
        array: wireQueryDataTypeToQueryDataTypeDefinition(input.subType),
        type: "array",
        nullable: isNullableQueryDataType(input.subType)
      };
    case "set":
      return {
        type: "set",
        set: wireQueryDataTypeToQueryDataTypeDefinition(input.subType),
        nullable: false
      };
    case "union":
      const allowNulls = isNullableQueryDataType(input);
      if (allowNulls && input.unionTypes.length === 2) {
        const nonNull = input.unionTypes.find((t) => t.type != null);
        if (nonNull) {
          return {
            ...wireQueryDataTypeToQueryDataTypeDefinition(nonNull),
            nullable: true
          };
        }
      }
      return {
        type: "union",
        union: input.unionTypes.reduce((acc, t) => {
          if (t.type === "null") {
            return acc;
          }
          acc.push(wireQueryDataTypeToQueryDataTypeDefinition(t));
          return acc;
        }, []),
        nullable: allowNulls
      };
    case "struct":
      return {
        type: "struct",
        struct: Object.fromEntries(input.fields.map((f) => [f.name, wireQueryDataTypeToQueryDataTypeDefinition(f.fieldType)])),
        nullable: false
      };
    case "twoDimensionalAggregation":
      return {
        type: "twoDimensionalAggregation",
        twoDimensionalAggregation: get2DQueryAggregationProps(input),
        nullable: false
      };
    case "threeDimensionalAggregation":
      return {
        type: "threeDimensionalAggregation",
        threeDimensionalAggregation: get3DQueryAggregationProps(input),
        nullable: false
      };
    case "entrySet":
      const keyType = wireQueryDataTypeToQueryDataTypeDefinition(input.keyType);
      if (!validMapKeyTypes.includes(keyType.type)) {
        throw new Error("Map types with a key type of " + keyType.type + " are not supported" + validMapKeyTypes.toString());
      }
      if (keyType.type === "array") {
        throw new Error("Map types cannot have keys as arrays");
      }
      return {
        type: "map",
        nullable: false,
        keyType,
        valueType: wireQueryDataTypeToQueryDataTypeDefinition(input.valueType)
      };
    case "null":
    case "unsupported":
      throw new Error(`Unable to process query because the server indicated an unsupported QueryDataType.type: ${input.type}. Please check that your query is using supported types.`);
    default:
      throw new Error(`Unsupported QueryDataType.type ${input.type}`);
  }
}
function get2DQueryAggregationProps(input) {
  if (input.keyType.type === "range") {
    return {
      keyType: input.keyType.type,
      keySubtype: input.keyType.subType.type,
      valueType: input.valueType.type
    };
  } else {
    if (guardInvalidKeyTypes(input.keyType)) {
      return {
        keyType: input.keyType.type,
        valueType: input.valueType.type
      };
    }
    throw new Error(`Cannot create 2D aggregation with ${input.keyType.type} as its type`);
  }
}
function get3DQueryAggregationProps(input) {
  if (input.keyType.type === "range") {
    return {
      keyType: input.keyType.type,
      keySubtype: input.keyType.subType.type,
      valueType: get2DQueryAggregationProps(input.valueType)
    };
  } else {
    if (guardInvalidKeyTypes(input.keyType)) {
      return {
        keyType: input.keyType.type,
        valueType: get2DQueryAggregationProps(input.valueType)
      };
    }
    throw new Error(`Cannot create 3D aggregation with ${input.keyType.type} as its type`);
  }
}
function guardInvalidKeyTypes(key) {
  return key.type === "string" || key.type === "boolean";
}
var validMapKeyTypes = ["string", "object", "double", "float", "integer", "long", "date", "timestamp", "byte", "datetime", "decimal", "marking", "short", "objectType"];

// src/wireQueryTypeV2ToSdkQueryMetadata.ts
function wireQueryTypeV2ToSdkQueryMetadata(input) {
  return {
    type: "query",
    apiName: input.apiName,
    description: input.description,
    displayName: input.displayName,
    version: input.version,
    parameters: Object.fromEntries(Object.entries(input.parameters).map(([name, parameter]) => [name, wireQueryParameterV2ToQueryParameterDefinition(parameter)])),
    output: wireQueryDataTypeToQueryDataTypeDefinition(input.output),
    rid: input.rid
  };
}
function wireQueryTypeV2ToSdkQueryDefinitionNoParams(input) {
  return {
    type: "query",
    apiName: input.apiName,
    description: input.description,
    displayName: input.displayName,
    version: input.version,
    rid: input.rid
  };
}
function wireQueryParameterV2ToQueryParameterDefinition(parameter) {
  return {
    description: parameter.description,
    ...wireQueryDataTypeToQueryDataTypeDefinition(parameter.dataType)
  };
}

exports.__UNSTABLE_wireInterfaceTypeV2ToSdkObjectDefinition = __UNSTABLE_wireInterfaceTypeV2ToSdkObjectDefinition;
exports.wireActionTypeV2ToSdkActionMetadata = wireActionTypeV2ToSdkActionMetadata;
exports.wireObjectTypeFullMetadataToSdkObjectMetadata = wireObjectTypeFullMetadataToSdkObjectMetadata;
exports.wirePropertyV2ToSdkPropertyDefinition = wirePropertyV2ToSdkPropertyDefinition;
exports.wireQueryDataTypeToQueryDataTypeDefinition = wireQueryDataTypeToQueryDataTypeDefinition;
exports.wireQueryParameterV2ToQueryParameterDefinition = wireQueryParameterV2ToQueryParameterDefinition;
exports.wireQueryTypeV2ToSdkQueryDefinitionNoParams = wireQueryTypeV2ToSdkQueryDefinitionNoParams;
exports.wireQueryTypeV2ToSdkQueryMetadata = wireQueryTypeV2ToSdkQueryMetadata;
