// @osdk/workshop-iframe-custom-widget/build/types/loadingState.js
/**
Copyright 2024 Palantir Technologies, Inc.

Licensed under Palantir's License;
you may not use this file except in compliance with the License.
You may obtain a copy of the License from the root of this repository at LICENSE.md

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
 */
/**
 * Helper function for creating "not started" async loading state.
 */
export function asyncValueNotStarted() {
    return {
        status: undefined,
    };
}
/**
 * Helper function for creating "loading" async loading state.
 */
export function asyncValueLoading() {
    return {
        status: "LOADING",
    };
}
/**
 * Helper function for creating the "loaded" async loading state.
 *
 * @param value: The loaded value
 * @type V: The value type
 */
export function asyncValueLoaded(value) {
    return {
        status: "LOADED",
        value,
    };
}
/**
 * Helper function for creating the "failed" async loading state.
 *
 * @param error: The error associated with the reason behind the failure
 * @type E: The error type
 */
export function asyncValueFailed(error) {
    return {
        status: "FAILED",
        error,
    };
}
/**
 * Helper function for creating the "failed" async loading state.
 *
 * @param value: The previously loaded value
 * @param progress: The loading progress percentage
 * @type V: The value type
 */
export function asyncValueReloading(value, progress) {
    return {
        progress,
        status: "RELOADING",
        value,
    };
}
/**
 * Type guard for "not started" async loaded state.
 *
 * @param state: The async loaded state to type check
 * @type V: The value type
 * @type E: The error type
 * @returns true only if the asyncloaded state is "not started".
 */
export function isAsyncValue_NotStarted(state) {
    return (state == null || state.status == null || state.status === "NOT_STARTED");
}
/**
 * Type guard for "loading" async loaded state.
 *
 * @param state: The async loaded state to type check
 * @type V: The value type
 * @type E: The error type
 * @returns true only if the asyncloaded state is "loading".
 */
export function isAsyncValue_Loading(state) {
    return state != null && state.status === "LOADING";
}
/**
 * Type guard for "loaded" async loaded state.
 *
 * @param state: The async loaded state to type check
 * @type V: The value type
 * @type E: The error type
 * @returns true only if the asyncloaded state is "loaded".
 */
export function isAsyncValue_Loaded(state) {
    return state != null && state.status === "LOADED";
}
/**
 * Type guard for "reloading" async loaded state.
 *
 * @param state: The async loaded state to type check
 * @type V: The value type
 * @type E: The error type
 * @returns true only if the asyncloaded state is "reloading".
 */
export function isAsyncValue_Reloading(state) {
    return state != null && state.status === "RELOADING";
}
/**
 * Type guard for "failed loading" async loaded state.
 *
 * @param state: The async loaded state to type check
 * @type V: The value type
 * @type E: The error type
 * @returns true only if the asyncloaded state is "failed loading".
 */
export function isAsyncValue_FailedLoading(state) {
    return state != null && state.status === "FAILED";
}
/**
 * A visitor function to visit an IAsyncValue
 */
export function visitLoadingState(state, visitor) {
    if (isAsyncValue_Loading(state) || isAsyncValue_NotStarted(state)) {
        return visitor.loading();
    }
    else if (isAsyncValue_Reloading(state)) {
        return visitor.reloading(state.value);
    }
    else if (isAsyncValue_FailedLoading(state)) {
        return visitor.failed(state.error);
    }
    // state.status === "LOADED"
    return visitor.succeeded(state.value);
}
