// @osdk/client/build/browser/ontology/StandardOntologyProvider.js
/*
 * Copyright 2024 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { createAsyncClientCache } from "../object/Cache.js";
import { deepFreeze } from "../util/deepFreeze.js";
import { loadActionMetadata } from "./loadActionMetadata.js";
import { loadFullObjectMetadata } from "./loadFullObjectMetadata.js";
import { loadInterfaceMetadata } from "./loadInterfaceMetadata.js";
import { loadQueryMetadata } from "./loadQueryMetadata.js";
import { InterfaceDefinitions } from "./OntologyProvider.js";
export const createStandardOntologyProviderFactory = () => {
  return client => {
    function makeGetter(fn) {
      const cache = createAsyncClientCache((client, key) => fn(client, key, false));
      return async apiName => {
        return await cache.get(client, apiName);
      };
    }
    const ret = {
      getObjectDefinition: makeGetter(async function (client, key) {
        const objectDef = await loadFullObjectMetadata(client, key);

        // ensure we have all of the interfaces loaded
        const interfaceDefs = Object.fromEntries((await Promise.all(objectDef.implements?.map(i => ret.getInterfaceDefinition(i)) ?? [])).map(i => [i.apiName, {
          def: i,
          handler: undefined
        }]));
        const fullObjectDef = {
          ...objectDef,
          [InterfaceDefinitions]: interfaceDefs
        };
        return deepFreeze(fullObjectDef);
      }),
      getInterfaceDefinition: makeGetter(async function (client, key) {
        return deepFreeze(await loadInterfaceMetadata(client, key));
      }),
      getActionDefinition: makeGetter(async function (client, key) {
        const r = await loadActionMetadata(client, key);
        return r;
      }),
      getQueryDefinition: function (client, fn) {
        const queryCache = createAsyncClientCache((client, key) => {
          return fn(client, key);
        });
        return async (apiName, version) => {
          const key = version ? `${apiName}:${version}` : apiName;
          return await queryCache.get(client, key);
        };
      }(client, async function (client, key) {
        return loadQueryMetadata(client, key);
      })
    };
    return ret;
  };
};
