// @osdk/client/build/cjs/public/unstable-do-not-use.cjs
'use strict';

var chunkO3SKYLBQ_cjs = require('../chunk-O3SKYLBQ.cjs');
var chunkEHWB7KEN_cjs = require('../chunk-EHWB7KEN.cjs');
var chunkLDTMSHUZ_cjs = require('../chunk-LDTMSHUZ.cjs');
var trie = require('@wry/trie');
var deepEqual = require('fast-deep-equal');
var invariant5 = require('tiny-invariant');
var shared_net_fetch = require('@osdk/shared.net.fetch');
var rxjs = require('rxjs');
var mnemonist = require('mnemonist');
var groupBy = require('object.groupby');
var shared_net_errors = require('@osdk/shared.net.errors');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var deepEqual__default = /*#__PURE__*/_interopDefault(deepEqual);
var invariant5__default = /*#__PURE__*/_interopDefault(invariant5);
var groupBy__default = /*#__PURE__*/_interopDefault(groupBy);

// src/public/unstable-do-not-use.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/computeObjectSetCacheKey.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/OrderByCanonicalizer.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/WeakRefTrie.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
var defaultMakeData = () => /* @__PURE__ */ Object.create(null);
var WeakRefTrie = class {
  #finalizer = new FinalizationRegistry((orderBy) => {
    this.#trie.removeArray(Object.entries(orderBy).flat());
  });
  #trie;
  constructor(makeData = defaultMakeData) {
    this.#trie = new trie.Trie(false, (array) => {
      const data = makeData(array);
      this.#finalizer.register(data, array);
      return new WeakRef(data);
    });
  }
  lookupArray(array) {
    const maybe = this.#trie.lookupArray(array);
    let ret = maybe.deref();
    if (maybe && !ret) {
      this.#trie.removeArray(array);
      ret = this.#trie.lookupArray(array).deref();
    }
    return ret;
  }
  peekArray(array) {
    const maybe = this.#trie.peekArray(array);
    const ret = maybe?.deref();
    if (maybe && !ret) {
      this.#trie.removeArray(array);
    }
    return ret;
  }
  removeArray(array) {
    return this.#trie.removeArray(array)?.deref();
  }
};

// src/observable/internal/OrderByCanonicalizer.ts
var OrderByCanonicalizer = class {
  #trie = new WeakRefTrie((array) => {
    const pairs = array.reduce(function(result, _, index, array2) {
      if (index % 2 === 0 && array2[index] != null) {
        result.push(array2.slice(index, index + 2));
      }
      return result;
    }, []);
    const data = Object.fromEntries(pairs);
    return data;
  });
  canonicalize = (orderBy) => {
    const strings = Object.entries(orderBy).flat();
    return this.#trie.lookupArray(strings);
  };
};

// src/observable/internal/WhereClauseCanonicalizer.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
var WhereClauseCanonicalizer = class {
  /**
   * This is a shortcut cache for any WhereClause's that we have
   * seen and already canonicalized. The theory behind this
   * is that well behaving React applications will either `useMemo`
   * their where clause, or store it in state or pass it through as
   * props such that we are likely to get the same WhereClause
   * object multiple times and we can skip unnecessary work.
   */
  #cache = /* @__PURE__ */ new WeakMap();
  /**
   * This is a trie that stores the sorted collapsed keys of a where clause to
   * the cache key for canonicalized options. In theory this keeps the number of
   * deepEqual comparisons down to a minimum but its probably overkill.
   */
  #trie = new trie.Trie();
  /**
   * This is a cache from the cacheKey provided by `this.#trie` to the potential
   * canonicalized options.
   */
  #existingOptions = /* @__PURE__ */ new Map();
  canonicalize(where) {
    if (this.#cache.has(where)) {
      return this.#cache.get(where);
    }
    const keysSet = /* @__PURE__ */ new Set();
    const calculatedCanon = this.#toCanon(where, keysSet);
    const cacheKey = this.#trie.lookupArray(Array.from(keysSet).sort());
    const lookupEntry = this.#existingOptions.get(cacheKey) ?? {
      options: []
    };
    this.#existingOptions.set(cacheKey, lookupEntry);
    const canon = lookupEntry.options.find((ref) => deepEqual__default.default(ref.deref(), calculatedCanon))?.deref() ?? calculatedCanon;
    if (canon === calculatedCanon) {
      lookupEntry.options.push(new WeakRef(canon));
    }
    this.#cache.set(where, canon);
    return canon;
  }
  #toCanon = (where, set = /* @__PURE__ */ new Set()) => {
    if ("$and" in where) {
      if (process.env.NODE_ENV !== "production") {
        !Array.isArray(where.$and) ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "expected $and to be an array") : invariant5__default.default(false) : void 0;
        !(Object.keys(where).length === 1) ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "expected only $and to be present") : invariant5__default.default(false) : void 0;
      }
      if (where.$and.length === 0) {
        return {};
      }
      if (where.$and.length === 1) {
        return this.#toCanon(where.$and[0], set);
      }
    }
    return Object.fromEntries(Object.entries(where).sort(([a], [b]) => a.localeCompare(b)).map(([k, v]) => {
      set.add(k);
      if (k === "$and" || k === "$or") {
        return [k, v.map((x) => this.#toCanon(x, set))];
      }
      if (k !== "$not" && typeof v === "object" && v != null && "$eq" in v) {
        return [k, v["$eq"]];
      }
      return [k, v];
    }));
  };
};

// src/observable/computeObjectSetCacheKey.ts
var whereCanonicalizer = new WhereClauseCanonicalizer();
var orderByCanonicalizer = new OrderByCanonicalizer();
function computeObjectSetCacheKey(objectSet, options) {
  const keyParts = ["objectSet", chunkEHWB7KEN_cjs.getWireObjectSet(objectSet)];
  if (!options) {
    return JSON.stringify(keyParts);
  }
  if (options.where) {
    keyParts.push("where", whereCanonicalizer.canonicalize(options.where));
  }
  if (options.withProperties) {
    const propKeys = Object.keys(options.withProperties).sort();
    keyParts.push("props", propKeys);
  }
  if (options.union && options.union.length > 0) {
    keyParts.push("union", options.union.map((os) => chunkEHWB7KEN_cjs.getWireObjectSet(os)));
  }
  if (options.intersect && options.intersect.length > 0) {
    keyParts.push("intersect", options.intersect.map((os) => chunkEHWB7KEN_cjs.getWireObjectSet(os)));
  }
  if (options.subtract && options.subtract.length > 0) {
    keyParts.push("subtract", options.subtract.map((os) => chunkEHWB7KEN_cjs.getWireObjectSet(os)));
  }
  if (options.pivotTo) {
    keyParts.push("pivotTo", options.pivotTo);
  }
  if (options.orderBy) {
    keyParts.push("orderBy", orderByCanonicalizer.canonicalize(options.orderBy));
  }
  if (options.pageSize) {
    keyParts.push("pageSize", options.pageSize);
  }
  return JSON.stringify(keyParts);
}

// src/observable/ObservableClient.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/ObservableClientImpl.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/UnsubscribableWrapper.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
var UnsubscribableWrapper = class {
  #subscription;
  constructor(subscription) {
    this.#subscription = subscription;
  }
  unsubscribe() {
    this.#subscription?.unsubscribe();
  }
};

// src/observable/internal/ObservableClientImpl.ts
var ObservableClientImpl = class {
  constructor(store) {
    this.__experimentalStore = store;
    this.applyAction = store.applyAction.bind(store);
    this.validateAction = store.validateAction.bind(store);
  }
  observeObject = (apiName, pk, options, subFn) => {
    return this.__experimentalStore.objects.observe(
      {
        ...options,
        apiName,
        pk
      },
      // cast to cross typed to untyped barrier
      subFn
    );
  };
  observeList = (options, subFn) => {
    return this.__experimentalStore.lists.observe(
      options,
      // cast to cross typed to untyped barrier
      subFn
    );
  };
  observeAggregation = (options, subFn) => {
    return this.__experimentalStore.aggregations.observe(options, subFn);
  };
  observeFunction = (queryDef, params, options, subFn) => {
    const dependsOn = options.dependsOn?.map((dep) => typeof dep === "string" ? dep : dep.apiName);
    const dependsOnObjects = options.dependsOnObjects?.map((obj) => ({
      $apiName: obj.$objectType ?? obj.$apiName,
      $primaryKey: obj.$primaryKey
    }));
    return this.__experimentalStore.functions.observe({
      ...options,
      queryDef,
      params,
      dependsOn,
      dependsOnObjects
    }, subFn);
  };
  observeLinks = (objects, linkName, options, subFn) => {
    const objectsArray = Array.isArray(objects) ? objects : [objects];
    const observer = subFn;
    return objectsArray.length <= 1 ? observeSingleLink(this.__experimentalStore, objectsArray, linkName, options, observer) : observeMultiLinks(this.__experimentalStore, objectsArray, linkName, options, observer);
  };
  observeObjectSet(baseObjectSet, options, subFn) {
    return this.__experimentalStore.objectSets.observe(
      {
        baseObjectSet,
        ...options
      },
      // cast to cross typed to untyped barrier
      subFn
    );
  }
  invalidateAll() {
    return this.__experimentalStore.invalidateAll();
  }
  invalidateObjects(objects) {
    return this.__experimentalStore.invalidateObjects(objects);
  }
  invalidateObjectType(type) {
    return this.__experimentalStore.invalidateObjectType(type, void 0);
  }
  invalidateFunction(apiName, params) {
    return this.__experimentalStore.invalidateFunction(apiName, params);
  }
  invalidateFunctionsByObject(apiName, primaryKey) {
    return this.__experimentalStore.invalidateFunctionsByObject(apiName, primaryKey);
  }
  canonicalizeWhereClause(where) {
    return this.__experimentalStore.whereCanonicalizer.canonicalize(where);
  }
};
function observeSingleLink(store, objectsArray, linkName, options, observer) {
  if (objectsArray.length === 0) {
    observer.next({
      resolvedList: [],
      isOptimistic: false,
      lastUpdated: 0,
      fetchMore: () => Promise.resolve(),
      hasMore: false,
      status: "loaded",
      totalCount: "0"
    });
    return new UnsubscribableWrapper(new rxjs.Subscription());
  }
  const parentSub = new rxjs.Subscription();
  for (const obj of objectsArray) {
    parentSub.add(store.links.observe({
      ...options,
      srcType: {
        type: "object",
        apiName: obj.$objectType ?? obj.$apiName
      },
      linkName,
      pk: obj.$primaryKey
    }, observer));
  }
  return new UnsubscribableWrapper(parentSub);
}
function observeMultiLinks(store, objectsArray, linkName, options, observer) {
  const parentSub = new rxjs.Subscription();
  const totalExpected = objectsArray.length;
  const perObjectResults = /* @__PURE__ */ new Map();
  let errored = false;
  function mergeAndEmit() {
    if (errored) {
      return;
    }
    const seen = /* @__PURE__ */ new Map();
    const fetchMores = [];
    let latestUpdated = 0;
    let hasMore = false;
    let isOptimistic = false;
    for (const payload of perObjectResults.values()) {
      for (const obj of payload.resolvedList) {
        seen.set(`${obj.$objectType}:${obj.$primaryKey}`, obj);
      }
      if (payload.lastUpdated > latestUpdated) {
        latestUpdated = payload.lastUpdated;
      }
      if (payload.isOptimistic) {
        isOptimistic = true;
      }
      if (payload.hasMore) {
        hasMore = true;
        fetchMores.push(payload.fetchMore);
      }
    }
    const payloads = [...perObjectResults.values()];
    const loading = perObjectResults.size < totalExpected || payloads.some((p) => p.status === "init" || p.status === "loading");
    observer.next({
      resolvedList: Array.from(seen.values()),
      isOptimistic,
      lastUpdated: latestUpdated,
      fetchMore: hasMore ? () => Promise.all(fetchMores.map((fn) => fn())).then(() => {
      }) : () => Promise.resolve(),
      hasMore,
      status: loading ? "loading" : payloads.some((p) => p.status === "error") ? "error" : "loaded",
      ...!hasMore ? {
        totalCount: String(seen.size)
      } : {}
    });
  }
  for (const obj of objectsArray) {
    const objKey = `${obj.$objectType ?? obj.$apiName}:${obj.$primaryKey}`;
    parentSub.add(store.links.observe({
      ...options,
      srcType: {
        type: "object",
        apiName: obj.$objectType ?? obj.$apiName
      },
      linkName,
      pk: obj.$primaryKey
    }, {
      next: (payload) => {
        if (errored) {
          return;
        }
        perObjectResults.set(objKey, payload);
        mergeAndEmit();
      },
      error: (err) => {
        if (errored) {
          return;
        }
        errored = true;
        parentSub.unsubscribe();
        observer.error(err);
      },
      // store link queries are long-lived and do not complete
      complete: () => {
      }
    }));
  }
  return new UnsubscribableWrapper(parentSub);
}

// src/observable/internal/Store.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/DebugFlags.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
process.env.NODE_ENV !== "production" && false;
var DEBUG_CACHE_KEYS = process.env.NODE_ENV !== "production" && false;

// src/observable/internal/actions/ActionApplication.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/actions/OptimisticJob.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/OptimisticId.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
function createOptimisticId() {
  if (process.env.NODE_ENV !== "production") {
    if (createOptimisticId.counter === void 0) {
      createOptimisticId.counter = 0;
    }
    return {
      __optimisticId: createOptimisticId.counter++
    };
  }
  return /* @__PURE__ */ Object.create(null);
}

// src/observable/internal/actions/OptimisticJob.ts
var OptimisticJob = class {
  #result;
  constructor(store, optimisticId) {
    const updatedObjects = [];
    const addedObjectPromises = [];
    const deletedObjects = [];
    this.getResult = () => {
      return this.#result ??= (async () => {
        const addedObjects = await Promise.allSettled(addedObjectPromises);
        const {
          batchResult
        } = store.batch({
          optimisticId
        }, (batch) => {
          for (const obj of addedObjects) {
            if (obj.status === "fulfilled") {
              store.objects.getQuery({
                apiName: obj.value.$objectType,
                pk: obj.value.$primaryKey
              }, void 0).writeToStore(obj.value, "loading", batch);
            } else {
              throw obj;
            }
          }
          for (const obj of updatedObjects) {
            store.objects.getQuery({
              apiName: obj.$objectType,
              pk: obj.$primaryKey
            }, void 0).writeToStore(obj, "loading", batch);
          }
          for (const obj of deletedObjects) {
            store.objects.getQuery({
              apiName: obj.$objectType,
              pk: obj.$primaryKey
            }, void 0).deleteFromStore("loading", batch);
          }
        });
        return batchResult.changes;
      })();
    };
    this.context = {
      updateObject(value) {
        updatedObjects.push(value);
        return this;
      },
      createObject(type, pk, properties) {
        const create = store.client[chunkEHWB7KEN_cjs.additionalContext].objectFactory2(store.client[chunkEHWB7KEN_cjs.additionalContext], [{
          $primaryKey: pk,
          $apiName: type.apiName,
          $objectType: type.apiName,
          ...properties
        }], void 0, {}).then((objs) => {
          return objs[0];
        });
        addedObjectPromises.push(create);
        return this;
      },
      deleteObject(value) {
        deletedObjects.push(value);
        return this;
      }
    };
  }
};
function runOptimisticJob(store, optimisticUpdate) {
  if (!optimisticUpdate) {
    return () => Promise.resolve();
  }
  const optimisticId = createOptimisticId();
  const job = new OptimisticJob(store, optimisticId);
  optimisticUpdate(job.context);
  const optimisticApplicationDone = job.getResult();
  return () => {
    return optimisticApplicationDone.then(
      // we don't want to leak the result
      () => void 0
    ).finally(() => {
      store.layers.remove(optimisticId);
    });
  };
}

// src/observable/internal/actions/ActionApplication.ts
var ACTION_DELAY = process.env.NODE_ENV === "production" ? 0 : 1e3;
var ActionApplication = class {
  constructor(store) {
    this.store = store;
  }
  applyAction = async (action, args, {
    optimisticUpdate
  } = {}) => {
    const logger = process.env.NODE_ENV !== "production" ? this.store.logger?.child({
      methodName: "applyAction"
    }) : this.store.logger;
    const removeOptimisticResult = runOptimisticJob(this.store, optimisticUpdate);
    return await (async () => {
      try {
        if (Array.isArray(args)) {
          if (process.env.NODE_ENV !== "production") {
            logger?.debug("applying action to multiple args", args);
          }
          const results = await this.store.client(action).batchApplyAction(args, {
            $returnEdits: true
          });
          await this.#invalidateActionEditResponse(results);
          return results;
        }
        const actionResults = await this.store.client(action).applyAction(args, {
          $returnEdits: true
        });
        if (process.env.NODE_ENV !== "production") {
          if (ACTION_DELAY > 0) {
            logger?.debug("action done, pausing", actionResults);
            await new Promise((resolve) => setTimeout(resolve, ACTION_DELAY));
            logger?.debug("action done, pausing done");
          }
        }
        await this.#invalidateActionEditResponse(actionResults);
        return actionResults;
      } finally {
        if (process.env.NODE_ENV !== "production") {
          logger?.debug("optimistic action complete; remove the results");
        }
        await removeOptimisticResult();
      }
    })();
  };
  #invalidateActionEditResponse = async (actionEditResponse) => {
    if (actionEditResponse == null) {
      return;
    }
    const {
      deletedObjects,
      modifiedObjects,
      addedObjects,
      editedObjectTypes,
      type
    } = actionEditResponse;
    let changes;
    if (type === "edits") {
      const promisesToWait = [];
      for (const list of [deletedObjects, modifiedObjects, addedObjects]) {
        for (const obj of list ?? []) {
          promisesToWait.push(this.store.invalidateObject(obj.objectType, obj.primaryKey));
        }
      }
      this.store.batch({}, (batch) => {
        for (const {
          objectType,
          primaryKey
        } of deletedObjects ?? []) {
          const cacheKey = this.store.cacheKeys.get(
            "object",
            objectType,
            primaryKey,
            /* rdpConfig */
            void 0
          );
          this.store.queries.peek(cacheKey)?.deleteFromStore(
            "loaded",
            // this is probably not the best value to use
            batch
          );
        }
      });
      await Promise.all(promisesToWait);
    } else {
      for (const apiName of editedObjectTypes) {
        await this.store.invalidateObjectType(apiName, changes);
      }
    }
  };
};

// src/observable/internal/aggregation/AggregationCacheKey.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
var API_NAME_IDX = 1;
var WHERE_IDX = 2;
var RDP_IDX = 3;
var AGGREGATE_IDX = 4;

// src/observable/internal/aggregation/AggregationsHelper.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/AbstractHelper.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/QuerySubscription.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
var subscriptionIdCounter = 0;
var QuerySubscription = class extends UnsubscribableWrapper {
  /** @internal */
  /** @internal */
  /** @internal */
  constructor(query, subscription) {
    super(subscription);
    this.query = query;
    this.subscription = subscription;
    this.subscriptionId = `sub_${++subscriptionIdCounter}`;
    Object.defineProperties(this, {
      query: {
        enumerable: false
      },
      subscription: {
        enumerable: false
      },
      subscriptionId: {
        enumerable: false
      }
    });
  }
};

// src/observable/internal/AbstractHelper.ts
var AbstractHelper = class {
  constructor(store, cacheKeys) {
    this.store = store;
    this.cacheKeys = cacheKeys;
  }
  observe(options, subFn) {
    const query = this.getQuery(options);
    return this._subscribe(query, options, subFn);
  }
  _subscribe(query, options, subFn) {
    this.store.cacheKeys.retain(query.cacheKey);
    if (options.mode !== "offline") {
      query.revalidate(options.mode === "force").catch((e) => {
        subFn.error(e);
        if (this.store.logger) {
          this.store.logger.error("Unhandled error in observeObject", e);
        } else {
          throw e;
        }
      });
    }
    const sub = query.subscribe(subFn);
    const querySub = new QuerySubscription(query, sub);
    query.registerSubscriptionDedupeInterval(querySub.subscriptionId, options.dedupeInterval);
    sub.add(() => {
      query.unregisterSubscriptionDedupeInterval(querySub.subscriptionId);
      this.store.cacheKeys.release(query.cacheKey);
    });
    return querySub;
  }
};

// src/observable/internal/aggregation/ObjectAggregationQuery.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/aggregation/AggregationQuery.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/Query.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
var Query = class {
  retainCount = 0;
  #connectable;
  #subscription;
  #subject;
  #subscriptionDedupeIntervals = /* @__PURE__ */ new Map();
  /** @internal */
  constructor(store, observable, opts, cacheKey, logger) {
    this.options = opts;
    this.cacheKey = cacheKey;
    this.store = store;
    this.cacheKeys = store.cacheKeys;
    this.#subject = observable;
    this.logger = logger ?? (process.env.NODE_ENV === "production" ? store.client[chunkEHWB7KEN_cjs.additionalContext].logger : store.client[chunkEHWB7KEN_cjs.additionalContext].logger?.child({}, {
      msgPrefix: process.env.NODE_ENV !== "production" ? `Query<${cacheKey.type}, ${cacheKey.otherKeys.map((x) => JSON.stringify(x)).join(", ")}>` : "Query"
    }));
  }
  subscribe(observer) {
    this.#connectable ??= this._createConnectable(this.#subject);
    this.#subscription = this.#connectable.connect();
    const sub = this.#connectable.subscribe({
      next: (value) => {
        if (observer.next) {
          observer.next(value);
        }
      },
      error: (err) => {
        if (observer.error) {
          observer.error(err);
        }
      },
      complete: () => {
        if (observer.complete) {
          observer.complete();
        }
      }
    });
    return sub;
  }
  /**
   * Register a subscription's dedupeInterval value
   */
  registerSubscriptionDedupeInterval(subscriptionId, dedupeInterval) {
    if (dedupeInterval != null && dedupeInterval > 0) {
      this.#subscriptionDedupeIntervals.set(subscriptionId, dedupeInterval);
    }
  }
  /**
   * Unregister a subscription's dedupeInterval value
   */
  unregisterSubscriptionDedupeInterval(subscriptionId) {
    this.#subscriptionDedupeIntervals.delete(subscriptionId);
  }
  /**
   * Get the minimum dedupeInterval from all active subscriptions
   */
  getMinimumDedupeInterval() {
    if (this.#subscriptionDedupeIntervals.size === 0) {
      return this.options.dedupeInterval ?? 0;
    }
    return Math.min(...this.#subscriptionDedupeIntervals.values());
  }
  /**
   * Causes the query to revalidate. This will cause the query to fetch
   * the latest data from the server and update the store if it is deemed
   * "stale" or if `force` is true.
   *
   * @param force
   * @returns
   */
  async revalidate(force) {
    const logger = process.env.NODE_ENV !== "production" ? this.logger?.child({
      methodName: "revalidate"
    }) : this.logger;
    if (force) {
      this.abortController?.abort();
    }
    if (this.pendingFetch) {
      if (process.env.NODE_ENV !== "production") {
        logger?.debug("Fetch is already pending, using it");
      }
      await this.pendingFetch;
      return;
    }
    const minDedupeInterval = this.getMinimumDedupeInterval();
    if (minDedupeInterval > 0 && this.lastFetchStarted != null && Date.now() - this.lastFetchStarted < minDedupeInterval) {
      if (process.env.NODE_ENV !== "production") {
        logger?.debug("Within dupeInterval, aborting revalidate");
      }
      return Promise.resolve();
    }
    if (process.env.NODE_ENV !== "production") {
      logger?.debug("Starting actual revalidate");
    }
    this.store.batch({}, (batch) => {
      this.setStatus("loading", batch);
    });
    this._preFetch();
    this.lastFetchStarted = Date.now();
    if (process.env.NODE_ENV !== "production") {
      logger?.debug("calling _fetchAndStore()");
    }
    this.pendingFetch = this._fetchAndStore().finally(() => {
      logger?.debug("promise's finally for _fetchAndStore()");
      this.pendingFetch = void 0;
    });
    await this.pendingFetch;
    return;
  }
  _preFetch() {
  }
  /**
   * Sets the status of the query in the store (but does not store that in `changes`).
   *
   * @param status
   * @param batch
   * @returns
   */
  setStatus(status, batch) {
    if (process.env.NODE_ENV !== "production") {
      this.logger?.child({
        methodName: "setStatus"
      }).debug(`Attempting to set status to '${status}'`);
    }
    const existing = batch.read(this.cacheKey);
    if (existing?.status === status) {
      if (process.env.NODE_ENV !== "production") {
        this.logger?.child({
          methodName: "setStatus"
        }).debug(`Status is already set to '${status}'; aborting`);
      }
      return;
    }
    if (process.env.NODE_ENV !== "production") {
      this.logger?.child({
        methodName: "setStatus"
      }).debug(`Writing status '${status}' to cache`);
    }
    batch.write(this.cacheKey, existing?.value, status);
  }
  dispose() {
    if (this.abortController) {
      this.abortController.abort();
    }
    this.#subscription?.unsubscribe();
    this._dispose();
  }
  /**
   * Per query type dispose functionality
   */
  _dispose() {
  }
  /**
   * The purpose of this method is to provide a way for others to write
   * directly into the store for this query.
   *
   * @param data
   * @param status
   * @param batch
   */
  /**
   * @param changes
   * @param optimisticId
   * @returns If revalidation is needed, a promise that resolves after the
   *          revalidation is complete. Otherwise, undefined.
   */
};

// src/observable/internal/aggregation/AggregationQuery.ts
var AggregationQuery = class extends Query {
  constructor(store, subject, cacheKey, opts) {
    super(store, subject, opts, cacheKey, process.env.NODE_ENV !== "production" ? store.client[chunkEHWB7KEN_cjs.additionalContext].logger?.child({}, {
      msgPrefix: `AggregationQuery<${cacheKey.otherKeys.map((x) => JSON.stringify(x)).join(", ")}>`
    }) : void 0);
    this.apiName = cacheKey.otherKeys[API_NAME_IDX];
    this.canonicalWhere = cacheKey.otherKeys[WHERE_IDX];
    this.rdpConfig = cacheKey.otherKeys[RDP_IDX];
    this.canonicalAggregate = cacheKey.otherKeys[AGGREGATE_IDX];
  }
  _createConnectable(subject) {
    return rxjs.connectable(subject.pipe(rxjs.map((x) => {
      return {
        status: x.status,
        result: x.value,
        lastUpdated: x.lastUpdated,
        error: x.status === "error" ? new Error("Aggregation failed") : void 0
      };
    })), {
      connector: () => new rxjs.BehaviorSubject({
        status: "init",
        result: void 0,
        lastUpdated: 0
      })
    });
  }
  async _fetchAndStore() {
    if (process.env.NODE_ENV !== "production") {
      this.logger?.child({
        methodName: "_fetchAndStore"
      }).debug("calling _fetchAndStore");
    }
    try {
      const result = await this._fetchAggregation();
      this.store.batch({}, (batch) => {
        this.writeToStore(result, "loaded", batch);
      });
    } catch (err) {
      this.store.batch({}, (batch) => {
        this.writeToStore(void 0, "error", batch);
      });
    }
  }
  writeToStore(data, status, batch) {
    batch.write(this.cacheKey, data, status);
    batch.changes.modified.add(this.cacheKey);
    return batch.read(this.cacheKey);
  }
  invalidateObjectType = (objectType, changes) => {
    if (this.apiName === objectType) {
      changes?.modified.add(this.cacheKey);
      return this.revalidate(true);
    }
    return Promise.resolve();
  };
};

// src/observable/internal/aggregation/ObjectAggregationQuery.ts
var ObjectAggregationQuery = class extends AggregationQuery {
  async _fetchAggregation() {
    const type = this.cacheKey.otherKeys[API_NAME_IDX];
    let objectSet = this.store.client({
      type: "object",
      apiName: type
    });
    if (this.rdpConfig) {
      objectSet = objectSet.withProperties(this.rdpConfig);
    }
    objectSet = objectSet.where(this.canonicalWhere);
    return await objectSet.aggregate(this.canonicalAggregate);
  }
};

// src/observable/internal/aggregation/AggregationsHelper.ts
var AggregationsHelper = class extends AbstractHelper {
  constructor(store, cacheKeys, whereCanonicalizer2, rdpCanonicalizer) {
    super(store, cacheKeys);
    this.whereCanonicalizer = whereCanonicalizer2;
    this.rdpCanonicalizer = rdpCanonicalizer;
  }
  observe(options, subFn) {
    return super.observe(options, subFn);
  }
  getQuery(options) {
    const {
      type,
      where,
      withProperties,
      aggregate
    } = options;
    const {
      apiName
    } = type;
    const typeKind = "type" in type ? type.type : "interface";
    const canonWhere = this.whereCanonicalizer.canonicalize(where ?? {});
    const canonRdp = withProperties ? this.rdpCanonicalizer.canonicalize(withProperties) : void 0;
    const canonAggregate = this.canonicalizeAggregate(aggregate);
    const aggregationCacheKey = this.cacheKeys.get("aggregation", typeKind, apiName, canonWhere, canonRdp, canonAggregate);
    return this.store.queries.get(aggregationCacheKey, () => {
      if (typeKind !== "object") {
        throw new Error("Only ObjectTypeDefinition is currently supported for aggregations");
      }
      return new ObjectAggregationQuery(this.store, this.store.subjects.get(aggregationCacheKey), aggregationCacheKey, options);
    });
  }
  canonicalizeAggregate(aggregate) {
    return JSON.parse(JSON.stringify(aggregate));
  }
};

// src/observable/internal/CacheKey.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
function DEBUG_ONLY__cacheKeyToString(x) {
  if (process.env.NODE_ENV !== "production") {
    return `${x.type}CacheKey<${x.otherKeys.map((xx) => JSON.stringify(xx)).join(", ")}>`.replaceAll('"', "'");
  } else {
    throw new Error("not implemented");
  }
}
function DEBUG_ONLY__cacheKeysToString(x) {
  if (process.env.NODE_ENV !== "production") {
    return "\n  - " + x.map(DEBUG_ONLY__cacheKeyToString).join("\n  - ");
  } else {
    throw new Error("not implemented");
  }
}

// src/observable/internal/CacheKeys.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/RefCounts.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
var RefCounts = class {
  refCounts = /* @__PURE__ */ new Map();
  // keeps our objects around for some extended duration after they are no longer
  // needed which is good for quick clicks across tabs.
  gcMap = /* @__PURE__ */ new Map();
  constructor(keepAlive, cleanup) {
    this.keepAlive = keepAlive;
    this.cleanup = cleanup;
  }
  register(key) {
    if (!this.refCounts.has(key)) {
      this.gcMap.set(key, Date.now() + this.keepAlive);
    }
    return key;
  }
  retain(key) {
    const count = this.refCounts.get(key) ?? 0;
    this.refCounts.set(key, count + 1);
    if (this.gcMap.has(key)) {
      this.gcMap.delete(key);
    }
  }
  release(key) {
    const count = this.refCounts.get(key);
    if (count === void 0) ; else if (count === 1) {
      this.refCounts.delete(key);
      this.gcMap.set(key, Date.now() + this.keepAlive);
    } else {
      this.refCounts.set(key, count - 1);
    }
  }
  has(key) {
    return this.refCounts.has(key);
  }
  gc() {
    const now = Date.now();
    for (const [key, deathTime] of this.gcMap) {
      if (deathTime < now) {
        this.gcMap.delete(key);
        this.cleanup(key);
      }
    }
  }
};

// src/observable/internal/CacheKeys.ts
var CacheKeys = class {
  #cacheKeys = new trie.Trie(false, (keys) => {
    const cacheKey = {
      type: keys[0],
      otherKeys: keys.slice(1)
    };
    this.#onCreate?.(cacheKey);
    return cacheKey;
  });
  #refCounts = new RefCounts(6e4, (k) => this.#cleanupCacheKey(k));
  // we are currently only using this for debug logging and should just remove it in the future if that
  // continues to be true
  #finalizationRegistry;
  #onCreate;
  #onDestroy;
  constructor({
    onCreate,
    onDestroy
  }) {
    this.#onCreate = onCreate;
    this.#onDestroy = onDestroy;
    setInterval(() => {
      this.#refCounts.gc();
    }, 1e3);
    this.#finalizationRegistry = new FinalizationRegistry((cleanupCallback) => {
      try {
        cleanupCallback();
      } catch (e) {
        console.error("Caught an error while running a finalization callback", e);
      }
    });
  }
  get(type, ...args) {
    const normalizedArgs = [...args];
    while (normalizedArgs.length > 0 && normalizedArgs[normalizedArgs.length - 1] === void 0) {
      normalizedArgs.pop();
    }
    const cacheKeyArgs = [type, ...normalizedArgs];
    if (process.env.NODE_ENV !== "production" && DEBUG_CACHE_KEYS) ;
    const cacheKey = this.#cacheKeys.lookupArray(cacheKeyArgs);
    this.#refCounts.register(cacheKey);
    return cacheKey;
  }
  retain(cacheKey) {
    this.#refCounts.retain(cacheKey);
  }
  release(cacheKey) {
    this.#refCounts.release(cacheKey);
  }
  #remove(cacheKey) {
    this.#cacheKeys.remove(cacheKey.type, ...cacheKey.otherKeys);
  }
  /**
   * Called after a key is no longer retained and the timeout has elapsed
   * @param key
   */
  #cleanupCacheKey = (key) => {
    this.#onDestroy?.(key);
    this.#remove(key);
  };
};

// src/observable/internal/Changes.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
var Changes = class {
  modifiedObjects = new mnemonist.MultiMap();
  addedObjects = new mnemonist.MultiMap();
  added = /* @__PURE__ */ new Set();
  modified = /* @__PURE__ */ new Set();
  deleted = /* @__PURE__ */ new Set();
  registerObject = (cacheKey, data, isNew) => {
    this[isNew ? "addedObjects" : "modifiedObjects"].set(data.$objectType ?? data.$apiName, data);
    this[isNew ? "added" : "modified"].add(cacheKey);
  };
  deleteObject = (cacheKey) => {
    this.deleted.add(cacheKey);
  };
  registerList = (key) => {
    this.modified.add(key);
  };
  registerLink = (cacheKey) => {
    this.modified.add(cacheKey);
  };
  deleteLink = (cacheKey) => {
    this.deleted.add(cacheKey);
  };
  registerObjectSet = (key) => {
    this.modified.add(key);
  };
  registerFunction = (key) => {
    this.modified.add(key);
  };
  isEmpty() {
    return this.modifiedObjects.size === 0 && this.addedObjects.size === 0 && this.added.size === 0 && this.modified.size === 0 && this.deleted.size === 0;
  }
};
function createChangedObjects() {
  return new Changes();
}
function DEBUG_ONLY__changesToString(changes) {
  if (process.env.NODE_ENV !== "production") {
    return JSON.stringify({
      modifiedObjects: multimapHelper(changes.modifiedObjects),
      addedObjects: multimapHelper(changes.addedObjects),
      added: listHelper(changes.added),
      modified: listHelper(changes.modified)
    }, null, 2);
  } else {
    throw new Error("not implemented");
  }
}
function listHelper(set) {
  return Array.from(set).map(DEBUG_ONLY__cacheKeyToString);
}
function multimapHelper(multimap) {
  return Object.fromEntries(Array.from(multimap.associations()).map(([type, objects]) => {
    return [type, objects.map((o) => o.$primaryKey)];
  }));
}

// src/observable/internal/function/FunctionsHelper.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/function/FunctionCacheKey.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
var PARAMS_IDX = 2;

// src/observable/internal/function/FunctionParamsCanonicalizer.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
function isPrimitiveValue(value) {
  if (value == null) return true;
  const t = typeof value;
  return t === "string" || t === "number" || t === "boolean" || t === "bigint";
}
var FunctionParamsCanonicalizer = class {
  #inputCache = /* @__PURE__ */ new WeakMap();
  #trie = new trie.Trie(false);
  #canonicalByMarker = /* @__PURE__ */ new WeakMap();
  canonicalize(params) {
    if (params == null) {
      return void 0;
    }
    if (this.#inputCache.has(params)) {
      return this.#inputCache.get(params);
    }
    const seen = /* @__PURE__ */ new WeakSet();
    const path = [];
    const canonicalValue = this.#encodeAndBuild(params, path, seen);
    const marker = this.#trie.lookupArray(path);
    let canonical = this.#canonicalByMarker.get(marker);
    if (canonical === void 0) {
      canonical = canonicalValue;
      this.#canonicalByMarker.set(marker, canonical);
    }
    this.#inputCache.set(params, canonical);
    return canonical;
  }
  #encodeAndBuild(value, path, seen) {
    if (value == null) {
      path.push(value);
      return value;
    }
    if (isPrimitiveValue(value)) {
      path.push(value);
      return value;
    }
    if (seen.has(value)) {
      throw new Error("Circular reference in function parameters");
    }
    seen.add(value);
    if (value instanceof Date) {
      const iso = value.toISOString();
      path.push("$:date", iso);
      return iso;
    }
    if (Array.isArray(value)) {
      path.push("$:array");
      const arr = value.map((item) => this.#encodeAndBuild(item, path, seen));
      path.push("$:array_end");
      return arr;
    }
    if (value instanceof Set) {
      path.push("$:set");
      const sorted = this.#sortSetValues(Array.from(value));
      const arr = sorted.map((item) => this.#encodeAndBuild(item, path, seen));
      path.push("$:set_end");
      return arr;
    }
    if (value instanceof Map) {
      path.push("$:map");
      const sorted = this.#sortMapEntries(Array.from(value.entries()));
      const arr = sorted.map(([k, v]) => [this.#encodeAndBuild(k, path, seen), this.#encodeAndBuild(v, path, seen)]);
      path.push("$:map_end");
      return arr;
    }
    if (chunkO3SKYLBQ_cjs.isObjectSpecifiersObject(value)) {
      const objectType = value.$objectType ?? value.$apiName;
      path.push("$:osdk", objectType, value.$primaryKey);
      return {
        $apiName: objectType,
        $primaryKey: value.$primaryKey
      };
    }
    if (chunkEHWB7KEN_cjs.isObjectSet(value)) {
      const wire = chunkEHWB7KEN_cjs.getWireObjectSet(value);
      path.push("$:objectset", wire);
      return wire;
    }
    const obj = value;
    path.push("$:object");
    const canonical = {};
    for (const key of Object.keys(obj).sort()) {
      path.push(key);
      canonical[key] = this.#encodeAndBuild(obj[key], path, seen);
    }
    path.push("$:object_end");
    return canonical;
  }
  #comparePrimitives(a, b) {
    const ta = typeof a;
    const tb = typeof b;
    if (ta !== tb) return ta.localeCompare(tb);
    if (ta === "string") return a.localeCompare(b);
    if (ta === "number") {
      const an = a;
      const bn = b;
      if (Number.isNaN(an) && Number.isNaN(bn)) return 0;
      if (Number.isNaN(an)) return 1;
      if (Number.isNaN(bn)) return -1;
      return an - bn;
    }
    if (ta === "boolean") return (a ? 1 : 0) - (b ? 1 : 0);
    if (ta === "bigint") {
      const ab = a;
      const bb = b;
      return ab < bb ? -1 : ab > bb ? 1 : 0;
    }
    return 0;
  }
  #sortSetValues(items) {
    return items.slice().sort((a, b) => {
      if (isPrimitiveValue(a) && isPrimitiveValue(b)) {
        return this.#comparePrimitives(a, b);
      }
      return JSON.stringify(a).localeCompare(JSON.stringify(b));
    });
  }
  #sortMapEntries(entries) {
    return entries.slice().sort(([a], [b]) => {
      if (isPrimitiveValue(a) && isPrimitiveValue(b)) {
        return this.#comparePrimitives(a, b);
      }
      return JSON.stringify(a).localeCompare(JSON.stringify(b));
    });
  }
};

// src/observable/internal/function/FunctionQuery.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
var FunctionQuery = class extends Query {
  #apiName;
  #version;
  #params;
  #dependsOn;
  #dependsOnObjects;
  #queryDef;
  constructor(store, subject, queryDef, params, cacheKey, opts) {
    super(store, subject, opts, cacheKey, process.env.NODE_ENV !== "production" ? store.client[chunkEHWB7KEN_cjs.additionalContext].logger?.child({}, {
      msgPrefix: `FunctionQuery<${cacheKey.otherKeys.map((x) => JSON.stringify(x)).join(", ")}>`
    }) : void 0);
    this.#apiName = queryDef.apiName;
    this.#version = queryDef.isFixedVersion ? queryDef.version : void 0;
    this.#params = params;
    this.#dependsOn = opts.dependsOn;
    this.#dependsOnObjects = opts.dependsOnObjects;
    this.#queryDef = queryDef;
  }
  _createConnectable(subject) {
    return rxjs.connectable(subject.pipe(rxjs.map((x) => {
      const value = x.value;
      return {
        status: x.status,
        result: value?.result,
        lastUpdated: value?.executedAt ?? 0,
        error: value?.error
      };
    })), {
      connector: () => new rxjs.BehaviorSubject({
        status: "init",
        result: void 0,
        lastUpdated: 0
      })
    });
  }
  async _fetchAndStore() {
    if (process.env.NODE_ENV !== "production") {
      this.logger?.child({
        methodName: "_fetchAndStore"
      }).debug("calling _fetchAndStore");
    }
    try {
      const result = await chunkO3SKYLBQ_cjs.applyQuery(this.store.client[chunkEHWB7KEN_cjs.additionalContext], this.#queryDef, this.#params);
      const executedAt = Date.now();
      this.store.batch({}, (batch) => {
        this.writeToStore({
          result,
          executedAt
        }, "loaded", batch);
      });
    } catch (e) {
      if (process.env.NODE_ENV !== "production") {
        this.logger?.child({
          methodName: "_fetchAndStore"
        }).error("Error executing function", e);
      }
      const error = e instanceof Error ? e : new Error(String(e));
      this.store.batch({}, (batch) => {
        this.writeToStore({
          result: void 0,
          executedAt: 0,
          error
        }, "error", batch);
      });
    }
  }
  writeToStore(data, status, batch) {
    batch.write(this.cacheKey, data, status);
    return batch.read(this.cacheKey);
  }
  invalidateObjectType = (objectType, changes) => {
    if (this.#dependsOn?.includes(objectType)) {
      changes?.registerFunction(this.cacheKey);
      return this.revalidate(true);
    }
    return Promise.resolve();
  };
  dependsOnObject(apiName, primaryKey) {
    if (!this.#dependsOnObjects) {
      return false;
    }
    return this.#dependsOnObjects.some((obj) => obj.$apiName === apiName && obj.$primaryKey === primaryKey);
  }
  /**
   * Called during batch operations when objects change.
   * Checks if any objects in dependsOnObjects were modified/added
   * and triggers revalidation if so.
   */
  maybeUpdateAndRevalidate = (changes, _optimisticId) => {
    if (!this.#dependsOnObjects?.length) {
      return void 0;
    }
    for (const dep of this.#dependsOnObjects) {
      const modifiedObjects = changes.modifiedObjects.get(dep.$apiName);
      if (modifiedObjects?.some((obj) => obj.$primaryKey === dep.$primaryKey)) {
        return this.revalidate(true);
      }
      const addedObjects = changes.addedObjects.get(dep.$apiName);
      if (addedObjects?.some((obj) => obj.$primaryKey === dep.$primaryKey)) {
        return this.revalidate(true);
      }
    }
    return void 0;
  };
  get apiName() {
    return this.#apiName;
  }
  get version() {
    return this.#version;
  }
};

// src/observable/internal/function/FunctionsHelper.ts
var FunctionsHelper = class extends AbstractHelper {
  paramsCanonicalizer = new FunctionParamsCanonicalizer();
  constructor(store, cacheKeys) {
    super(store, cacheKeys);
  }
  observe(options, subFn) {
    return super.observe(options, subFn);
  }
  getQuery(options) {
    const {
      queryDef,
      params,
      ...observeOpts
    } = options;
    const apiName = queryDef.apiName;
    const version = queryDef.isFixedVersion ? queryDef.version : void 0;
    const canonicalParams = this.paramsCanonicalizer.canonicalize(params);
    const functionCacheKey = this.cacheKeys.get("function", apiName, version, canonicalParams);
    return this.store.queries.get(functionCacheKey, () => new FunctionQuery(this.store, this.store.subjects.get(functionCacheKey), queryDef, params, functionCacheKey, observeOpts));
  }
  async invalidateFunction(apiName, params) {
    const functionApiName = typeof apiName === "string" ? apiName : apiName.apiName;
    let canonicalParams;
    if (params !== void 0) {
      canonicalParams = this.paramsCanonicalizer.canonicalize(params);
    }
    const promises = [];
    for (const cacheKey of this.store.queries.keys()) {
      if (cacheKey.type !== "function") {
        continue;
      }
      const query = this.store.queries.peek(cacheKey);
      if (!query) {
        continue;
      }
      if (query.apiName !== functionApiName) {
        continue;
      }
      if (canonicalParams !== void 0) {
        const queryCacheKey = cacheKey;
        const queryParams = queryCacheKey.otherKeys[PARAMS_IDX];
        if (queryParams !== canonicalParams) {
          continue;
        }
      }
      promises.push(query.revalidate(true));
    }
    await Promise.allSettled(promises);
  }
  async invalidateFunctionsByObject(apiName, primaryKey) {
    const promises = [];
    for (const cacheKey of this.store.queries.keys()) {
      if (cacheKey.type !== "function") {
        continue;
      }
      const query = this.store.queries.peek(cacheKey);
      if (!query) {
        continue;
      }
      if (query.dependsOnObject(apiName, primaryKey)) {
        promises.push(query.revalidate(true));
      }
    }
    await Promise.allSettled(promises);
  }
};

// src/observable/internal/IntersectCanonicalizer.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/Canonicalizer.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
var CachingCanonicalizer = class {
  /**
   * Cache for input object identity.
   */
  inputCache = /* @__PURE__ */ new WeakMap();
  /**
   * Look up or create a canonical form for the given input.
   * This method handles the structural deduplication logic.
   *
   * @param input The input to canonicalize
   * @returns The canonical form
   */
  canonicalize(input) {
    if (!input) {
      return void 0;
    }
    if (this.inputCache.has(input)) {
      return this.inputCache.get(input);
    }
    const canonical = this.lookupOrCreate(input);
    this.inputCache.set(input, canonical);
    return canonical;
  }
};

// src/observable/internal/IntersectCanonicalizer.ts
var IntersectCanonicalizer = class extends CachingCanonicalizer {
  structuralCache = /* @__PURE__ */ new Map();
  constructor(whereCanonicalizer2) {
    super();
    this.whereCanonicalizer = whereCanonicalizer2;
  }
  lookupOrCreate(intersectWith) {
    const canonicalClauses = intersectWith.map((item) => this.whereCanonicalizer.canonicalize(item.where ?? {}));
    const structuralKey = canonicalClauses.map((clause) => JSON.stringify(clause)).join("||");
    let canonical = this.structuralCache.get(structuralKey);
    if (!canonical) {
      canonical = canonicalClauses;
      this.structuralCache.set(structuralKey, canonical);
    }
    return canonical;
  }
};

// src/observable/internal/Layers.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/createInitEntry.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
function createInitEntry(cacheKey) {
  return {
    cacheKey,
    status: "init",
    value: void 0,
    lastUpdated: 0
  };
}

// src/observable/internal/Layer.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/WeakMapWithEntries.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
var WeakMapWithEntries = class {
  #map = /* @__PURE__ */ new WeakMap();
  #list = [];
  #toClean = [];
  #needsCleaning = false;
  #registry = new FinalizationRegistry(() => {
    this.#toClean.push(new WeakRef({}));
    this.#needsCleaning = true;
  });
  constructor() {
    const weakThis = new WeakRef(this);
    const intervalId = setInterval(() => {
      const self = weakThis.deref();
      if (self) {
        if (this.#needsCleaning) {
          this.#clean();
        }
      } else {
        clearInterval(intervalId);
      }
    }, 1e3);
  }
  #clean() {
    this.#list = this.#list.filter((ref) => ref.deref() !== void 0);
  }
  // functions for WeakMap
  delete(key) {
    const ret = this.#map.delete(key);
    this.#toClean.push(new WeakRef(key));
    this.#needsCleaning = true;
    return ret;
  }
  get(key) {
    return this.#map.get(key);
  }
  has(key) {
    return this.#map.has(key);
  }
  /**
   * Adds a new element with a specified key and value.
   * @param key Must be an object or symbol.
   */
  set(key, value) {
    if (!this.#map.has(key)) {
      this.#list.push(new WeakRef(key));
    }
    this.#map.set(key, value);
    return this;
  }
  [Symbol.toStringTag] = "WeakMap";
  // functions for iterables
  /** Returns an iterable of entries in the map. */
  [Symbol.iterator]() {
    return this.entries();
  }
  /**
   * Returns an iterable of key, value pairs for every entry in the map.
   */
  entries() {
    const self = this;
    function* iter() {
      for (const ref of self.#list) {
        const key = ref.deref();
        if (key !== void 0) {
          yield [key, self.#map.get(key)];
        }
      }
    }
    return iter();
  }
  /**
   * Returns an iterable of keys in the map
   */
  keys() {
    const self = this;
    function* iter() {
      for (const ref of self.#list) {
        const key = ref.deref();
        if (key !== void 0) {
          yield key;
        }
      }
    }
    return iter();
  }
  /**
   * Returns an iterable of values in the map
   */
  values() {
    const self = this;
    function* iter() {
      for (const ref of self.#list) {
        const key = ref.deref();
        if (key !== void 0) {
          const value = self.#map.get(key);
          if (value !== void 0) {
            yield value;
          }
        }
      }
    }
    return iter();
  }
};

// src/observable/internal/Layer.ts
var Layer = class _Layer {
  #parent;
  #cache = new WeakMapWithEntries();
  #layerId;
  constructor(parent, layerId) {
    this.#parent = parent;
    this.#layerId = layerId;
  }
  get parentLayer() {
    return this.#parent;
  }
  get layerId() {
    return this.#layerId;
  }
  addLayer(layerId) {
    return new _Layer(this, layerId);
  }
  removeLayer(layerId) {
    if (layerId == null || this.#parent == null) {
      return this;
    }
    if (this.#layerId !== layerId) {
      this.#parent = this.#parent.removeLayer(layerId);
      return this;
    }
    return this.#parent.removeLayer(layerId);
  }
  entries() {
    return this.#cache.entries();
  }
  keys() {
    return this.#cache.keys();
  }
  get(cacheKey) {
    return this.#cache.get(cacheKey) ?? this.#parent?.get(cacheKey);
  }
  set(cacheKey, value) {
    this.#cache.set(cacheKey, value);
  }
};

// src/observable/internal/Subjects.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
var Subjects = class {
  #layers;
  // we can use a regular Map here because the refCounting will
  // handle cleanup.
  #cacheKeyToSubject = /* @__PURE__ */ new WeakMap();
  constructor({
    logger,
    layers
  }) {
    this.logger = logger;
    this.#layers = layers;
  }
  peek = (cacheKey) => {
    return this.#cacheKeyToSubject.get(cacheKey);
  };
  get = (cacheKey) => {
    let subject = this.#cacheKeyToSubject.get(cacheKey);
    if (!subject) {
      const initialValue = this.#layers.top.get(cacheKey) ?? createInitEntry(cacheKey);
      subject = new rxjs.BehaviorSubject({
        ...initialValue,
        isOptimistic: initialValue.value !== this.#layers.truth.get(cacheKey)?.value
      });
      this.#cacheKeyToSubject.set(cacheKey, subject);
    }
    return subject;
  };
  delete = (cacheKey) => {
    const subject = this.peek(cacheKey);
    if (subject) {
      subject.complete();
      this.#cacheKeyToSubject.delete(cacheKey);
    }
  };
};

// src/observable/internal/tombstone.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
var tombstone = void 0;

// src/observable/internal/Layers.ts
var Layers = class {
  #truthLayer = new Layer(void 0, void 0);
  #topLayer;
  #onRevalidate;
  constructor({
    logger,
    onRevalidate
  }) {
    this.logger = logger;
    this.#topLayer = this.#truthLayer;
    this.subjects = new Subjects({
      logger,
      layers: this
    });
    this.#onRevalidate = onRevalidate;
  }
  get top() {
    return this.#topLayer;
  }
  get truth() {
    return this.#truthLayer;
  }
  remove(layerId) {
    !(layerId != null) ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "undefined is the reserved layerId for the truth layer") : invariant5__default.default(false) : void 0;
    let currentLayer = this.#topLayer;
    const cacheKeys = /* @__PURE__ */ new Map();
    while (currentLayer != null && currentLayer.parentLayer != null) {
      if (currentLayer.layerId === layerId) {
        for (const [k, v] of currentLayer.entries()) {
          if (cacheKeys.has(k)) continue;
          cacheKeys.set(k, v);
        }
      }
      currentLayer = currentLayer.parentLayer;
    }
    this.#topLayer = this.#topLayer.removeLayer(layerId);
    for (const [k, oldEntry] of cacheKeys) {
      const currentEntry = this.#topLayer.get(k);
      if (oldEntry !== currentEntry) {
        const newEntry = currentEntry ?? createInitEntry(k);
        this.subjects.peek(k)?.next({
          ...newEntry,
          isOptimistic: currentEntry?.value !== this.#truthLayer.get(k)?.value
        });
      }
    }
  }
  batch({
    optimisticId,
    changes
  }, batchFn) {
    !(optimisticId === void 0 || !!optimisticId) ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "optimistic must be undefined or not falsy") : invariant5__default.default(false) : void 0;
    const batchContext = this.#createBatchContext({
      optimisticId,
      changes
    });
    const retVal = batchFn(batchContext);
    this.#onRevalidate(changes, optimisticId).catch((e) => {
      if (this.logger) {
        this.logger.error("Unhandled error in batch", e);
      } else {
        console.error("Unhandled error in batch", e);
        throw e;
      }
    });
    return {
      batchResult: batchContext,
      retVal,
      changes: batchContext.changes
    };
  }
  #createBatchContext({
    optimisticId,
    changes
  }) {
    let needsLayer = optimisticId !== void 0;
    const batchContext = {
      changes,
      createLayerIfNeeded: () => {
        if (needsLayer) {
          this.#topLayer = this.#topLayer.addLayer(optimisticId);
          needsLayer = false;
        }
      },
      optimisticWrite: !!optimisticId,
      write: (cacheKey, value, status) => {
        const oldTopValue = this.#topLayer.get(cacheKey);
        if (optimisticId) batchContext.createLayerIfNeeded();
        const writeLayer = optimisticId ? this.#topLayer : this.#truthLayer;
        const newValue = {
          cacheKey,
          value,
          lastUpdated: Date.now(),
          status
        };
        writeLayer.set(cacheKey, newValue);
        const newTopValue = this.#topLayer.get(cacheKey);
        if (oldTopValue !== newTopValue) {
          this.subjects.get(cacheKey)?.next({
            ...newValue,
            isOptimistic: newTopValue?.value !== this.#truthLayer.get(cacheKey)?.value
          });
        }
        return newValue;
      },
      delete: (cacheKey, status) => {
        return batchContext.write(cacheKey, tombstone, status);
      },
      read: (cacheKey) => {
        return optimisticId ? this.#topLayer.get(cacheKey) : this.#truthLayer.get(cacheKey);
      }
    };
    return batchContext;
  }
};

// src/observable/internal/links/LinksHelper.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/links/SpecificLinkQuery.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/base-list/BaseListQuery.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/isObjectInstance.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
function isObjectInstance(item) {
  return item != null && typeof item === "object" && "$primaryKey" in item;
}

// src/observable/internal/list/ListCacheKey.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
var API_NAME_IDX2 = 1;
var WHERE_IDX2 = 2;
var ORDER_BY_IDX = 3;
var RDP_IDX2 = 4;
var INTERSECT_IDX = 5;
var PIVOT_IDX = 6;
var RIDS_IDX = 7;

// src/observable/internal/sorting/SortingStrategy.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
var NoOpSortingStrategy = class {
  sortCacheKeys(objectCacheKeys, _batch) {
    return objectCacheKeys;
  }
};
var OrderBySortingStrategy = class {
  constructor(apiName, orderBy) {
    this.apiName = apiName;
    this.orderBy = orderBy;
    this.sortFns = createOrderBySortFns(orderBy);
  }
  sortCacheKeys(objectCacheKeys, batch) {
    if (Object.keys(this.orderBy).length === 0) {
      return objectCacheKeys;
    }
    return objectCacheKeys.sort((a, b) => {
      for (const sortFn of this.sortFns) {
        const ret = sortFn(batch.read(a)?.value?.$as(this.apiName), batch.read(b)?.value?.$as(this.apiName));
        if (ret !== 0) {
          return ret;
        }
      }
      return 0;
    });
  }
};
function createOrderBySortFns(orderBy) {
  return Object.entries(orderBy).map(([key, order]) => {
    return (a, b) => {
      const aValue = a?.[key];
      const bValue = b?.[key];
      if (aValue == null && bValue == null) {
        return 0;
      }
      if (aValue == null) {
        return 1;
      }
      if (bValue == null) {
        return -1;
      }
      const m = order === "asc" ? -1 : 1;
      return aValue < bValue ? m : aValue > bValue ? -m : 0;
    };
  });
}

// src/observable/internal/base-list/createCollectionConnectable.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
function createCollectionConnectable(subject, subjects, createPayload) {
  return rxjs.connectable(subject.pipe(rxjs.switchMap((listEntry) => {
    const resolvedData = listEntry?.value?.data == null || listEntry.value.data.length === 0 ? rxjs.of([]) : rxjs.combineLatest(listEntry.value.data.map((cacheKey) => subjects.get(cacheKey).pipe(rxjs.map((objectEntry) => objectEntry?.value), rxjs.distinctUntilChanged())));
    return rxjs.scheduled(rxjs.combineLatest({
      resolvedData,
      isOptimistic: rxjs.of(listEntry.isOptimistic),
      status: rxjs.of(listEntry.status),
      lastUpdated: rxjs.of(listEntry.lastUpdated),
      totalCount: rxjs.of(listEntry?.value?.totalCount)
    }).pipe(rxjs.map((params) => createPayload({
      resolvedData: Array.isArray(params.resolvedData) ? params.resolvedData : [],
      isOptimistic: params.isOptimistic,
      status: params.status,
      lastUpdated: params.lastUpdated,
      totalCount: params.totalCount
    }))), rxjs.asapScheduler);
  })), {
    resetOnDisconnect: false,
    connector: () => new rxjs.ReplaySubject(1)
  });
}

// src/observable/internal/base-list/removeDuplicates.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
function removeDuplicates(objectCacheKeys, batch) {
  const visited = /* @__PURE__ */ new Set();
  return objectCacheKeys.filter((key) => {
    batch.read(key);
    if (visited.has(key)) {
      return false;
    }
    visited.add(key);
    return true;
  });
}

// src/observable/internal/base-list/BaseListQuery.ts
var BaseListQuery = class extends Query {
  /**
   * The sorting strategy to use for this collection
   * @protected
   */
  sortingStrategy = new NoOpSortingStrategy();
  /**
   * Get RDP configuration from the cache key
   */
  get rdpConfig() {
    return this.cacheKey.otherKeys[RDP_IDX2];
  }
  // Collection-specific behavior is implemented by subclasses
  /**
   * Token for the next page of results
   * @protected
   */
  /**
   * Promise tracking an in-progress page fetch
   * @protected
   */
  //
  // Shared Implementations
  //
  /**
   * Standard method to update a list of objects
   * Handles common list update patterns for both ListQuery and SpecificLinkQuery
   *
   * @param items Objects or cache keys to add to the list
   * @param status Status to set for the list
   * @param batch Batch context to use
   * @param append Whether to append to the existing list or replace it
   * @param totalCount Optional total count from API response
   * @returns The updated entry
   */
  _updateList(items, status, batch, append = false, totalCount) {
    if (process.env.NODE_ENV !== "production") {
      this.logger?.child({
        methodName: "updateList"
      }).debug(`{status: ${status}, append: ${append}}`, JSON.stringify(items, null, 2));
    }
    let objectCacheKeys;
    if (items.length === 0) {
      objectCacheKeys = [];
    } else if (isObjectInstance(items[0])) {
      objectCacheKeys = this.store.objects.storeOsdkInstances(items, batch, this.rdpConfig);
    } else {
      objectCacheKeys = items;
    }
    objectCacheKeys = this.#retainReleaseAppend(batch, append, objectCacheKeys);
    objectCacheKeys = this._sortCacheKeys(objectCacheKeys, batch);
    objectCacheKeys = removeDuplicates(objectCacheKeys, batch);
    return this.writeToStore({
      data: objectCacheKeys,
      totalCount
    }, status, batch);
  }
  /**
   * Common implementation for writing to store for collection-based queries
   * @param data The collection data to write to the store
   * @param status The status to set
   * @param batch The batch context
   */
  writeToStore(data, status, batch) {
    const entry = batch.read(this.cacheKey);
    if (entry && deepEqual__default.default(data, entry.value)) {
      if (entry.status === status) {
        if (process.env.NODE_ENV !== "production") {
          this.logger?.child({
            methodName: "writeToStore"
          }).debug(`Collection data was deep equal and status unchanged (${status}), skipping update`);
        }
        return entry;
      }
      if (process.env.NODE_ENV !== "production") {
        this.logger?.child({
          methodName: "writeToStore"
        }).debug(`Collection data was deep equal, just updating status from ${entry.status} to ${status}`);
      }
      return batch.write(this.cacheKey, entry.value, status);
    }
    if (process.env.NODE_ENV !== "production") {
      this.logger?.child({
        methodName: "writeToStore"
      }).debug(`{status: ${status}},`, DEBUG_ONLY__cacheKeysToString(data.data));
    }
    const ret = batch.write(this.cacheKey, data, status);
    this.registerCacheChanges(batch);
    return ret;
  }
  /**
   * Register changes to the cache based on the specific collection type
   * Implemented by subclasses to handle specific change registration
   */
  /**
   * Common method for managing object reference counting and appending results
   * Used by collection queries when updating object references
   *
   * @param batch The batch context to use
   * @param append Whether to append to existing objects or replace them
   * @param objectCacheKeys Array of object cache keys to process
   * @returns The final array of object cache keys after retain/release/append
   */
  #retainReleaseAppend(batch, append, objectCacheKeys) {
    const existingList = batch.read(this.cacheKey);
    if (!batch.optimisticWrite) {
      if (!append) {
        for (const objectCacheKey of existingList?.value?.data ?? []) {
          this.store.cacheKeys.release(objectCacheKey);
        }
      }
      for (const objectCacheKey of objectCacheKeys) {
        this.store.cacheKeys.retain(objectCacheKey);
      }
    }
    if (append) {
      objectCacheKeys = [...existingList?.value?.data ?? [], ...objectCacheKeys];
    }
    return objectCacheKeys;
  }
  _dispose() {
    this.store.batch({}, (batch) => {
      const entry = batch.read(this.cacheKey);
      if (entry) {
        for (const objectCacheKey of entry.value?.data ?? []) {
          this.store.cacheKeys.release(objectCacheKey);
        }
      }
    });
  }
  /**
   * Creates a payload from collection parameters
   * Default implementation that covers common fields for all collection types
   * Subclasses may override to add type-specific fields if needed
   *
   * @param params Common collection parameters
   * @returns A typed payload for the specific collection type
   */
  createPayload(params) {
    return {
      resolvedList: params.resolvedData,
      isOptimistic: params.isOptimistic,
      fetchMore: this.fetchMore,
      hasMore: this.nextPageToken != null,
      status: params.status,
      lastUpdated: params.lastUpdated,
      totalCount: params.totalCount
    };
  }
  /**
   * Creates a connectable observable for this collection
   * Common implementation shared by all collection types
   *
   * @param subject The subject to connect to
   * @returns A connectable observable of the collection's payload type
   */
  _createConnectable(subject) {
    return createCollectionConnectable(subject, this.store.subjects, (params) => this.createPayload(params));
  }
  /**
   * @override Reset pagination state before a fetch
   */
  _preFetch() {
    this.nextPageToken = void 0;
    super._preFetch();
  }
  /**
   * Common fetchMore implementation for pagination
   * Handles pending request management and loading states
   */
  fetchMore = () => {
    if (this.pendingPageFetch) {
      return this.pendingPageFetch;
    }
    if (this.pendingFetch) {
      this.pendingPageFetch = (async () => {
        await this.pendingFetch;
        await this.fetchMore();
      })().finally(() => {
        this.pendingPageFetch = void 0;
      });
      return this.pendingPageFetch;
    }
    if (this.nextPageToken == null) {
      return Promise.resolve(void 0);
    }
    this.store.batch({}, (batch) => {
      this.setStatus("loading", batch);
    });
    this.pendingFetch = this.fetchPageAndUpdate("loaded", this.abortController?.signal).then(() => void 0).finally(() => {
      this.pendingFetch = void 0;
    });
    return this.pendingFetch;
  };
  /**
   * Minimum number of results to load initially
   * May be overridden by subclasses for specific collection types
   * @protected
   */
  minResultsToLoad = 0;
  /**
   * Common _fetchAndStore implementation for pagination
   * Uses fetchPageAndUpdate to load the initial set of data
   * Will load multiple pages if necessary to reach minResultsToLoad
   */
  async _fetchAndStore() {
    if (process.env.NODE_ENV !== "production") {
      this.logger?.child({
        methodName: "_fetchAndStore"
      }).debug("fetching pages");
    }
    while (true) {
      const entry = await this.fetchPageAndUpdate("loading", this.abortController?.signal);
      if (!entry) {
        return;
      }
      const count = entry.value?.data.length || 0;
      if (count >= this.minResultsToLoad || this.nextPageToken == null) {
        break;
      }
      await Promise.resolve();
    }
    this.store.batch({}, (batch) => {
      this.setStatus("loaded", batch);
    });
    return Promise.resolve();
  }
  /**
   * Template method for fetching a page of data and updating the store
   * Provides common error handling and abort signal checking
   *
   * @param status The status to set for the entry
   * @param signal Optional AbortSignal for cancellation
   * @returns A promise that resolves to the updated entry or undefined if aborted
   */
  async fetchPageAndUpdate(status, signal) {
    if (process.env.NODE_ENV !== "production") {
      this.logger?.child({
        methodName: "fetchPageAndUpdate"
      }).debug(`Fetching data with status: ${status}`);
    }
    if (signal?.aborted) {
      return void 0;
    }
    try {
      const hadPreviousPage = this.nextPageToken != null;
      const result = await this.fetchPageData(signal);
      this.currentTotalCount = result.totalCount;
      if (signal?.aborted) {
        return void 0;
      }
      const {
        retVal
      } = this.store.batch({}, (batch) => {
        const append = hadPreviousPage;
        const finalStatus = result.nextPageToken ? status : "loaded";
        const objectKeys = this.store.objects.storeOsdkInstances(result.data, batch, this.rdpConfig);
        return this._updateList(objectKeys, finalStatus, batch, append, this.currentTotalCount);
      });
      return retVal;
    } catch (error) {
      if (process.env.NODE_ENV !== "production") {
        this.logger?.child({
          methodName: "fetchPageAndUpdate"
        }).error("Error fetching data", error);
      }
      if (!signal?.aborted) {
        const {
          retVal
        } = this.store.batch({}, (batch) => {
          return this.handleFetchError(error, status, batch);
        });
        return retVal;
      }
      return void 0;
    }
  }
  /**
   * Abstract method that subclasses implement for their specific data fetching logic
   *
   * @param signal Optional AbortSignal for cancellation
   * @returns A promise that resolves to the fetched data
   */
  /**
   * Handle fetch errors by setting appropriate error state
   * Default implementation that subclasses can override
   *
   * @param error The error that occurred
   * @param status The intended status if successful
   * @param batch The batch context to use
   * @returns The updated entry with error status
   */
  handleFetchError(_error, _status, batch) {
    const existingTotalCount = batch.read(this.cacheKey)?.value?.totalCount;
    return this.writeToStore({
      data: [],
      totalCount: existingTotalCount
    }, "error", batch);
  }
  /**
   * Sort the collection items using the configured sorting strategy
   * @param objectCacheKeys - The cache keys to sort
   * @param batch - The batch context
   * @returns Sorted array of cache keys
   */
  _sortCacheKeys(objectCacheKeys, batch) {
    return this.sortingStrategy.sortCacheKeys(objectCacheKeys, batch);
  }
  /**
   * Unified method for updating collection data in the store
   * Handles storing, sorting, deduplication, and reference counting
   *
   * @param items - Either object cache keys or object instances to update
   * @param options - Configuration options for the update
   * @param batch - The batch context to use
   * @returns The updated entry
   */
  updateCollection(items, options, batch) {
    if (process.env.NODE_ENV !== "production") {
      const logger = process.env.NODE_ENV !== "production" ? this.logger?.child({
        methodName: "updateCollection"
      }) : this.logger;
      logger?.debug(`{status: ${options.status}, append: ${options.append}}`, JSON.stringify(items, null, 2));
    }
    let objectCacheKeys;
    if (items.length === 0) {
      objectCacheKeys = [];
    } else if (isObjectInstance(items[0])) {
      objectCacheKeys = this.store.objects.storeOsdkInstances(items, batch, this.rdpConfig);
    } else {
      objectCacheKeys = items;
    }
    objectCacheKeys = this.#retainReleaseAppend(batch, options.append ?? false, objectCacheKeys);
    objectCacheKeys = this._sortCacheKeys(objectCacheKeys, batch);
    objectCacheKeys = removeDuplicates(objectCacheKeys, batch);
    const existingTotalCount = batch.read(this.cacheKey)?.value?.totalCount;
    return this.writeToStore({
      data: objectCacheKeys,
      totalCount: existingTotalCount
    }, options.status, batch);
  }
  //
  // Shared Websocket Subscription Methods
  //
  /**
   * Create standard websocket subscription handlers for an ObjectSet.
   * Subclasses can override individual handlers for custom behavior.
   *
   * @param objectSet The ObjectSet to subscribe to
   * @param sub The parent subscription to add cleanup to
   * @param methodName The method name for logging purposes
   */
  createWebsocketSubscription(objectSet, sub, methodName = "registerStreamUpdates") {
    const logger = process.env.NODE_ENV !== "production" ? this.logger?.child({
      methodName
    }) : this.logger;
    if (process.env.NODE_ENV !== "production") {
      logger?.child({
        methodName
      }).info("Subscribing from websocket");
    }
    try {
      const websocketSubscription = objectSet.subscribe({
        onChange: this.onOswChange.bind(this),
        onError: this.onOswError.bind(this),
        onOutOfDate: this.onOswOutOfDate.bind(this),
        onSuccessfulSubscription: this.onOswSuccessfulSubscription.bind(this)
      });
      sub.add(() => {
        if (process.env.NODE_ENV !== "production") {
          logger?.child({
            methodName
          }).info("Unsubscribing from websocket");
        }
        websocketSubscription.unsubscribe();
      });
    } catch (error) {
      if (this.logger) {
        this.logger.child({
          methodName
        }).error("Failed to register stream updates", error);
      }
      this.onOswError({
        subscriptionClosed: true,
        error
      });
    }
  }
  /**
   * Handler called when websocket subscription is successfully established.
   */
  onOswSuccessfulSubscription() {
    if (process.env.NODE_ENV !== "production") {
      this.logger?.child({
        methodName: "onSuccessfulSubscription"
      }).debug("");
    }
  }
  /**
   * Handler called when subscribed data becomes out of date.
   */
  onOswOutOfDate() {
    if (process.env.NODE_ENV !== "production") {
      this.logger?.child({
        methodName: "onOutOfDate"
      }).debug("");
    }
  }
  /**
   * Handler called when websocket subscription encounters an error.
   */
  onOswError(errors) {
    if (this.logger) {
      this.logger?.child({
        methodName: "onError"
      }).error("subscription errors", errors);
    }
  }
  /**
   * Handler called when an object in the subscribed set is added or updated.
   * Default implementation stores the object with RDP config if available.
   *
   * @param update The object update notification
   */
  onOswChange({
    object,
    state
  }) {
    const logger = process.env.NODE_ENV !== "production" ? this.logger?.child({
      methodName: "registerStreamUpdates"
    }) : this.logger;
    if (process.env.NODE_ENV !== "production") {
      logger?.child({
        methodName: "onChange"
      }).debug(`Got an update of type: ${state}`, object);
    }
    if (state === "ADDED_OR_UPDATED") {
      this.store.batch({}, (batch) => {
        this.store.objects.storeOsdkInstances(
          [object],
          batch,
          this.rdpConfig
          // Safe - null for queries without RDPs
        );
      });
    } else if (state === "REMOVED") {
      this.onOswRemoved(object);
    }
  }
  /**
   * Handler called when an object is removed from the subscribed set.
   * Default implementation deletes the object from cache.
   * ListQuery overrides this for list-specific removal logic.
   *
   * @param object The removed object
   */
  onOswRemoved(object) {
    if (process.env.NODE_ENV !== "production") {
      this.logger?.child({
        methodName: "onRemoved"
      }).debug("Removing object", object);
    }
    this.store.batch({}, (batch) => {
      const objectCacheKey = this.store.cacheKeys.get("object", object.$objectType ?? object.$apiName, object.$primaryKey);
      batch.delete(objectCacheKey, "loaded");
    });
  }
};

// src/observable/internal/links/SpecificLinkQuery.ts
var SpecificLinkQuery = class extends BaseListQuery {
  #sourceApiName;
  #sourcePk;
  #linkName;
  #whereClause;
  #orderBy;
  /**
   * Register changes to the cache specific to SpecificLinkQuery
   */
  registerCacheChanges(batch) {
    batch.changes.modified.add(this.cacheKey);
  }
  constructor(store, subject, cacheKey, opts) {
    super(store, subject, opts, cacheKey, process.env.NODE_ENV !== "production" ? store.client[chunkEHWB7KEN_cjs.additionalContext].logger?.child({}, {
      msgPrefix: `SpecificLinkQuery<${cacheKey.otherKeys.map((x) => JSON.stringify(x)).join(", ")}>`
    }) : void 0);
    [this.#sourceApiName, this.#sourcePk, this.#linkName, this.#whereClause, this.#orderBy] = cacheKey.otherKeys;
  }
  // _fetchAndStore is now implemented in BaseCollectionQuery
  /**
   * Implements fetchPageData from the BaseCollectionQuery template method pattern
   * Fetches a page of linked objects
   */
  async fetchPageData(signal) {
    const client = this.store.client;
    const sourceObjectDef = {
      type: "object",
      apiName: this.#sourceApiName
    };
    const sourceMetadata = await client[chunkEHWB7KEN_cjs.additionalContext].ontologyProvider.getObjectDefinition(this.#sourceApiName);
    if (this.#orderBy && Object.keys(this.#orderBy).length > 0) {
      const linkDef = sourceMetadata.links?.[this.#linkName];
      if (!linkDef?.targetType) {
        throw new Error(`Missing link definition or targetType for link '${this.#linkName}' on object type '${this.#sourceApiName}'`);
      }
      this.sortingStrategy = new OrderBySortingStrategy(linkDef.targetType, this.#orderBy);
    }
    const sourceQuery = client(sourceObjectDef).where({
      [sourceMetadata.primaryKeyApiName]: this.#sourcePk
    });
    const linkQuery = sourceQuery.pivotTo(this.#linkName);
    if (signal?.aborted) {
      throw new Error("Aborted");
    }
    const queryParams = {
      $pageSize: this.options.pageSize || 100,
      $nextPageToken: this.nextPageToken,
      $includeRid: true
    };
    if (this.#orderBy && Object.keys(this.#orderBy).length > 0) {
      queryParams.$orderBy = this.#orderBy;
    }
    if (this.#whereClause && Object.keys(this.#whereClause).length > 0) {
      queryParams.$where = this.#whereClause;
    }
    const response = await linkQuery.fetchPage(queryParams);
    this.nextPageToken = response.nextPageToken;
    return response;
  }
  /**
   * Removes a link query from the store
   */
  deleteFromStore(status, batch) {
    const entry = batch.read(this.cacheKey);
    if (entry && deepEqual__default.default(tombstone, entry.value)) {
      if (process.env.NODE_ENV !== "production") {
        this.logger?.child({
          methodName: "deleteFromStore"
        }).debug(`Links were already deleted, just setting status`);
      }
      return batch.write(this.cacheKey, entry.value, status);
    }
    if (process.env.NODE_ENV !== "production") {
      this.logger?.child({
        methodName: "deleteFromStore"
      }).debug(JSON.stringify({
        status
      }));
    }
    if (!entry || !entry.value) {
      return;
    }
    const ret = batch.delete(this.cacheKey, status);
    batch.changes.deleted.add(this.cacheKey);
    return ret;
  }
  /**
   * Implements Query.maybeUpdateAndRevalidate to handle cache invalidation
   */
  maybeUpdateAndRevalidate = async (changes, _optimisticId) => {
    if (changes.modified.has(this.cacheKey)) {
      return this.revalidate(true);
    }
    return Promise.resolve();
  };
  invalidateObjectType = (objectType, changes) => {
    if (this.#sourceApiName === objectType) {
      changes?.modified.add(this.cacheKey);
      return this.revalidate(true);
    } else {
      return (async () => {
        const sourceMetadata = await this.store.client[chunkEHWB7KEN_cjs.additionalContext].ontologyProvider.getObjectDefinition(this.#sourceApiName);
        const linkDef = sourceMetadata.links?.[this.#linkName];
        if (!linkDef || linkDef.targetType !== objectType) return;
        const promise = this.revalidate(true);
        changes?.modified.add(this.cacheKey);
        return promise;
      })();
    }
  };
};

// src/observable/internal/links/LinksHelper.ts
var LinksHelper = class extends AbstractHelper {
  constructor(store, cacheKeys, whereCanonicalizer2, orderByCanonicalizer2) {
    super(store, cacheKeys);
    this.whereCanonicalizer = whereCanonicalizer2;
    this.orderByCanonicalizer = orderByCanonicalizer2;
  }
  getQuery(options) {
    const {
      apiName
    } = options.srcType;
    const canonWhere = this.whereCanonicalizer.canonicalize(options.where ?? {});
    const canonOrderBy = this.orderByCanonicalizer.canonicalize(options.orderBy ?? {});
    const linkCacheKey = this.cacheKeys.get("specificLink", apiName, options.pk, options.linkName, canonWhere, canonOrderBy);
    return this.store.queries.get(linkCacheKey, () => {
      return new SpecificLinkQuery(this.store, this.store.subjects.get(linkCacheKey), linkCacheKey, options);
    });
  }
};

// src/observable/internal/list/ListsHelper.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/list/InterfaceListQuery.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/list/ListQuery.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/getObjectTypesThatInvalidate.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
async function getObjectTypesThatInvalidate(mc, objectSet) {
  const counts = {};
  const resultType = await calcObjectSet(objectSet, {
    counts,
    methodInput: void 0,
    ontologyProvider: mc.ontologyProvider
  });
  const tweaked = {
    ...counts,
    [resultType.apiName]: counts[resultType.apiName] - 1
  };
  return {
    resultType,
    counts,
    invalidationSet: new Set(Object.entries(tweaked).filter(([, v]) => v > 0).map(([k]) => k))
  };
}
async function calcObjectSet(os, ctx) {
  const op = ctx.ontologyProvider;
  async function bumpObject(apiName) {
    const objectType = await op.getObjectDefinition(apiName);
    ctx.counts[apiName] = (ctx.counts[apiName] ?? 0) + 1;
    return objectType;
  }
  async function bumpInterface(apiName) {
    const interfaceDef = await op.getInterfaceDefinition(apiName);
    for (const s of interfaceDef.implementedBy ?? []) {
      ctx.counts[s] = (ctx.counts[s] ?? 0) + 1;
    }
    return interfaceDef;
  }
  switch (os.type) {
    case "base":
      return await bumpObject(os.objectType);
    case "interfaceBase":
      return await bumpInterface(os.interfaceType);
    case "interfaceLinkSearchAround": {
      const srcDef = await calcObjectSet(os.objectSet, ctx);
      !(srcDef.type === "interface") ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false) : invariant5__default.default(false) : void 0;
      for (const [k, v] of Object.entries(srcDef.links)) {
        if (k === os.interfaceLink) {
          if (v.targetType === "object") {
            return await bumpObject(v.targetType);
          }
          return await bumpInterface(v.targetType);
        }
      }
      throw new Error(`Could not find link ${os.interfaceLink} in object set ${JSON.stringify(os.objectSet)}`);
    }
    case "searchAround": {
      const contextDef = await calcObjectSet(os.objectSet, ctx);
      !(contextDef.type === "object") ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false) : invariant5__default.default(false) : void 0;
      for (const [k, v] of Object.entries(contextDef.links)) {
        if (k === os.link) {
          return await bumpObject(v.targetType);
        }
      }
      throw new Error(`Could not find link ${os.link} in object set ${JSON.stringify(os.objectSet)}`);
    }
    case "filter":
      return calcObjectSet(os.objectSet, ctx);
    case "union":
    case "subtract":
    case "intersect":
      const returnTypes = await Promise.all(os.objectSets.map(async (os2) => {
        const counts = {};
        const r = await calcObjectSet(os2, {
          ...ctx,
          counts
        });
        return {
          r,
          counts
        };
      }));
      for (const {
        counts
      } of returnTypes) {
        for (const [k, v] of Object.entries(counts)) {
          ctx.counts[k] = Math.max(ctx.counts[k] ?? 0, v);
        }
      }
      if (returnTypes.length === 0) {
        throw new Error(`Could not find any context types for set operation`);
      }
      const allMatch = returnTypes.every(({
        r
      }) => r.apiName === returnTypes[0].r.apiName);
      if (!allMatch) {
        throw new Error(`Incompatible context types found for set operation`);
      }
      return returnTypes[0].r;
    case "withProperties":
      for (const [, v] of Object.entries(os.derivedProperties)) {
        await calcRdp(v, {
          ...ctx,
          methodInput: os.objectSet
        });
      }
      return calcObjectSet(os.objectSet, {
        ...ctx,
        methodInput: os.objectSet
      });
    // used by rdps
    case "methodInput":
      !ctx.methodInput ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "methodInput must be provided") : invariant5__default.default(false) : void 0;
      return await calcObjectSet(ctx.methodInput, {
        ...ctx,
        counts: {}
      });
    case "asType":
    // we don't currently support this anywhere.
    case "asBaseObjectTypes":
    // We don't currently support this because it could return multiple object types conceptually
    // internally, we actually use it this way but we shouldn't be finding that object sets.
    case "reference":
    // reference is particularly problematic because we cannot answer the question about
    // which types without loading the object set definition (and it can change).
    case "static":
      throw new Error(`Unsupported ObjectSet type ${os.type}`);
    case "nearestNeighbors":
      return calcObjectSet(os.objectSet, ctx);
    default:
      throw new Error(`Unhandled ObjectSet type ${os.type}`);
  }
}
async function calcRdp(dpd, ctx) {
  switch (dpd.type) {
    // Operates on object sets
    case "selection":
      return await calcObjectSet(dpd.objectSet, ctx);
    // Operates on single property
    case "negate":
    case "extract":
    case "absoluteValue":
      return await calcRdp(dpd.property, ctx);
    // Operates on many (unordered) properties
    case "least":
    case "greatest":
    case "add":
    case "multiply":
      return await Promise.all(dpd.properties.map((innerDpd) => calcRdp(innerDpd, ctx)));
    // Operates on 2 ordered properties
    case "subtract":
    case "divide":
      return await Promise.all([calcRdp(dpd.left, ctx), calcRdp(dpd.right, ctx)]);
    // Operates on a single property name
    case "property":
      return;
    default:
      throw new Error(`Unhandled DerivedPropertyDefinition type ${dpd.type}`);
  }
}

// src/observable/internal/objectMatchesWhereClause.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/evaluateFilter.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
function evaluateFilter(f, realValue, expected, strict) {
  switch (f) {
    case "$eq":
      return realValue === expected;
    case "$gt":
      return realValue > expected;
    case "$lt":
      return realValue < expected;
    case "$gte":
      return realValue >= expected;
    case "$lte":
      return realValue <= expected;
    case "$ne":
      return realValue !== expected;
    case "$in":
      return Array.isArray(expected) && expected.includes(realValue);
    case "$isNull":
      return realValue == null;
    case "$startsWith":
      return realValue.startsWith(expected);
    case "$contains":
    case "$containsAllTerms":
    case "$containsAllTermsInOrder":
    case "$containsAnyTerm":
    case "$intersects":
    case "$within":
      return !strict;
    default:
      if (process.env.NODE_ENV !== "production") {
        process.env.NODE_ENV !== "production" ? invariant5__default.default(false, `Unknown where filter ${f}`) : invariant5__default.default(false) ;
      }
      return !strict;
  }
}

// src/observable/internal/objectMatchesWhereClause.ts
function is$and(whereClause) {
  if (process.env.NODE_ENV !== "production") {
    if ("$and" in whereClause) {
      !Array.isArray(whereClause.$and) ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "expected $and to be an array") : invariant5__default.default(false) : void 0;
      !(Object.keys(whereClause).length === 1) ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "expected only $and to be present") : invariant5__default.default(false) : void 0;
    }
  }
  return "$and" in whereClause;
}
function is$or(whereClause) {
  if (process.env.NODE_ENV !== "production") {
    if ("$or" in whereClause) {
      !Array.isArray(whereClause.$or) ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "expected $or to be an array") : invariant5__default.default(false) : void 0;
      !(Object.keys(whereClause).length === 1) ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "expected only $or to be present") : invariant5__default.default(false) : void 0;
    }
  }
  return "$or" in whereClause;
}
function is$not(whereClause) {
  if (process.env.NODE_ENV !== "production") {
    if ("$not" in whereClause) {
      !(Object.keys(whereClause).length === 1) ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "expected only $not to be present") : invariant5__default.default(false) : void 0;
    }
  }
  return "$not" in whereClause;
}
function objectSortaMatchesWhereClause(o, whereClause, strict) {
  if (deepEqual__default.default({}, whereClause)) {
    return true;
  }
  if (is$and(whereClause)) {
    return whereClause.$and.every((w) => objectSortaMatchesWhereClause(o, w, strict));
  }
  if (is$or(whereClause)) {
    return whereClause.$or.some((w) => objectSortaMatchesWhereClause(o, w, strict));
  }
  if (is$not(whereClause)) {
    return !objectSortaMatchesWhereClause(o, whereClause.$not, strict);
  }
  return Object.entries(whereClause).every(([key, filter]) => {
    if (typeof filter === "object" && filter != null) {
      const realValue = o[key];
      const [f] = Object.keys(filter);
      const expected = filter[f];
      return evaluateFilter(f, realValue, expected, strict);
    }
    if (key in o) {
      if (o[key] === filter) {
        return true;
      }
    }
    return false;
  });
}

// src/observable/internal/list/ListQuery.ts
var ListQuery = class extends BaseListQuery {
  // pageSize?: number; // this is the internal page size. we need to track this properly
  #whereClause;
  // Using base class minResultsToLoad instead of a private property
  #orderBy;
  #intersectWith;
  #pivotInfo;
  #objectSet;
  /**
   * Register changes to the cache specific to ListQuery
   */
  registerCacheChanges(batch) {
    batch.changes.registerList(this.cacheKey);
  }
  constructor(store, subject, apiName, cacheKey, opts) {
    super(store, subject, opts, cacheKey, process.env.NODE_ENV !== "production" ? store.client[chunkEHWB7KEN_cjs.additionalContext].logger?.child({}, {
      msgPrefix: `ListQuery<${cacheKey.otherKeys.map((x) => JSON.stringify(x)).join(", ")}>`
    }) : void 0);
    this.apiName = apiName;
    this.#whereClause = cacheKey.otherKeys[WHERE_IDX2];
    this.#orderBy = cacheKey.otherKeys[ORDER_BY_IDX];
    this.#intersectWith = cacheKey.otherKeys[INTERSECT_IDX];
    this.#pivotInfo = cacheKey.otherKeys[PIVOT_IDX];
    this.#objectSet = this.createObjectSet(store);
    if (!this.#pivotInfo) {
      this.sortingStrategy = new OrderBySortingStrategy(this.apiName, this.#orderBy);
    }
    if (opts.autoFetchMore === true) {
      this.minResultsToLoad = Number.MAX_SAFE_INTEGER;
    } else if (typeof opts.autoFetchMore === "number") {
      this.minResultsToLoad = Math.max(0, opts.autoFetchMore);
    } else {
      this.minResultsToLoad = 0;
    }
  }
  get canonicalWhere() {
    return this.#whereClause;
  }
  get canonicalIntersectWith() {
    return this.#intersectWith;
  }
  get canonicalPivotInfo() {
    return this.#pivotInfo;
  }
  /**
   * Create the ObjectSet for this query.
   */
  /**
   * Implements fetchPageData from BaseCollectionQuery template method
   * Fetches a page of data
   */
  async fetchPageData(signal) {
    if (Object.keys(this.#orderBy).length > 0 && !(this.sortingStrategy instanceof OrderBySortingStrategy)) {
      const wireObjectSet = chunkEHWB7KEN_cjs.getWireObjectSet(this.#objectSet);
      const {
        resultType
      } = await getObjectTypesThatInvalidate(this.store.client[chunkEHWB7KEN_cjs.additionalContext], wireObjectSet);
      this.sortingStrategy = new OrderBySortingStrategy(resultType.apiName, this.#orderBy);
    }
    const resp = await this.#objectSet.fetchPage({
      $nextPageToken: this.nextPageToken,
      $pageSize: this.options.pageSize,
      $includeRid: true,
      // For now this keeps the shared test code from falling apart
      // but shouldn't be needed ideally
      ...Object.keys(this.#orderBy).length > 0 ? {
        $orderBy: this.#orderBy
      } : {}
    });
    if (signal?.aborted) {
      throw new Error("Aborted");
    }
    this.nextPageToken = resp.nextPageToken;
    const fetchedData = await this.postProcessFetchedData(resp.data);
    return {
      ...resp,
      data: fetchedData
    };
  }
  /**
   * Handle fetch errors by setting appropriate error state and notifying subscribers
   */
  handleFetchError(error, _status, batch) {
    this.logger?.error("error", error);
    this.store.subjects.get(this.cacheKey).error(error);
    const existingTotalCount = batch.read(this.cacheKey)?.value?.totalCount;
    return this.writeToStore({
      data: [],
      totalCount: existingTotalCount
    }, "error", batch);
  }
  /**
   * Will revalidate the list if its query is affected by invalidating the
   * apiName of the object type passed in.
   *
   * @param apiName to invalidate
   * @returns
   */
  /**
   * Postprocess fetched data.
   */
  invalidateObjectType = async (objectType, changes) => {
    if (this.apiName === objectType) {
      changes?.modified.add(this.cacheKey);
      return this.revalidate(true);
    }
  };
  /**
   * Note: This method is not async because I want it to return right after it
   *       finishes the synchronous updates. The promise that is returned
   *       will resolve after the revalidation is complete.
   * @param changes
   * @param optimisticId
   * @returns If revalidation is needed, a promise that resolves after the
   *          revalidation is complete. Otherwise, undefined.
   */
  maybeUpdateAndRevalidate = (changes, optimisticId) => {
    if (process.env.NODE_ENV !== "production") {
      this.logger?.child({
        methodName: "maybeUpdateAndRevalidate"
      }).debug(DEBUG_ONLY__changesToString(changes));
      this.logger?.child({
        methodName: "maybeUpdateAndRevalidate"
      }).debug(`Already in changes? ${changes.modified.has(this.cacheKey)}`);
    }
    if (changes.modified.has(this.cacheKey)) return;
    changes.modified.add(this.cacheKey);
    try {
      const relevantObjects = this._extractAndCategorizeRelevantObjects(changes);
      const status = optimisticId || relevantObjects.added.sortaMatches.size > 0 || relevantObjects.modified.sortaMatches.size > 0 ? "loading" : "loaded";
      const newList = [];
      let needsRevalidation = false;
      this.store.batch({
        optimisticId,
        changes
      }, (batch) => {
        const existingList = new Set(batch.read(this.cacheKey)?.value?.data);
        const toAdd = new Set(
          // easy case. objects are new to the cache and they match this filter
          relevantObjects.added.strictMatches
        );
        const toRemove = new Set(changes.deleted);
        for (const obj of relevantObjects.modified.all) {
          if (relevantObjects.modified.strictMatches.has(obj)) {
            const objectCacheKey = this.getObjectCacheKey(obj);
            if (!existingList.has(objectCacheKey)) {
              toAdd.add(obj);
            }
            continue;
          } else if (batch.optimisticWrite) {
            continue;
          } else {
            const existingObjectCacheKey = this.getObjectCacheKey(obj);
            toRemove.add(existingObjectCacheKey);
            if (relevantObjects.modified.sortaMatches.has(obj)) {
              needsRevalidation = true;
            }
          }
        }
        for (const key of existingList) {
          if (toRemove.has(key)) continue;
          newList.push(key);
        }
        for (const obj of toAdd) {
          newList.push(this.getObjectCacheKey(obj));
        }
        const existingTotalCount = batch.read(this.cacheKey)?.value?.totalCount;
        this._updateList(
          newList,
          status,
          batch,
          /* append */
          false,
          existingTotalCount
        );
      });
      if (needsRevalidation) {
        return this.revalidate(true);
      }
      return void 0;
    } finally {
      if (process.env.NODE_ENV !== "production") {
        this.logger?.child({
          methodName: "maybeUpdateAndRevalidate"
        }).debug("in finally");
      }
    }
  };
  #matchType(obj) {
    if (objectSortaMatchesWhereClause(obj, this.#whereClause, true)) {
      return "strict";
    }
    if (objectSortaMatchesWhereClause(obj, this.#whereClause, false)) {
      return "sorta";
    }
    return false;
  }
  _extractAndCategorizeRelevantObjects(changes) {
    const relevantObjects = this.extractRelevantObjects(changes);
    for (const group of Object.values(relevantObjects)) {
      for (const obj of group.all ?? []) {
        const matchType = this.#matchType(obj);
        if (matchType) {
          group[`${matchType}Matches`].add(obj);
        }
      }
    }
    return relevantObjects;
  }
  /**
   * Extract relevant objects for this query type.
   */
  registerStreamUpdates(sub) {
    this.createWebsocketSubscription(this.#objectSet, sub, "observeList");
  }
  onOswChange({
    object: objOrIface,
    state
  }) {
    const logger = process.env.NODE_ENV !== "production" ? this.logger?.child({
      methodName: "registerStreamUpdates"
    }) : this.logger;
    if (process.env.NODE_ENV !== "production") {
      logger?.child({
        methodName: "onChange"
      }).debug(`Got an update of type: ${state}`, objOrIface);
    }
    if (state === "ADDED_OR_UPDATED") {
      const object = objOrIface.$apiName !== objOrIface.$objectType ? objOrIface.$as(objOrIface.$objectType) : objOrIface;
      this.store.batch({}, (batch) => {
        this.store.objects.storeOsdkInstances([object], batch, this.rdpConfig);
      });
    } else if (state === "REMOVED") {
      this.onOswRemoved(objOrIface);
    }
  }
  onOswRemoved(objOrIface) {
    const logger = process.env.NODE_ENV !== "production" ? this.logger?.child({
      methodName: "onOswRemoved"
    }) : this.logger;
    this.store.batch({}, (batch) => {
      const existing = batch.read(this.cacheKey);
      !existing ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, "the truth value for our list should exist as we already subscribed") : invariant5__default.default(false) : void 0;
      if (existing.status === "loaded") {
        const objectCacheKey = this.getObjectCacheKey(objOrIface);
        const newObjects = existing.value?.data.filter((o) => o !== objectCacheKey);
        if (newObjects?.length !== existing.value?.data.length) {
          batch.changes.registerList(this.cacheKey);
          const existingTotalCount = existing.value?.totalCount;
          batch.write(this.cacheKey, {
            data: newObjects ?? [],
            totalCount: existingTotalCount
          }, "loaded");
        }
        return;
      }
      if (process.env.NODE_ENV !== "production") {
        logger?.info("Removing an object from an object list that is in the middle of being loaded.", existing);
      }
      this.revalidate(
        /* force */
        true
      ).catch((e) => {
        if (logger) {
          logger?.error("Uncaught error while revalidating list", e);
        } else {
          console.error("Uncaught error while revalidating list", e);
        }
      });
    });
  }
  /**
   * Get cache key for object.
   */
  getObjectCacheKey(obj) {
    const pk = obj.$primaryKey;
    return this.cacheKeys.get("object", obj.$objectType, pk, this.rdpConfig ?? void 0);
  }
};

// src/observable/internal/list/InterfaceListQuery.ts
var InterfaceListQuery = class extends ListQuery {
  createObjectSet(store) {
    const rdpConfig = this.cacheKey.otherKeys[RDP_IDX2];
    const pivotInfo = this.cacheKey.otherKeys[PIVOT_IDX];
    const rids = this.cacheKey.otherKeys[RIDS_IDX];
    if (pivotInfo != null) {
      const sourceSet = createSourceSetForPivot(store, pivotInfo, rids);
      let objectSet2 = sourceSet.where(this.canonicalWhere).pivotTo(pivotInfo.linkName);
      if (rdpConfig != null) {
        objectSet2 = objectSet2.withProperties(rdpConfig);
      }
      return objectSet2;
    }
    const type = "interface";
    const objectTypeDef = {
      type,
      apiName: this.apiName
    };
    const clientCtx = store.client[chunkEHWB7KEN_cjs.additionalContext];
    let objectSet;
    if (rids != null) {
      objectSet = clientCtx.objectSetFactory(objectTypeDef, clientCtx, {
        type: "static",
        objects: [...rids]
      });
    } else {
      objectSet = store.client(objectTypeDef);
    }
    if (rdpConfig != null) {
      objectSet = objectSet.withProperties(rdpConfig);
    }
    return objectSet.where(this.canonicalWhere);
  }
  async revalidateObjectType(apiName) {
    const objectMetadata = await this.store.client.fetchMetadata({
      type: "object",
      apiName
    });
    if (this.apiName in objectMetadata.interfaceMap) {
      await this.revalidate(
        /* force */
        true
      );
    }
  }
  async postProcessFetchedData(data) {
    return reloadDataAsFullObjects(this.store.client, data);
  }
  createPayload(params) {
    const resolvedList = params.resolvedData.map((obj) => obj.$as(this.apiName));
    return {
      resolvedList,
      isOptimistic: params.isOptimistic,
      fetchMore: this.fetchMore,
      hasMore: this.nextPageToken != null,
      status: params.status,
      lastUpdated: params.lastUpdated
    };
  }
  extractRelevantObjects(changes) {
    const matchesApiName = ([, object]) => {
      return this.apiName in object[chunkO3SKYLBQ_cjs.ObjectDefRef].interfaceMap;
    };
    const added = Array.from(changes.addedObjects).filter(matchesApiName).map(([, object]) => object.$as(this.apiName));
    const modified = Array.from(changes.modifiedObjects).filter(matchesApiName).map(([, object]) => object.$as(this.apiName));
    return {
      added: {
        all: added,
        strictMatches: /* @__PURE__ */ new Set(),
        sortaMatches: /* @__PURE__ */ new Set()
      },
      modified: {
        all: modified,
        strictMatches: /* @__PURE__ */ new Set(),
        sortaMatches: /* @__PURE__ */ new Set()
      }
    };
  }
};
function createSourceSetForPivot(store, pivotInfo, rids) {
  const clientCtx = store.client[chunkEHWB7KEN_cjs.additionalContext];
  if (rids != null) {
    return clientCtx.objectSetFactory({
      type: "object",
      apiName: pivotInfo.sourceType
    }, clientCtx, {
      type: "static",
      objects: [...rids]
    });
  }
  if (pivotInfo.sourceTypeKind === "interface") {
    return store.client({
      type: "interface",
      apiName: pivotInfo.sourceType
    });
  }
  return store.client({
    type: "object",
    apiName: pivotInfo.sourceType
  });
}
async function reloadDataAsFullObjects(client, data) {
  if (data.length === 0) {
    return data;
  }
  const groups = groupBy__default.default(data, (x) => x.$objectType);
  const objectTypeToPrimaryKeyToObject = Object.fromEntries(await Promise.all(Object.entries(groups).map(async ([apiName, objects]) => {
    const objectDef = await client.fetchMetadata({
      type: "object",
      apiName
    });
    const where = {
      [objectDef.primaryKeyApiName]: {
        $in: objects.map((x) => x.$primaryKey)
      }
    };
    const result = await client(objectDef).where(where).fetchPage({
      $includeRid: true
    });
    return [apiName, Object.fromEntries(result.data.map((x) => [x.$primaryKey, x]))];
  })));
  return data.map((obj) => {
    const fullObject = objectTypeToPrimaryKeyToObject[obj.$objectType][obj.$primaryKey];
    !fullObject ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false, `Could not find object ${obj.$objectType} ${obj.$primaryKey}`) : invariant5__default.default(false) : void 0;
    return fullObject;
  });
}

// src/observable/internal/list/ObjectListQuery.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
var ObjectListQuery = class extends ListQuery {
  createObjectSet(store) {
    const rdpConfig = this.cacheKey.otherKeys[RDP_IDX2];
    const intersectWith = this.cacheKey.otherKeys[INTERSECT_IDX];
    const pivotInfo = this.cacheKey.otherKeys[PIVOT_IDX];
    const rids = this.cacheKey.otherKeys[RIDS_IDX];
    const clientCtx = store.client[chunkEHWB7KEN_cjs.additionalContext];
    const typeDefinition = {
      type: "object",
      apiName: this.apiName
    };
    if (pivotInfo != null) {
      let sourceSet;
      if (rids != null) {
        sourceSet = clientCtx.objectSetFactory({
          type: "object",
          apiName: pivotInfo.sourceType
        }, clientCtx, {
          type: "static",
          objects: [...rids]
        });
      } else {
        sourceSet = pivotInfo.sourceTypeKind === "interface" ? store.client({
          type: "interface",
          apiName: pivotInfo.sourceType
        }) : store.client({
          type: "object",
          apiName: pivotInfo.sourceType
        });
      }
      let objectSet2 = sourceSet.where(this.canonicalWhere).pivotTo(pivotInfo.linkName);
      if (rdpConfig != null) {
        objectSet2 = objectSet2.withProperties(rdpConfig);
      }
      return objectSet2;
    }
    let objectSet;
    if (rids != null) {
      objectSet = clientCtx.objectSetFactory(typeDefinition, clientCtx, {
        type: "static",
        objects: [...rids]
      });
    } else {
      objectSet = store.client(typeDefinition);
    }
    if (rdpConfig != null) {
      objectSet = objectSet.withProperties(rdpConfig);
    }
    objectSet = objectSet.where(this.canonicalWhere);
    if (intersectWith != null && intersectWith.length > 0) {
      const intersectSets = intersectWith.map((whereClause) => {
        let intersectSet = store.client({
          type: "object",
          apiName: this.apiName
        });
        if (rdpConfig != null) {
          intersectSet = intersectSet.withProperties(rdpConfig);
        }
        return intersectSet.where(whereClause);
      });
      objectSet = objectSet.intersect(...intersectSets);
    }
    return objectSet;
  }
  async revalidateObjectType(apiName) {
    if (this.apiName === apiName) {
      await this.revalidate(
        /* force */
        true
      );
    }
  }
  postProcessFetchedData(data) {
    return Promise.resolve(data);
  }
  extractRelevantObjects(changes) {
    return {
      added: {
        all: changes.addedObjects.get(this.cacheKey.otherKeys[API_NAME_IDX2]) ?? [],
        strictMatches: /* @__PURE__ */ new Set(),
        sortaMatches: /* @__PURE__ */ new Set()
      },
      modified: {
        all: changes.modifiedObjects.get(this.cacheKey.otherKeys[API_NAME_IDX2]) ?? [],
        strictMatches: /* @__PURE__ */ new Set(),
        sortaMatches: /* @__PURE__ */ new Set()
      }
    };
  }
};

// src/observable/internal/list/ListsHelper.ts
var ListsHelper = class extends AbstractHelper {
  constructor(store, cacheKeys, whereCanonicalizer2, orderByCanonicalizer2, rdpCanonicalizer, intersectCanonicalizer, pivotCanonicalizer, ridListCanonicalizer) {
    super(store, cacheKeys);
    this.whereCanonicalizer = whereCanonicalizer2;
    this.orderByCanonicalizer = orderByCanonicalizer2;
    this.rdpCanonicalizer = rdpCanonicalizer;
    this.intersectCanonicalizer = intersectCanonicalizer;
    this.pivotCanonicalizer = pivotCanonicalizer;
    this.ridListCanonicalizer = ridListCanonicalizer;
  }
  observe(options, subFn) {
    const ret = super.observe(options, subFn);
    if (options.streamUpdates) {
      ret.query.registerStreamUpdates(ret.subscription);
    }
    return ret;
  }
  getQuery(options) {
    const {
      type: typeDefinition,
      where,
      orderBy,
      withProperties,
      intersectWith,
      pivotTo,
      rids
    } = options;
    const {
      apiName,
      type
    } = typeDefinition;
    const canonWhere = this.whereCanonicalizer.canonicalize(where ?? {});
    const canonOrderBy = this.orderByCanonicalizer.canonicalize(orderBy ?? {});
    const canonRdp = withProperties ? this.rdpCanonicalizer.canonicalize(withProperties) : void 0;
    const canonIntersect = intersectWith && intersectWith.length > 0 ? this.intersectCanonicalizer.canonicalize(intersectWith) : void 0;
    const canonPivot = pivotTo ? this.pivotCanonicalizer.canonicalize(apiName, type, pivotTo) : void 0;
    const canonRids = rids != null ? this.ridListCanonicalizer.canonicalize(rids) : void 0;
    const listCacheKey = this.cacheKeys.get("list", type, apiName, canonWhere, canonOrderBy, canonRdp, canonIntersect, canonPivot, canonRids);
    return this.store.queries.get(listCacheKey, () => {
      const QueryClass = type === "object" ? ObjectListQuery : InterfaceListQuery;
      return new QueryClass(this.store, this.store.subjects.get(listCacheKey), apiName, listCacheKey, options);
    });
  }
};

// src/observable/internal/object/ObjectCacheKey.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
var API_NAME_IDX3 = 0;
var RDP_CONFIG_IDX = 2;

// src/observable/internal/object/ObjectCacheKeyRegistry.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/utils/rdpFieldOperations.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
function extractRdpFieldNames(rdpConfig) {
  if (!rdpConfig) {
    return /* @__PURE__ */ new Set();
  }
  return new Set(Object.keys(rdpConfig));
}
function stripRdpFields(value, rdpFields) {
  if (rdpFields.size === 0) {
    return value;
  }
  const underlying = value[chunkO3SKYLBQ_cjs.UnderlyingOsdkObject];
  const objectDef = value[chunkO3SKYLBQ_cjs.ObjectDefRef];
  const newProps = {
    $apiName: underlying.$apiName,
    $objectType: underlying.$objectType,
    $primaryKey: underlying.$primaryKey,
    $title: underlying.$title,
    $rid: underlying.$rid
  };
  for (const key of Object.keys(underlying)) {
    if (key in objectDef.properties && !rdpFields.has(key)) {
      newProps[key] = underlying[key];
    }
  }
  return chunkO3SKYLBQ_cjs.createOsdkObject(value[chunkO3SKYLBQ_cjs.ClientRef], objectDef, newProps);
}
function isSuperset(superset, subset) {
  for (const field of subset) {
    if (!superset.has(field)) {
      return false;
    }
  }
  return true;
}
function filterToRdpFields(value, rdpFieldsToKeep, sourceRdpFields) {
  const underlying = value[chunkO3SKYLBQ_cjs.UnderlyingOsdkObject];
  const objectDef = value[chunkO3SKYLBQ_cjs.ObjectDefRef];
  const newProps = {
    $apiName: underlying.$apiName,
    $objectType: underlying.$objectType,
    $primaryKey: underlying.$primaryKey,
    $title: underlying.$title,
    $rid: underlying.$rid
  };
  for (const key of Object.keys(underlying)) {
    if (key in objectDef.properties) {
      const isRdpField = sourceRdpFields.has(key);
      if (!isRdpField || rdpFieldsToKeep.has(key)) {
        newProps[key] = underlying[key];
      }
    }
  }
  return chunkO3SKYLBQ_cjs.createOsdkObject(value[chunkO3SKYLBQ_cjs.ClientRef], objectDef, newProps);
}
function mergeObjectFields(sourceValue, sourceRdpFields, targetRdpFields, targetCurrentValue) {
  if (targetRdpFields.size === 0) {
    return stripRdpFields(sourceValue, sourceRdpFields);
  }
  if (isSuperset(sourceRdpFields, targetRdpFields)) {
    if (sourceRdpFields.size === targetRdpFields.size) {
      return sourceValue;
    }
    return filterToRdpFields(sourceValue, targetRdpFields, sourceRdpFields);
  }
  const sourceUnderlying = sourceValue[chunkO3SKYLBQ_cjs.UnderlyingOsdkObject];
  const objectDef = sourceValue[chunkO3SKYLBQ_cjs.ObjectDefRef];
  const newProps = {
    $apiName: sourceUnderlying.$apiName,
    $objectType: sourceUnderlying.$objectType,
    $primaryKey: sourceUnderlying.$primaryKey,
    $title: sourceUnderlying.$title,
    $rid: sourceUnderlying.$rid
  };
  for (const key of Object.keys(sourceUnderlying)) {
    if (key in objectDef.properties && (!sourceRdpFields.has(key) || targetRdpFields.has(key))) {
      newProps[key] = sourceUnderlying[key];
    }
  }
  if (targetCurrentValue) {
    const targetUnderlying = targetCurrentValue[chunkO3SKYLBQ_cjs.UnderlyingOsdkObject];
    for (const field of targetRdpFields) {
      if (!sourceRdpFields.has(field) && field in targetUnderlying) {
        newProps[field] = targetUnderlying[field];
      }
    }
  }
  return chunkO3SKYLBQ_cjs.createOsdkObject(sourceValue[chunkO3SKYLBQ_cjs.ClientRef], objectDef, newProps);
}

// src/observable/internal/object/ObjectCacheKeyRegistry.ts
var ObjectCacheKeyRegistry = class {
  /**
   * Map from base key (apiName:primaryKey) to all related cache key variants and metadata
   */
  baseToVariants = /* @__PURE__ */ new Map();
  /**
   * Metadata for each cache key (apiName, primaryKey, rdpConfig)
   */
  keyMetadata = /* @__PURE__ */ new WeakMap();
  /**
   * Register a cache key variant for an object
   */
  register(cacheKey, apiName, primaryKey, rdpConfig) {
    const baseKey = this.makeBaseKey(apiName, primaryKey);
    const primaryKeyStr = String(primaryKey);
    this.keyMetadata.set(cacheKey, {
      apiName,
      primaryKey: primaryKeyStr,
      rdpConfig,
      rdpFieldSet: rdpConfig ? extractRdpFieldNames(rdpConfig) : void 0
    });
    let entry = this.baseToVariants.get(baseKey);
    if (!entry) {
      entry = {
        variants: /* @__PURE__ */ new Set(),
        apiName,
        primaryKey: primaryKeyStr
      };
      this.baseToVariants.set(baseKey, entry);
    }
    entry.variants.add(cacheKey);
  }
  /**
   * Get all variant cache keys for a specific object
   */
  getVariants(apiName, primaryKey) {
    const baseKey = this.makeBaseKey(apiName, primaryKey);
    const entry = this.baseToVariants.get(baseKey);
    return new Set(entry?.variants ?? []);
  }
  /**
   * Unregister a cache key when it's being cleaned up
   */
  unregister(cacheKey) {
    const metadata = this.keyMetadata.get(cacheKey);
    if (!metadata) return;
    const baseKey = this.makeBaseKey(metadata.apiName, metadata.primaryKey);
    const entry = this.baseToVariants.get(baseKey);
    if (entry) {
      entry.variants.delete(cacheKey);
      if (entry.variants.size === 0) {
        this.baseToVariants.delete(baseKey);
      }
    }
    this.keyMetadata.delete(cacheKey);
  }
  /**
   * Get metadata for a cache key
   */
  getMetadata(cacheKey) {
    return this.keyMetadata.get(cacheKey);
  }
  /**
   * Get the count of variants for a specific object
   */
  getVariantCount(apiName, primaryKey) {
    const baseKey = this.makeBaseKey(apiName, primaryKey);
    return this.baseToVariants.get(baseKey)?.variants.size ?? 0;
  }
  /**
   * Generate a base key from apiName and primaryKey
   */
  makeBaseKey(apiName, primaryKey) {
    return `${apiName}:${primaryKey}`;
  }
  /**
   * Check if a cache key has RDP configuration
   */
  hasRdpConfig(cacheKey) {
    return this.keyMetadata.get(cacheKey)?.rdpConfig != null;
  }
  /**
   * Get the RDP configuration for a cache key
   */
  getRdpConfig(cacheKey) {
    return this.keyMetadata.get(cacheKey)?.rdpConfig;
  }
  /**
   * Get the cached RDP field set for a cache key
   */
  getRdpFieldSet(cacheKey) {
    return this.keyMetadata.get(cacheKey)?.rdpFieldSet ?? /* @__PURE__ */ new Set();
  }
};

// src/observable/internal/object/ObjectsHelper.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/object/ObjectQuery.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/BulkObjectLoader.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// ../../node_modules/.pnpm/p-defer@4.0.1/node_modules/p-defer/index.js
chunkLDTMSHUZ_cjs.init_cjs_shims();
function pDefer() {
  const deferred = {};
  deferred.promise = new Promise((resolve, reject) => {
    deferred.resolve = resolve;
    deferred.reject = reject;
  });
  return deferred;
}

// src/observable/internal/BulkObjectLoader.ts
var weakCache = new mnemonist.DefaultWeakMap((c) => new BulkObjectLoader(c));
function getBulkObjectLoader(client) {
  return weakCache.get(client);
}
var BulkObjectLoader = class {
  #client;
  #m = new mnemonist.DefaultMap(() => ({
    data: [],
    timer: void 0
  }));
  #logger;
  #maxWait;
  #maxEntries;
  constructor(client, maxWait = 25, maxEntries = 100) {
    this.#client = client;
    this.#logger = client[chunkEHWB7KEN_cjs.additionalContext].logger;
    this.#maxWait = maxWait;
    this.#maxEntries = maxEntries;
  }
  async fetch(apiName, primaryKey, defType = "object") {
    const deferred = pDefer();
    const entry = this.#m.get(apiName);
    entry.data.push({
      primaryKey,
      deferred
    });
    if (entry.defType === void 0) {
      entry.defType = defType;
    } else if (entry.defType !== defType) {
      deferred.reject(new shared_net_errors.PalantirApiError(`Conflicting defType for ${apiName}: existing=${entry.defType}, new=${defType}`));
      return deferred.promise;
    }
    if (!entry.timer) {
      entry.timer = setTimeout(() => {
        this.#loadObjects(apiName, entry.data, entry.defType ?? "object");
      }, this.#maxWait);
    }
    if (entry.data.length >= this.#maxEntries) {
      clearTimeout(entry.timer);
      this.#loadObjects(apiName, entry.data, entry.defType ?? "object");
    }
    return await deferred.promise;
  }
  #loadObjects(apiName, arr, defType) {
    this.#m.delete(apiName);
    const loadFn = defType === "interface" ? this.#loadInterfaceObjects(apiName, arr) : this.#loadObjectTypeObjects(apiName, arr);
    loadFn.catch((e) => {
      this.#logger?.error("Unhandled exception", e);
      for (const {
        primaryKey,
        deferred
      } of arr) {
        const errorMessage = e instanceof Error ? e.message : String(e);
        deferred.reject(new shared_net_errors.PalantirApiError(`Failed to load ${apiName} with pk ${primaryKey}: ${errorMessage}`));
      }
    });
  }
  async #loadObjectTypeObjects(apiName, arr) {
    const objectDef = {
      type: "object",
      apiName
    };
    const objMetadata = await this.#client.fetchMetadata(objectDef);
    const pks = arr.map((x) => x.primaryKey);
    const whereClause = pks.length === 1 ? {
      [objMetadata.primaryKeyApiName]: {
        $eq: pks[0]
      }
    } : {
      [objMetadata.primaryKeyApiName]: {
        $in: pks
      }
    };
    const {
      data
    } = await this.#client(objectDef).where(whereClause).fetchPage({
      $pageSize: pks.length,
      $includeRid: true
    });
    for (const {
      primaryKey,
      deferred
    } of arr) {
      const object = data.find((x) => x.$primaryKey === primaryKey);
      if (object) {
        deferred.resolve(object);
      } else {
        deferred.reject(new shared_net_errors.PalantirApiError(`Object not found: ${primaryKey}`));
      }
    }
  }
  async #loadInterfaceObjects(apiName, arr) {
    const pks = arr.map((x) => x.primaryKey);
    const interfaceDef = {
      type: "interface",
      apiName
    };
    const interfaceMetadata = await this.#client.fetchMetadata(interfaceDef);
    const implementingTypes = interfaceMetadata.implementedBy ?? [];
    const foundObjects = /* @__PURE__ */ new Map();
    for (const objectTypeName of implementingTypes) {
      const objectDef = {
        type: "object",
        apiName: objectTypeName
      };
      const objMetadata = await this.#client.fetchMetadata(objectDef);
      const remainingPks = pks.filter((pk) => !foundObjects.has(pk));
      if (remainingPks.length === 0) {
        break;
      }
      const whereClause = remainingPks.length === 1 ? {
        [objMetadata.primaryKeyApiName]: {
          $eq: remainingPks[0]
        }
      } : {
        [objMetadata.primaryKeyApiName]: {
          $in: remainingPks
        }
      };
      const {
        data
      } = await this.#client(objectDef).where(whereClause).fetchPage({
        $pageSize: remainingPks.length
      });
      for (const obj of data) {
        foundObjects.set(obj.$primaryKey, obj);
      }
    }
    for (const {
      primaryKey,
      deferred
    } of arr) {
      const object = foundObjects.get(primaryKey);
      if (object) {
        deferred.resolve(object);
      } else {
        deferred.reject(new shared_net_errors.PalantirApiError(`Interface ${apiName} object not found: ${primaryKey}`));
      }
    }
  }
};

// src/observable/internal/object/ObjectQuery.ts
var ObjectQuery = class extends Query {
  #apiName;
  #pk;
  #defType;
  #implementingTypes;
  constructor(store, subject, type, pk, cacheKey, opts, defType = "object") {
    super(store, subject, opts, cacheKey, process.env.NODE_ENV !== "production" ? store.client[chunkEHWB7KEN_cjs.additionalContext].logger?.child({}, {
      msgPrefix: `ObjectQuery<${cacheKey.otherKeys.map((x) => JSON.stringify(x)).join(", ")}>`
    }) : void 0);
    this.#apiName = type;
    this.#pk = pk;
    this.#defType = defType;
  }
  _createConnectable(subject) {
    return rxjs.connectable(subject.pipe(rxjs.map((x) => {
      return {
        status: x.status,
        object: x.value,
        lastUpdated: x.lastUpdated,
        isOptimistic: x.isOptimistic
      };
    })), {
      connector: () => new rxjs.BehaviorSubject({
        status: "init",
        object: void 0,
        lastUpdated: 0,
        isOptimistic: false
      })
    });
  }
  async _fetchAndStore() {
    if (process.env.NODE_ENV !== "production") {
      this.logger?.child({
        methodName: "_fetchAndStore"
      }).debug("calling _fetchAndStore");
    }
    const rdpConfig = this.cacheKey.otherKeys[RDP_CONFIG_IDX];
    let obj;
    if (rdpConfig) {
      const miniDef = {
        type: this.#defType,
        apiName: this.#apiName
      };
      const fetched = await this.store.client(miniDef).withProperties(rdpConfig).fetchOne(this.#pk, {
        $includeRid: true
      });
      obj = fetched;
    } else {
      obj = await getBulkObjectLoader(this.store.client).fetch(this.#apiName, this.#pk, this.#defType);
    }
    this.store.batch({}, (batch) => {
      this.writeToStore(obj, "loaded", batch);
    });
  }
  writeToStore(data, status, batch) {
    batch.read(this.cacheKey);
    const rdpConfig = this.cacheKey.otherKeys[RDP_CONFIG_IDX];
    this.store.objectCacheKeyRegistry.register(this.cacheKey, this.#apiName, this.#pk, rdpConfig);
    this.store.objects.propagateWrite(this.cacheKey, data, status, batch);
    return batch.read(this.cacheKey);
  }
  deleteFromStore(status, batch) {
    const rdpConfig = this.cacheKey.otherKeys[RDP_CONFIG_IDX];
    this.store.objectCacheKeyRegistry.register(this.cacheKey, this.#apiName, this.#pk, rdpConfig);
    this.store.objects.propagateWrite(this.cacheKey, tombstone, status, batch);
    return batch.read(this.cacheKey);
  }
  invalidateObjectType = async (objectType, changes) => {
    if (this.#defType === "object") {
      if (this.#apiName === objectType) {
        changes?.modified.add(this.cacheKey);
        return this.revalidate(true);
      }
      return;
    }
    if (!this.#implementingTypes) {
      const interfaceDef = {
        type: "interface",
        apiName: this.#apiName
      };
      const metadata = await this.store.client.fetchMetadata(interfaceDef);
      this.#implementingTypes = new Set(metadata.implementedBy ?? []);
    }
    if (this.#implementingTypes.has(objectType)) {
      changes?.modified.add(this.cacheKey);
      return this.revalidate(true);
    }
  };
};

// src/observable/internal/object/ObjectsHelper.ts
var ObjectsHelper = class extends AbstractHelper {
  observe(options, subFn) {
    return super.observe(options, subFn);
  }
  getQuery(options, rdpConfig) {
    const apiName = typeof options.apiName === "string" ? options.apiName : options.apiName.apiName;
    const {
      pk
    } = options;
    const defType = chunkO3SKYLBQ_cjs.getDefType(options.apiName);
    const objectCacheKey = this.cacheKeys.get("object", apiName, pk, rdpConfig ?? void 0);
    return this.store.queries.get(objectCacheKey, () => new ObjectQuery(this.store, this.store.subjects.get(objectCacheKey), apiName, pk, objectCacheKey, {
      dedupeInterval: 0
    }, defType));
  }
  /**
   * Internal helper method for writing objects to the store and returning their
   * object keys. For list queries with RDPs, the rdpConfig is included in the
   * cache key to ensure proper data isolation.
   * @internal
   */
  storeOsdkInstances(values, batch, rdpConfig) {
    return values.map((v) => this.getQuery({
      apiName: v.$objectType ?? v.$apiName,
      pk: v.$primaryKey
    }, rdpConfig).writeToStore(v, "loaded", batch).cacheKey);
  }
  /**
   * Write an object to cache and propagate to all related cache keys
   * @internal
   */
  propagateWrite(sourceCacheKey, value, status, batch) {
    const existing = batch.read(sourceCacheKey);
    const dataChanged = !existing || existing.value === void 0 || value === tombstone || !deepEqual__default.default(existing.value, value);
    const statusChanged = !existing || existing.status !== status;
    if (!dataChanged && !statusChanged) {
      return;
    }
    const valueToWrite = !dataChanged && existing ? existing.value : value;
    batch.write(sourceCacheKey, valueToWrite, status);
    if (value !== tombstone) {
      batch.changes.registerObject(sourceCacheKey, value, !existing);
    }
    const metadata = this.store.objectCacheKeyRegistry.getMetadata(sourceCacheKey);
    const relatedKeys = metadata ? this.store.objectCacheKeyRegistry.getVariants(metadata.apiName, metadata.primaryKey) : /* @__PURE__ */ new Set([sourceCacheKey]);
    for (const targetKey of relatedKeys) {
      if (targetKey === sourceCacheKey || !this.isKeyActive(targetKey)) {
        continue;
      }
      if (value === tombstone) {
        batch.write(targetKey, tombstone, status);
        continue;
      }
      const targetCurrentValue = batch.read(targetKey)?.value;
      const merged = this.mergeForTarget(value, targetCurrentValue && this.isObjectHolder(targetCurrentValue) ? targetCurrentValue : void 0, sourceCacheKey, targetKey);
      batch.write(targetKey, merged, status);
    }
  }
  /**
   * Check if a cache key is actively observed
   */
  isKeyActive(key) {
    const subject = this.store.subjects.peek(key);
    return subject?.observed === true;
  }
  /**
   * Type guard to check if a value is an ObjectHolder
   */
  isObjectHolder(value) {
    return value != null && typeof value === "object" && "$apiName" in value && "$primaryKey" in value;
  }
  /**
   * Merge object data for a specific target cache key, preserving RDP fields
   */
  mergeForTarget(sourceValue, targetCurrentValue, sourceCacheKey, targetCacheKey) {
    const sourceRdpFields = this.store.objectCacheKeyRegistry.getRdpFieldSet(sourceCacheKey);
    const targetRdpFields = this.store.objectCacheKeyRegistry.getRdpFieldSet(targetCacheKey);
    return mergeObjectFields(sourceValue, sourceRdpFields, targetRdpFields, targetCurrentValue);
  }
};

// src/observable/internal/objectset/ObjectSetHelper.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();

// src/observable/internal/objectset/ObjectSetQuery.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
var ObjectSetQuery = class extends BaseListQuery {
  #baseObjectSetWire;
  #operations;
  #composedObjectSet;
  #objectTypes;
  constructor(store, subject, baseObjectSetWire, operations, cacheKey, opts) {
    super(store, subject, opts, cacheKey, process.env.NODE_ENV !== "production" ? store.client[chunkEHWB7KEN_cjs.additionalContext].logger?.child({}, {
      msgPrefix: `ObjectSetQuery<${cacheKey.otherKeys.map((x) => JSON.stringify(x)).join(", ")}>`
    }) : void 0);
    this.#baseObjectSetWire = baseObjectSetWire;
    this.#operations = operations;
    this.#composedObjectSet = this.#composeObjectSet(opts);
    this.#objectTypes = this.#extractObjectTypes(opts);
    if (opts.autoFetchMore === true) {
      this.minResultsToLoad = Number.MAX_SAFE_INTEGER;
    } else if (typeof opts.autoFetchMore === "number") {
      this.minResultsToLoad = Math.max(0, opts.autoFetchMore);
    } else {
      this.minResultsToLoad = opts.pageSize || 0;
    }
  }
  #composeObjectSet(opts) {
    let result = opts.baseObjectSet;
    if (opts.withProperties) {
      result = result.withProperties(opts.withProperties);
    }
    if (opts.where) {
      result = result.where(opts.where);
    }
    if (opts.union && opts.union.length > 0) {
      result = result.union(...opts.union);
    }
    if (opts.intersect && opts.intersect.length > 0) {
      result = result.intersect(...opts.intersect);
    }
    if (opts.subtract && opts.subtract.length > 0) {
      result = result.subtract(...opts.subtract);
    }
    if (opts.pivotTo) {
      result = result.pivotTo(opts.pivotTo);
    }
    return result;
  }
  #extractObjectTypes(opts) {
    const types = /* @__PURE__ */ new Set();
    const baseWire = JSON.parse(this.#baseObjectSetWire);
    if (baseWire.type) {
      types.add(baseWire.type);
    }
    if (opts.union) {
      for (const os of opts.union) {
        const wire = chunkEHWB7KEN_cjs.getWireObjectSet(os);
        if (wire.type) {
          types.add(wire.type);
        }
      }
    }
    if (opts.intersect) {
      for (const os of opts.intersect) {
        const wire = chunkEHWB7KEN_cjs.getWireObjectSet(os);
        if (wire.type) {
          types.add(wire.type);
        }
      }
    }
    if (opts.subtract) {
      for (const os of opts.subtract) {
        const wire = chunkEHWB7KEN_cjs.getWireObjectSet(os);
        if (wire.type) {
          types.add(wire.type);
        }
      }
    }
    return types;
  }
  /**
   * Register changes to the cache specific to ObjectSetQuery
   */
  registerCacheChanges(batch) {
    batch.changes.registerObjectSet(this.cacheKey);
  }
  /**
   * Implements fetchPageData from BaseListQuery template method
   * Fetches a page of data from the composed ObjectSet
   */
  async fetchPageData(signal) {
    if (this.#operations.orderBy && Object.keys(this.#operations.orderBy).length > 0 && !(this.sortingStrategy instanceof OrderBySortingStrategy)) {
      const wireObjectSet = chunkEHWB7KEN_cjs.getWireObjectSet(this.#composedObjectSet);
      const {
        resultType
      } = await getObjectTypesThatInvalidate(this.store.client[chunkEHWB7KEN_cjs.additionalContext], wireObjectSet);
      this.sortingStrategy = new OrderBySortingStrategy(resultType.apiName, this.#operations.orderBy);
    }
    const resp = await this.#composedObjectSet.fetchPage({
      $nextPageToken: this.nextPageToken,
      $pageSize: this.options.pageSize,
      $includeRid: true,
      // OrderBy is already applied in the composed ObjectSet
      ...this.#operations.orderBy && Object.keys(this.#operations.orderBy).length > 0 ? {
        $orderBy: this.#operations.orderBy
      } : {}
    });
    if (signal?.aborted) {
      throw new Error("Aborted");
    }
    this.nextPageToken = resp.nextPageToken;
    return resp;
  }
  handleFetchError(error, _status, batch) {
    this.logger?.error("error", error);
    this.store.subjects.get(this.cacheKey).error(error);
    const existingTotalCount = batch.read(this.cacheKey)?.value?.totalCount;
    return this.writeToStore({
      data: [],
      totalCount: existingTotalCount
    }, "error", batch);
  }
  registerStreamUpdates(sub) {
    this.createWebsocketSubscription(this.#composedObjectSet, sub, "observeObjectSet");
  }
  invalidateObjectType = async (objectType, changes) => {
    if (this.#objectTypes.has(objectType)) {
      changes?.modified.add(this.cacheKey);
      return this.revalidate(true);
    }
    return Promise.resolve();
  };
  createPayload(params) {
    return {
      resolvedList: params.resolvedData,
      isOptimistic: params.isOptimistic,
      fetchMore: this.fetchMore,
      hasMore: this.nextPageToken != null,
      status: params.status,
      lastUpdated: params.lastUpdated,
      objectSet: this.#composedObjectSet,
      totalCount: params.totalCount
    };
  }
};

// src/observable/internal/objectset/ObjectSetHelper.ts
var ObjectSetHelper = class extends AbstractHelper {
  constructor(store, cacheKeys, whereCanonicalizer2, orderByCanonicalizer2) {
    super(store, cacheKeys);
    this.whereCanonicalizer = whereCanonicalizer2;
    this.orderByCanonicalizer = orderByCanonicalizer2;
  }
  observe(options, subFn) {
    const ret = super.observe(options, subFn);
    if (options.streamUpdates) {
      ret.query.registerStreamUpdates(ret.subscription);
    }
    return ret;
  }
  getQuery(options) {
    const {
      baseObjectSet
    } = options;
    const baseObjectSetWire = JSON.stringify(chunkEHWB7KEN_cjs.getWireObjectSet(baseObjectSet));
    const operations = this.buildCanonicalizedOperations(options);
    const objectSetCacheKey = this.cacheKeys.get("objectSet", baseObjectSetWire, operations);
    return this.store.queries.get(objectSetCacheKey, () => {
      return new ObjectSetQuery(this.store, this.store.subjects.get(objectSetCacheKey), baseObjectSetWire, operations, objectSetCacheKey, options);
    });
  }
  buildCanonicalizedOperations(options) {
    const operations = {};
    if (options.where) {
      operations.where = this.whereCanonicalizer.canonicalize(options.where);
    }
    if (options.withProperties) {
      operations.withProperties = Object.keys(options.withProperties).sort();
    }
    if (options.union && options.union.length > 0) {
      operations.union = options.union.map((os) => JSON.stringify(chunkEHWB7KEN_cjs.getWireObjectSet(os)));
    }
    if (options.intersect && options.intersect.length > 0) {
      operations.intersect = options.intersect.map((os) => JSON.stringify(chunkEHWB7KEN_cjs.getWireObjectSet(os)));
    }
    if (options.subtract && options.subtract.length > 0) {
      operations.subtract = options.subtract.map((os) => JSON.stringify(chunkEHWB7KEN_cjs.getWireObjectSet(os)));
    }
    if (options.pivotTo) {
      operations.pivotTo = options.pivotTo;
    }
    if (options.orderBy) {
      operations.orderBy = this.orderByCanonicalizer.canonicalize(options.orderBy);
    }
    if (options.pageSize) {
      operations.pageSize = options.pageSize;
    }
    return operations;
  }
};

// src/observable/internal/PivotCanonicalizer.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
var PivotCanonicalizer = class {
  #cache = /* @__PURE__ */ new Map();
  canonicalize(sourceType, sourceTypeKind, linkName) {
    const key = `${sourceTypeKind}:${sourceType}::${linkName}`;
    let canonical = this.#cache.get(key);
    if (!canonical) {
      canonical = {
        sourceType,
        sourceTypeKind,
        linkName
      };
      this.#cache.set(key, canonical);
    }
    return canonical;
  }
};

// src/observable/internal/Queries.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
var Queries = class {
  // we can use a regular Map here because the refCounting will
  // handle cleanup.
  map = /* @__PURE__ */ new Map();
  peek(cacheKey) {
    return this.map.get(cacheKey);
  }
  get(cacheKey, createQuery) {
    let query = this.peek(cacheKey);
    if (!query) {
      query = createQuery();
      this.map.set(cacheKey, query);
    }
    return query;
  }
  keys() {
    return this.map.keys();
  }
  delete(cacheKey) {
    this.map.get(cacheKey)?.dispose();
    this.map.delete(cacheKey);
  }
};

// src/observable/internal/RdpCanonicalizer.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
var RdpCanonicalizer = class extends CachingCanonicalizer {
  structuralCache = /* @__PURE__ */ new Map();
  lookupOrCreate(rdp) {
    const definitionMap = /* @__PURE__ */ new Map();
    const computedProperties = {};
    const objectTypeHolder = {
      type: "object",
      apiName: "__rdp_canonicalizer_holder__"
    };
    for (const [key, rdpFunction] of Object.entries(rdp)) {
      const builder = chunkEHWB7KEN_cjs.createWithPropertiesObjectSet(
        objectTypeHolder,
        {
          type: "methodInput"
        },
        definitionMap,
        /* fromBaseObjectSet */
        true
      );
      const result = rdpFunction(builder);
      const definition = definitionMap.get(result);
      if (definition) {
        computedProperties[key] = definition;
      }
    }
    const sortedKeys = Object.keys(computedProperties).sort();
    const sortedDefinitions = {};
    for (const key of sortedKeys) {
      sortedDefinitions[key] = computedProperties[key];
    }
    const definitionsKey = JSON.stringify(sortedDefinitions);
    let canonical = this.structuralCache.get(definitionsKey);
    if (!canonical) {
      const sortedRdp = {};
      for (const key of Object.keys(rdp).sort()) {
        sortedRdp[key] = rdp[key];
      }
      canonical = sortedRdp;
      this.structuralCache.set(definitionsKey, canonical);
    }
    return canonical;
  }
};

// src/observable/internal/RidListCanonicalizer.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
var RidListCanonicalizer = class {
  cache = /* @__PURE__ */ new Map();
  canonicalize(rids) {
    const sorted = [...new Set(rids)].sort();
    const key = sorted.join("\0");
    let canonical = this.cache.get(key);
    if (!canonical) {
      canonical = sorted;
      this.cache.set(key, canonical);
    }
    return canonical;
  }
};

// src/observable/internal/Store.ts
var Store = class {
  whereCanonicalizer = new WhereClauseCanonicalizer();
  orderByCanonicalizer = new OrderByCanonicalizer();
  rdpCanonicalizer = new RdpCanonicalizer();
  intersectCanonicalizer = new IntersectCanonicalizer(this.whereCanonicalizer);
  pivotCanonicalizer = new PivotCanonicalizer();
  ridListCanonicalizer = new RidListCanonicalizer();
  /** @internal */
  queries = new Queries();
  objectCacheKeyRegistry = new ObjectCacheKeyRegistry();
  layers = new Layers({
    logger: this.logger,
    onRevalidate: this.#maybeRevalidateQueries.bind(this)
  });
  subjects = this.layers.subjects;
  // these are hopefully temporary
  constructor(client) {
    this.logger = client[chunkEHWB7KEN_cjs.additionalContext].logger?.child({}, {
      msgPrefix: "Store"
    });
    this.client = client;
    this.cacheKeys = new CacheKeys({
      onDestroy: this.#cleanupCacheKey
    });
    this.aggregations = new AggregationsHelper(this, this.cacheKeys, this.whereCanonicalizer, this.rdpCanonicalizer);
    this.functions = new FunctionsHelper(this, this.cacheKeys);
    this.lists = new ListsHelper(this, this.cacheKeys, this.whereCanonicalizer, this.orderByCanonicalizer, this.rdpCanonicalizer, this.intersectCanonicalizer, this.pivotCanonicalizer, this.ridListCanonicalizer);
    this.objects = new ObjectsHelper(this, this.cacheKeys);
    this.links = new LinksHelper(this, this.cacheKeys, this.whereCanonicalizer, this.orderByCanonicalizer);
    this.objectSets = new ObjectSetHelper(this, this.cacheKeys, this.whereCanonicalizer, this.orderByCanonicalizer);
  }
  /**
   * Called after a key is no longer retained and the timeout has elapsed
   * @param key
   */
  #cleanupCacheKey = (key) => {
    const subject = this.subjects.peek(key);
    if (process.env.NODE_ENV !== "production") {
      !subject ? process.env.NODE_ENV !== "production" ? invariant5__default.default(false) : invariant5__default.default(false) : void 0;
    }
    this.subjects.delete(key);
    this.queries.delete(key);
  };
  applyAction = async (action, args, opts) => {
    return await new ActionApplication(this).applyAction(action, args, opts);
  };
  validateAction = async (action, args) => {
    const result = await this.client(action).applyAction(args, {
      $validateOnly: true,
      $returnEdits: false
    });
    return result;
  };
  getValue(cacheKey) {
    return this.layers.top.get(cacheKey);
  }
  batch({
    optimisticId,
    changes = createChangedObjects()
  }, batchFn) {
    return this.layers.batch({
      optimisticId,
      changes
    }, batchFn);
  }
  invalidateObject(apiName, pk) {
    if (typeof apiName !== "string") {
      apiName = apiName.apiName;
    }
    const variants = this.objectCacheKeyRegistry.getVariants(apiName, pk);
    const promises = [];
    if (variants.size === 0) {
      promises.push(this.objects.getQuery({
        apiName,
        pk
      }, void 0).revalidate(
        /* force */
        true
      ));
    } else {
      for (const key of variants) {
        const query = this.queries.peek(key);
        if (query) {
          promises.push(query.revalidate(
            /* force */
            true
          ));
        }
      }
    }
    return Promise.allSettled(promises);
  }
  async #maybeRevalidateQueries(changes, optimisticId) {
    const logger = process.env.NODE_ENV !== "production" ? this.logger?.child({
      methodName: "maybeRevalidateQueries"
    }) : void 0;
    if (changes.isEmpty()) {
      if (process.env.NODE_ENV !== "production") {
        logger?.debug("No changes, aborting");
      }
      return;
    }
    if (process.env.NODE_ENV !== "production") {
      logger?.debug(DEBUG_ONLY__changesToString(changes), {
        optimisticId
      });
    }
    try {
      const promises = [];
      for (const cacheKey of this.queries.keys()) {
        const query = this.queries.peek(cacheKey);
        if (!query?.maybeUpdateAndRevalidate) {
          continue;
        }
        if (!this.#shouldPropagateToQuery({
          cacheKey,
          maybeUpdateAndRevalidate: query.maybeUpdateAndRevalidate
        }, changes, optimisticId)) {
          continue;
        }
        const promise = query.maybeUpdateAndRevalidate(changes, optimisticId);
        if (promise) promises.push(promise);
      }
      await Promise.all(promises);
    } finally {
      if (process.env.NODE_ENV !== "production") {
        logger?.debug("in finally", DEBUG_ONLY__changesToString(changes));
      }
    }
  }
  /**
   * Determines whether changes should propagate to a specific query.
   * Prevents unnecessary observable pipeline execution for cross-propagation.
   *
   * @param query - The query to check
   * @param changes - The changes that occurred
   * @param optimisticId - Optional optimistic update ID
   * @returns true if the query should be notified of these changes
   */
  #shouldPropagateToQuery(query, changes, optimisticId) {
    if (optimisticId) {
      return true;
    }
    if (changes.modified.has(query.cacheKey)) {
      return true;
    }
    if (this.#shouldPropagateForObjectTypeChanges(query.cacheKey, changes)) {
      return true;
    }
    const queryRdpConfig = this.#getQueryRdpConfig(query.cacheKey);
    return queryRdpConfig != null;
  }
  /**
   * Checks if changes to an object type should propagate to a query.
   * This ensures queries receive updates when objects of their type are added/modified.
   *
   * @param cacheKey - The cache key of the query
   * @param changes - The changes that occurred
   * @returns true if the query should be notified based on object type changes
   */
  #shouldPropagateForObjectTypeChanges(cacheKey, changes) {
    const queryObjectType = this.#getQueryObjectType(cacheKey);
    if (!queryObjectType) {
      return false;
    }
    const affected = this.#changesAffectObjectType(changes, queryObjectType);
    if (process.env.NODE_ENV !== "production") {
      this.logger?.child({
        methodName: "shouldPropagateToQuery"
      }).debug(`Query type: ${queryObjectType}, affected: ${affected}`, {
        queryKey: DEBUG_ONLY__cacheKeyToString(cacheKey),
        addedCount: changes.addedObjects.get(queryObjectType)?.length ?? 0,
        modifiedCount: changes.modifiedObjects.get(queryObjectType)?.length ?? 0
      });
    }
    return affected;
  }
  /**
   * Extracts RDP configuration from a cache key if present.
   *
   * @param cacheKey - The cache key to check
   * @returns The RDP configuration, null, or undefined
   */
  #getQueryRdpConfig(cacheKey) {
    if ("otherKeys" in cacheKey && Array.isArray(cacheKey.otherKeys)) {
      if (cacheKey.type === "object") {
        return cacheKey.otherKeys[RDP_CONFIG_IDX];
      } else if (cacheKey.type === "list") {
        return cacheKey.otherKeys[RDP_IDX2];
      } else if (cacheKey.type === "aggregation") {
        return cacheKey.otherKeys[RDP_IDX];
      }
    }
    return void 0;
  }
  /**
   * Extracts the object type (apiName) from a cache key.
   *
   * @param cacheKey - The cache key to check
   * @returns The object type/apiName, or undefined if not applicable
   */
  #getQueryObjectType(cacheKey) {
    if ("otherKeys" in cacheKey && Array.isArray(cacheKey.otherKeys)) {
      if (cacheKey.type === "object") {
        return cacheKey.otherKeys[API_NAME_IDX3];
      } else if (cacheKey.type === "list") {
        return cacheKey.otherKeys[API_NAME_IDX2];
      } else if (cacheKey.type === "aggregation") {
        return cacheKey.otherKeys[API_NAME_IDX];
      }
    }
    return void 0;
  }
  /**
   * Checks if changes affect a specific object type.
   *
   * @param changes - The changes to check
   * @param objectType - The object type to check for
   * @returns true if the changes include added or modified objects of this type
   */
  #changesAffectObjectType(changes, objectType) {
    const addedForType = changes.addedObjects.get(objectType);
    if (addedForType && addedForType.length > 0) {
      return true;
    }
    const modifiedForType = changes.modifiedObjects.get(objectType);
    if (modifiedForType && modifiedForType.length > 0) {
      return true;
    }
    return false;
  }
  /**
   * Invalidates all cache entries for a specific object type.
   * This will revalidate:
   * 1. All objects of the specified type
   * 2. All lists of the specified type
   * 3. All links where the source object is of the specified type
   *
   * @param apiName - The API name of the object type to invalidate
   * @param changes - Optional changes object to track what has been modified
   * @returns Promise that resolves when all invalidations are complete
   */
  invalidateObjectType(apiName, changes) {
    if (typeof apiName !== "string") {
      apiName = apiName.apiName;
    }
    if (process.env.NODE_ENV !== "production") {
      this.logger?.child({
        methodName: "invalidateObjectType"
      }).info(changes ? DEBUG_ONLY__changesToString(changes) : void 0);
    }
    const promises = [];
    for (const cacheKey of this.layers.truth.keys()) {
      if (changes && changes.modified.has(cacheKey)) {
        continue;
      }
      const query = this.queries.peek(cacheKey);
      if (!query) continue;
      promises.push(query.invalidateObjectType(apiName, changes));
    }
    return Promise.allSettled(promises).then(() => void 0);
  }
  async invalidateAll() {
    const promises = [];
    for (const cacheKey of this.queries.keys()) {
      const query = this.queries.peek(cacheKey);
      if (query) {
        promises.push(query.revalidate(true));
      }
    }
    return Promise.allSettled(promises).then(() => void 0);
  }
  async invalidateObjects(objects) {
    const objectsArray = Array.isArray(objects) ? objects : [objects];
    const promises = [];
    for (const obj of objectsArray) {
      promises.push(this.invalidateObject(obj.$objectType, obj.$primaryKey));
    }
    return Promise.allSettled(promises).then(() => void 0);
  }
  async invalidateFunction(apiName, params) {
    return this.functions.invalidateFunction(apiName, params);
  }
  async invalidateFunctionsByObject(apiName, primaryKey) {
    return this.functions.invalidateFunctionsByObject(apiName, primaryKey);
  }
};

// src/observable/ObservableClient.ts
function createObservableClient(client) {
  const tweakedClient = chunkO3SKYLBQ_cjs.createClientFromContext({
    ...client[chunkEHWB7KEN_cjs.additionalContext],
    fetch: shared_net_fetch.createFetchHeaderMutator(client[chunkEHWB7KEN_cjs.additionalContext].fetch, (headers) => {
      headers.set("Fetch-User-Agent", [headers.get("Fetch-User-Agent"), chunkO3SKYLBQ_cjs.OBSERVABLE_USER_AGENT].filter((x) => x && x?.length > 0).join(" "));
      return headers;
    })
  });
  return new ObservableClientImpl(new Store(tweakedClient));
}

// src/public-utils/osdkConfig.ts
chunkLDTMSHUZ_cjs.init_cjs_shims();
function getMetaTagContent(name) {
  const element = document.querySelector(`meta[name="${name}"]`);
  const val = element ? element.getAttribute("content") : null;
  if (val == null) {
    throw new Error(`Missing meta tag: ${name}`);
  }
  return val;
}
function getViteEnvVar(name) {
  const val = undefined[name];
  if (val == null) {
    throw new Error(`Missing environment variable: ${name}`);
  }
  return val;
}
function isProduction() {
  return process.env.NODE_ENV === "production";
}
function getConfigValue(metaTagName, viteEnvVarName) {
  return isProduction() ? getMetaTagContent(metaTagName) : getViteEnvVar(viteEnvVarName);
}
function getOntologyRid(ontologyRid) {
  return isProduction() ? getMetaTagContent("osdk-ontologyRid") : ontologyRid;
}
function getOsdkConfig(ontologyRid) {
  return {
    clientId: getConfigValue("osdk-clientId", "VITE_FOUNDRY_CLIENT_ID"),
    redirectUrl: getConfigValue("osdk-redirectUrl", "VITE_FOUNDRY_REDIRECT_URL"),
    foundryUrl: getConfigValue("osdk-foundryUrl", "VITE_FOUNDRY_API_URL"),
    ontologyRid: getOntologyRid(ontologyRid)
  };
}

Object.defineProperty(exports, "createClientWithTransaction", {
  enumerable: true,
  get: function () { return chunkO3SKYLBQ_cjs.createClientWithTransaction; }
});
Object.defineProperty(exports, "augment", {
  enumerable: true,
  get: function () { return chunkEHWB7KEN_cjs.augment; }
});
exports.computeObjectSetCacheKey = computeObjectSetCacheKey;
exports.createObservableClient = createObservableClient;
exports.getMetaTagContent = getMetaTagContent;
exports.getOsdkConfig = getOsdkConfig;
