// @osdk/client/build/esm/observable/internal/function/FunctionQuery.js
/*
 * Copyright 2025 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { BehaviorSubject, connectable, map } from "rxjs";
import { additionalContext } from "../../../Client.js";
import { applyQuery } from "../../../queries/applyQuery.js";
import { Query } from "../Query.js";
export class FunctionQuery extends Query {
  #apiName;
  #version;
  #params;
  #dependsOn;
  #dependsOnObjects;
  #queryDef;
  constructor(store, subject, queryDef, params, cacheKey, opts) {
    super(store, subject, opts, cacheKey, process.env.NODE_ENV !== "production" ? store.client[additionalContext].logger?.child({}, {
      msgPrefix: `FunctionQuery<${cacheKey.otherKeys.map(x => JSON.stringify(x)).join(", ")}>`
    }) : undefined);
    this.#apiName = queryDef.apiName;
    this.#version = queryDef.isFixedVersion ? queryDef.version : undefined;
    this.#params = params;
    this.#dependsOn = opts.dependsOn;
    this.#dependsOnObjects = opts.dependsOnObjects;
    this.#queryDef = queryDef;
  }
  _createConnectable(subject) {
    return connectable(subject.pipe(map(x => {
      const value = x.value;
      return {
        status: x.status,
        result: value?.result,
        lastUpdated: value?.executedAt ?? 0,
        error: value?.error
      };
    })), {
      connector: () => new BehaviorSubject({
        status: "init",
        result: undefined,
        lastUpdated: 0
      })
    });
  }
  async _fetchAndStore() {
    if (process.env.NODE_ENV !== "production") {
      this.logger?.child({
        methodName: "_fetchAndStore"
      }).debug("calling _fetchAndStore");
    }
    try {
      // Type assertion needed because FunctionQuery we determine types dynamically
      // at runtime without compile-time parameter type info. applyQuery internally
      // converts params via remapQueryParams which handles the type safely.
      const result = await applyQuery(this.store.client[additionalContext], this.#queryDef, this.#params);
      const executedAt = Date.now();
      this.store.batch({}, batch => {
        this.writeToStore({
          result,
          executedAt
        }, "loaded", batch);
      });
    } catch (e) {
      if (process.env.NODE_ENV !== "production") {
        this.logger?.child({
          methodName: "_fetchAndStore"
        }).error("Error executing function", e);
      }
      const error = e instanceof Error ? e : new Error(String(e));
      this.store.batch({}, batch => {
        this.writeToStore({
          result: undefined,
          executedAt: 0,
          error
        }, "error", batch);
      });
    }
  }
  writeToStore(data, status, batch) {
    batch.write(this.cacheKey, data, status);
    return batch.read(this.cacheKey);
  }
  invalidateObjectType = (objectType, changes) => {
    // Check if this function depends on the given object type
    if (this.#dependsOn?.includes(objectType)) {
      changes?.registerFunction(this.cacheKey);
      return this.revalidate(true);
    }
    return Promise.resolve();
  };
  dependsOnObject(apiName, primaryKey) {
    if (!this.#dependsOnObjects) {
      return false;
    }
    return this.#dependsOnObjects.some(obj => obj.$apiName === apiName && obj.$primaryKey === primaryKey);
  }

  /**
   * Called during batch operations when objects change.
   * Checks if any objects in dependsOnObjects were modified/added
   * and triggers revalidation if so.
   */
  maybeUpdateAndRevalidate = changes => {
    if (!this.#dependsOnObjects?.length) {
      return undefined;
    }
    for (const dep of this.#dependsOnObjects) {
      const modifiedObjects = changes.modifiedObjects.get(dep.$apiName);
      if (modifiedObjects?.some(obj => obj.$primaryKey === dep.$primaryKey)) {
        return this.revalidate(true);
      }
      const addedObjects = changes.addedObjects.get(dep.$apiName);
      if (addedObjects?.some(obj => obj.$primaryKey === dep.$primaryKey)) {
        return this.revalidate(true);
      }
    }
    return undefined;
  };
  get apiName() {
    return this.#apiName;
  }
  get version() {
    return this.#version;
  }
}
