// @osdk/client/build/browser/object/formatting/formatNumber.js
/*
 * Copyright 2025 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { resolvePropertyReference } from "./propertyFormattingUtils.js";

/**
 * Extended NumberFormatOptions that includes rounding mode support
 * which is available in modern browsers but not yet in TypeScript types.
 * Only includes modes that are actually supported by Intl.NumberFormat and used in our API.
 */

/**
 * Formats a number value according to the specified formatting rule
 */
export function formatNumber(value, numberType, objectData, locale) {
  switch (numberType.type) {
    case "standard":
      return formatStandardNumber(value, numberType.baseFormatOptions, locale);
    case "fixedValues":
      return numberType.values[value];
    case "currency":
      return formatCurrency(value, numberType, objectData, locale);
    case "standardUnit":
      return formatStandardUnit(value, numberType, objectData, locale);
    case "customUnit":
      return formatCustomUnit(value, numberType, objectData, locale);
    case "affix":
      return formatAffix(value, numberType, objectData, locale);
    case "duration":
      // TODO (duration is a bit more complex)
      return undefined;
    case "scale":
      return formatScale(value, numberType, locale);
    case "ratio":
      return formatRatio(value, numberType, locale);
    default:
      return undefined;
  }
}
function formatStandardNumber(value, options, locale) {
  return formatNumberWithAffixes(value, options, locale);
}
function formatCurrency(value, rule, objectData, locale) {
  const currencyCode = resolvePropertyReference(rule.currencyCode, objectData);
  if (currencyCode == null) {
    return formatStandardNumber(value, rule.baseFormatOptions, locale);
  }
  const options = {
    ...convertToIntlOptions(rule.baseFormatOptions),
    style: "currency",
    currency: currencyCode,
    currencyDisplay: rule.style === "COMPACT" ? "narrowSymbol" : "symbol"
  };
  return formatWithIntl(value, options, locale);
}
function formatStandardUnit(value, rule, objectData, locale) {
  const unit = resolvePropertyReference(rule.unit, objectData);
  if (unit == null) {
    return formatStandardNumber(value, rule.baseFormatOptions, locale);
  }
  const options = {
    ...convertToIntlOptions(rule.baseFormatOptions),
    style: "unit",
    unit: unit
  };
  try {
    return formatWithIntl(value, options, locale);
  } catch (error) {
    return formatNumberWithAffixes(value, rule.baseFormatOptions, locale, "", ` ${unit}`);
  }
}
function formatCustomUnit(value, rule, objectData, locale) {
  const unit = resolvePropertyReference(rule.unit, objectData);
  const suffix = unit != null ? ` ${unit}` : "";
  return formatNumberWithAffixes(value, rule.baseFormatOptions, locale, "", suffix);
}
function formatAffix(value, rule, objectData, locale) {
  const prefix = rule.affix.prefix != null ? resolvePropertyReference(rule.affix.prefix, objectData) ?? "" : "";
  const suffix = rule.affix.postfix != null ? resolvePropertyReference(rule.affix.postfix, objectData) ?? "" : "";
  return formatNumberWithAffixes(value, rule.baseFormatOptions, locale, prefix, suffix);
}

/**
 * Gets the scale divisor for a given scale type
 */
function getScaleDivisor(scaleType) {
  switch (scaleType) {
    case "THOUSANDS":
      return 1e3;
    case "MILLIONS":
      return 1e6;
    case "BILLIONS":
      return 1e9;
    default:
      throw new Error(`Unknown scale type: ${scaleType}`);
  }
}

/**
 * Extracts the localized compact suffix for a given scale.
 * Uses Intl.NumberFormat with compact notation to get the proper i18n suffix.
 */
function getLocalizedCompactSuffix(scaleDivisor, locale) {
  const compactFormatter = new Intl.NumberFormat(locale, {
    notation: "compact",
    compactDisplay: "short"
  });
  const parts = compactFormatter.formatToParts(scaleDivisor);
  const compactPart = parts.find(p => p.type === "compact");
  return compactPart?.value ?? "";
}
function formatScale(value, rule, locale) {
  const scaleDivisor = getScaleDivisor(rule.scaleType);
  const suffix = getLocalizedCompactSuffix(scaleDivisor, locale);
  return formatNumberWithAffixes(value / scaleDivisor, rule.baseFormatOptions, locale, "", suffix);
}
function getRatioScaledValue(value, ratioType) {
  switch (ratioType) {
    case "PERCENTAGE":
      return value * 100;
    case "PER_MILLE":
      return value * 1000;
    case "BASIS_POINTS":
      return value * 10000;
    default:
      throw new Error(`Unknown ratio type: ${ratioType}`);
  }
}
function formatRatio(value, rule, locale) {
  // Special case: PERCENTAGE uses Intl's native percent style
  if (rule.ratioType === "PERCENTAGE") {
    const options = {
      ...convertToIntlOptions(rule.baseFormatOptions),
      style: "percent"
    };
    return formatWithIntl(value, options, locale);
  }

  // Other ratio types: scale and add suffix
  const scaledValue = getRatioScaledValue(value, rule.ratioType);
  let suffix;
  switch (rule.ratioType) {
    case "PER_MILLE":
      suffix = "‰";
      break;
    case "BASIS_POINTS":
      suffix = " bps";
      break;
    default:
      rule.ratioType;
      throw new Error(`Unknown ratio type: ${rule.ratioType}`);
  }
  return formatNumberWithAffixes(scaledValue, rule.baseFormatOptions, locale, "", suffix);
}

/**
 * Maps notation from OSDK format to Intl format
 */
function mapNotation(notation) {
  switch (notation) {
    case "STANDARD":
      return "standard";
    case "SCIENTIFIC":
      return "scientific";
    case "ENGINEERING":
      return "engineering";
    case "COMPACT":
      return "compact";
    default:
  }
}

/**
 * Maps rounding mode from OSDK format to Intl format
 */
function mapRoundingMode(mode) {
  switch (mode) {
    case "CEIL":
      return "ceil";
    case "FLOOR":
      return "floor";
    case "ROUND_CLOSEST":
      return "halfExpand";
    default:
      return undefined;
  }
}
function convertToIntlOptions(options) {
  return {
    useGrouping: options.useGrouping,
    minimumIntegerDigits: options.minimumIntegerDigits,
    minimumFractionDigits: options.minimumFractionDigits,
    maximumFractionDigits: options.maximumFractionDigits,
    minimumSignificantDigits: options.minimumSignificantDigits,
    maximumSignificantDigits: options.maximumSignificantDigits,
    notation: options.notation ? mapNotation(options.notation) : undefined,
    roundingMode: options.roundingMode ? mapRoundingMode(options.roundingMode) : undefined,
    convertNegativeToParenthesis: options.convertNegativeToParenthesis
  };
}

/**
 * Applies negative-to-parenthesis conversion if requested
 */
function maybeConvertNegativeToParenthesis(formatted, value, shouldConvert) {
  if (shouldConvert && value < 0) {
    return formatted.replace(/^-/, "(") + ")";
  }
  return formatted;
}

/**
 * Formats a number using Intl.NumberFormat with the given options
 */
function formatWithIntl(value, options, locale) {
  const formatter = new Intl.NumberFormat(locale, options);
  const formatted = formatter.format(value);
  return maybeConvertNegativeToParenthesis(formatted, value, options.convertNegativeToParenthesis ?? false);
}

/**
 * Formats a number and optionally adds prefix/suffix
 */
function formatNumberWithAffixes(value, baseOptions, locale, prefix, suffix) {
  const intlOptions = convertToIntlOptions(baseOptions);
  const formatted = formatWithIntl(value, intlOptions, locale);
  return `${prefix || ""}${formatted}${suffix || ""}`;
}
