// @osdk/foundry.orchestration/build/browser/public/Schedule.js
/*
 * Copyright 2024 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { foundryPlatformFetch as $foundryPlatformFetch } from "@osdk/shared.net.platformapi";
//
const _create = [1, "/v2/orchestration/schedules", 3];
/**
 * Creates a new Schedule.
 *
 * @beta
 *
 * Required Scopes: [api:orchestration-write]
 * URL: /v2/orchestration/schedules
 */
export function create($ctx, ...args) {
  return $foundryPlatformFetch($ctx, _create, ...args);
}
const _deleteSchedule = [3, "/v2/orchestration/schedules/{0}"];
/**
 * Delete the Schedule with the specified rid.
 *
 * @public
 *
 * Required Scopes: [api:orchestration-write]
 * URL: /v2/orchestration/schedules/{scheduleRid}
 */
export function deleteSchedule($ctx, ...args) {
  return $foundryPlatformFetch($ctx, _deleteSchedule, ...args);
}
const _get = [0, "/v2/orchestration/schedules/{0}", 2];
/**
 * Get the Schedule with the specified rid.
 *
 * @beta
 *
 * Required Scopes: [api:orchestration-read]
 * URL: /v2/orchestration/schedules/{scheduleRid}
 */
export function get($ctx, ...args) {
  return $foundryPlatformFetch($ctx, _get, ...args);
}
const _getBatch = [1, "/v2/orchestration/schedules/getBatch", 3];
/**
 * Fetch multiple schedules in a single request. Schedules not found or inaccessible to the user will be
 * omitted from the response.
 *
 * The maximum batch size for this endpoint is 1000.
 *
 * @beta
 *
 * Required Scopes: [api:orchestration-read]
 * URL: /v2/orchestration/schedules/getBatch
 */
export function getBatch($ctx, ...args) {
  return $foundryPlatformFetch($ctx, _getBatch, ...args);
}
const _replace = [2, "/v2/orchestration/schedules/{0}", 3];
/**
 * Replace the Schedule with the specified rid.
 *
 * @beta
 *
 * Required Scopes: [api:orchestration-write]
 * URL: /v2/orchestration/schedules/{scheduleRid}
 */
export function replace($ctx, ...args) {
  return $foundryPlatformFetch($ctx, _replace, ...args);
}
const _run = [1, "/v2/orchestration/schedules/{0}/run"];
/**
 * @public
 *
 * Required Scopes: [api:orchestration-write]
 * URL: /v2/orchestration/schedules/{scheduleRid}/run
 */
export function run($ctx, ...args) {
  return $foundryPlatformFetch($ctx, _run, ...args);
}
const _pause = [1, "/v2/orchestration/schedules/{0}/pause"];
/**
 * @public
 *
 * Required Scopes: [api:orchestration-write]
 * URL: /v2/orchestration/schedules/{scheduleRid}/pause
 */
export function pause($ctx, ...args) {
  return $foundryPlatformFetch($ctx, _pause, ...args);
}
const _unpause = [1, "/v2/orchestration/schedules/{0}/unpause"];
/**
 * @public
 *
 * Required Scopes: [api:orchestration-write]
 * URL: /v2/orchestration/schedules/{scheduleRid}/unpause
 */
export function unpause($ctx, ...args) {
  return $foundryPlatformFetch($ctx, _unpause, ...args);
}
const _getAffectedResources = [1, "/v2/orchestration/schedules/{0}/getAffectedResources", 2];
/**
 * @beta
 *
 * Required Scopes: [api:orchestration-read]
 * URL: /v2/orchestration/schedules/{scheduleRid}/getAffectedResources
 */
export function getAffectedResources($ctx, ...args) {
  return $foundryPlatformFetch($ctx, _getAffectedResources, ...args);
}
const _runs = [0, "/v2/orchestration/schedules/{0}/runs", 2];
/**
 * Get the most recent runs of a Schedule. If no page size is provided, a page size of 100 will be used.
 *
 * @public
 *
 * Required Scopes: [api:orchestration-read]
 * URL: /v2/orchestration/schedules/{scheduleRid}/runs
 */
export function runs($ctx, ...args) {
  return $foundryPlatformFetch($ctx, _runs, ...args);
}
