// @osdk/client/build/esm/object/formatting/formatDateTime.js
/*
 * Copyright 2025 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { resolvePropertyReference } from "./propertyFormattingUtils.js";

/**
 * Formats a date/timestamp value according to the specified formatting rule
 */
export function formatDateTime(value, format, timezone, objectData, locale, userTimezoneOverride) {
  const date = value instanceof Date ? value : new Date(value);
  if (isNaN(date.getTime())) {
    return undefined;
  }
  const resolvedTimezone = resolveTimezone(timezone, objectData, userTimezoneOverride);
  switch (format.type) {
    case "localizedFormat":
      return formatLocalized(date, format, locale, resolvedTimezone);
    case "stringFormat":
      // TODO - pattern formatting
      return undefined;
    default:
      return undefined;
  }
}
const INVALID_DATE_STRING = "Invalid date or timezone";
function formatLocalized(date, format, locale, timezone) {
  if (format.format === "DATE_FORMAT_ISO_INSTANT") {
    return date.toISOString();
  }
  if (format.format === "DATE_FORMAT_RELATIVE_TO_NOW") {
    return formatRelativeToNow(date, locale, timezone);
  }
  const options = getLocalizedFormatOptions(format.format);
  try {
    return new Intl.DateTimeFormat(locale, timezone ? {
      ...options,
      timeZone: timezone
    } : options).format(date);
  } catch (_e) {
    // If a property reference is an invalid timezone we specifically say that it's invalid instead of returning undefined;
    return INVALID_DATE_STRING;
  }
}
function getLocalizedFormatOptions(format) {
  switch (format) {
    case "DATE_FORMAT_DATE":
      return FORMAT_DATE;
    case "DATE_FORMAT_YEAR_AND_MONTH":
      return FORMAT_YEAR_AND_MONTH;
    case "DATE_FORMAT_DATE_TIME":
      return FORMAT_DATE_TIME;
    case "DATE_FORMAT_DATE_TIME_SHORT":
      return FORMAT_DATE_TIME_SHORT;
    case "DATE_FORMAT_TIME":
      return FORMAT_TIME;
    default:
      return {};
  }
}
const FORMAT_DATE = {
  day: "numeric",
  month: "short",
  weekday: "short",
  year: "numeric"
};
const FORMAT_YEAR_AND_MONTH = {
  month: "short",
  year: "numeric"
};
const FORMAT_DATE_TIME = {
  day: "numeric",
  month: "short",
  weekday: "short",
  year: "numeric",
  hour: "numeric",
  minute: "numeric",
  second: "numeric"
};
const FORMAT_DATE_TIME_SHORT = {
  day: "numeric",
  month: "short",
  year: "numeric",
  hour: "numeric",
  minute: "numeric"
};
const FORMAT_TIME = {
  hour: "numeric",
  minute: "numeric",
  second: "numeric"
};
const DATE_TIME_SHORT_WITH_WEEKDAY = {
  day: "numeric",
  month: "short",
  weekday: "short",
  year: "numeric",
  hour: "numeric",
  minute: "numeric"
};
function resolveTimezone(timezone, objectData, userTimezoneOverride) {
  if (userTimezoneOverride != null) {
    return userTimezoneOverride;
  }
  if (timezone == null) {
    return undefined;
  }
  switch (timezone.type) {
    case "user":
      return undefined;
    case "static":
      return resolvePropertyReference(timezone.zoneId, objectData);
    default:
      return undefined;
  }
}
const SECOND_MS = 1_000;
const MINUTE_MS = 60 * SECOND_MS;
const HOUR_MS = 60 * MINUTE_MS;
const DAY_MS = 24 * HOUR_MS;
function formatRelativeToNow(date, locale, timezone) {
  const now = Date.now();
  const diff = date.valueOf() - now;
  const absDiff = Math.abs(diff);

  // More than 1 day - fall back to absolute date formatting
  if (absDiff >= DAY_MS) {
    const dtf = new Intl.DateTimeFormat(locale, {
      ...DATE_TIME_SHORT_WITH_WEEKDAY,
      timeZone: timezone
    });
    return dtf.format(date);
  }

  // Within 1 day - use relative formatting
  const rtf = new Intl.RelativeTimeFormat(locale, {
    numeric: "auto"
  });
  const units = [["hour", HOUR_MS], ["minute", MINUTE_MS], ["second", SECOND_MS]];
  for (const [unit, ms] of units) {
    if (absDiff >= ms) {
      return rtf.format(Math.floor(diff / ms), unit);
    }
  }
  return rtf.format(0, "second");
}
