// @osdk/client/build/browser/observable/internal/sorting/SortingStrategy.js
/*
 * Copyright 2025 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Strategy interface for collection sorting
 */

/**
 * No-operation sorting strategy that preserves original order
 */
export class NoOpSortingStrategy {
  sortCacheKeys(objectCacheKeys) {
    return objectCacheKeys;
  }
}
/**
 * Sorting strategy for OrderBy clauses
 */
export class OrderBySortingStrategy {
  constructor(apiName, orderBy) {
    this.apiName = apiName;
    this.orderBy = orderBy;
    this.sortFns = createOrderBySortFns(orderBy);
  }
  sortCacheKeys(objectCacheKeys, batch) {
    if (Object.keys(this.orderBy).length === 0) {
      return objectCacheKeys;
    }
    return objectCacheKeys.sort((a, b) => {
      for (const sortFn of this.sortFns) {
        const ret = sortFn(batch.read(a)?.value?.$as(this.apiName), batch.read(b)?.value?.$as(this.apiName));
        if (ret !== 0) {
          return ret;
        }
      }
      return 0;
    });
  }
}

/**
 * Creates sort functions for an orderBy clause
 * @param orderBy - The order by clause
 * @returns Array of sort functions
 */
export function createOrderBySortFns(orderBy) {
  return Object.entries(orderBy).map(([key, order]) => {
    return (a, b) => {
      const aValue = a?.[key];
      const bValue = b?.[key];
      if (aValue == null && bValue == null) {
        return 0;
      }
      if (aValue == null) {
        return 1;
      }
      if (bValue == null) {
        return -1;
      }
      const m = order === "asc" ? -1 : 1;
      return aValue < bValue ? m : aValue > bValue ? -m : 0;
    };
  });
}
